""" 
This pipline runs the Measle SSPE project analysis.

Author: Will Hannon â€“ https://www.linkedin.com/in/williamhannon/
"""

#### ----------------------- Imports ----------------------- ####

import pandas as pd 
import itertools
from os.path import join
from snakemake.utils import min_version

min_version('6.0')  # require `snakemake` >= 6.0

#### -------------------- Configuration -------------------- ####

configfile: "config/config.yml"

#### ------------------------ Utils ------------------------ ####

include: "rules/common.smk"

#### ----------------------- Targets ----------------------- ####
SAMPLE_DF = pd.read_csv(config['samples']['file'])
samples = list(set(SAMPLE_DF['Run']))

rule all:
    input: 
        join(config['variant_dir'], "variants.csv"),
        join(config['notebook_dir'], "identify-backgrounds.html"),
        # join(config['haplotype_dir'], "haplotypes.csv"),
        # join(config['spatial_dir'], "bottleneck.csv"), 
        # join(config['bridging_dir'], "bridging_reads.csv"),
        # join(config['split_dir'], "genotyped.csv"),
        join(config['coverage_dir'], "merged.depth")

      
rule clean:
    shell:
        """
        rm -rf logs/
        rm -rf tmp/
        rm -f slurm*.out
        """

#### ------------------------ Rules ------------------------ ####

include: "rules/preprocessing.smk"
include: "rules/format-reference.smk"
include: "rules/align-reads.smk"
include: "rules/call-variants.smk"
include: "rules/download-tools.smk"
include: "rules/call-haplotypes.smk"
include: "rules/calculate-coverage.smk"
include: "rules/split-reads.smk"
include: "rules/process-variants.smk"
