""" 
This pipline runs the Measle SSPE variant analysis.

Author: Will Hannon 
Date: 2022
"""

#### ----------------------- Imports ----------------------- ####

import pandas as pd 
import itertools
from os.path import join
from snakemake.utils import min_version

min_version('7.0')  # require `snakemake` >= 7.0

#### -------------------- Configuration -------------------- ####

configfile: "config/config.yml"

#### ------------------------ Utils ------------------------ ####

include: "rules/common.smk"

#### ----------------------- Targets ----------------------- ####
SAMPLE_DF = pd.read_csv(config['samples']['file'])
samples = list(set(SAMPLE_DF['Run']))

rule all:
    input: 
        join(config['notebook_dir'], "11-cluster-tissues-spatially.html")
      
rule clean:
    shell:
        """
        rm -rf logs/
        rm -rf tmp/
        rm -f slurm*.out
        """

#### ------------------------ Rules ------------------------ ####

# 0. Download tool dependencies not available on conda
include: "rules/download-tools.smk"
# 1. Download, trim, and filter reads
include: "rules/preprocess-reads.smk"
# 2. Align reads to MevChiTok reference genome
include: "rules/first-alignment.smk"
# 3. Build and Align reads an 'SSPE' genome 
include: "rules/second-alignment.smk"
# 4. Calculate coverage of the sequencing reads
include: "rules/calculate-coverage.smk"
# 5. Call variants using the 'SSPE' reference genome
include: "rules/call-variants.smk"
# 6. Count reads that cover multiple variants 
include: "rules/bridging-reads.smk"
# 7. Run R notebooks that cluster and haplotype variants by frequency
include: "rules/process-variants.smk"

