---
title: "5. Validate Haplotype Assignments"
author:
    - "Will Hannon"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
  highlight: tango
  number_sections: no
  theme: default
  toc: yes
  toc_depth: 3
  toc_float:
    collapsed: no
    smooth_scroll: yes
---

The goal of this notebook is take the haplotypes that were identified in the previous notebooks and assigned to `Genome 1` and `Genome 2`, and check for problems like homoplasy or other issues. I'll use bridging reads over pairs of SNPs to look for issues along with checking for violations in our assumptions about the sum of allele frequencies in each tissue specimen. **The output of this notebook is a final set of phased variants with cluster/haplotype assignments.**

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```

```{r Required Packages, message=FALSE, warning=FALSE, echo=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed
packages = c("tidyverse", "foreach", "emdbook", "RColorBrewer", "data.table")

## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```

```{r Inputs, echo=T}

## ==== File paths input data ==== ##

if (exists("snakemake")) {

  # Phased variants with haplotype annotations assigned to background Genome 1 or Genome 2
  updated.data = snakemake@params[["incsv"]]
  
  # Original unfiltered variant calls
  variant.data = snakemake@params[["rawvariants"]]
  
  # Annotations for MeV genes
  annotations.filepath = snakemake@params[["annotations"]]

  # Counts of all reads bridging *all* pairs of SNP positions
  bridging.data = snakemake@params[["bridging"]]
  
  # 'Ancestral' mutation included in the reference
  ancestral.data = snakemake@params[['ancestral']]
  

} else {

  # Phased variants with haplotype annotations assigned to background Genome 1 or Genome 2
  updated.data = "../../results/variants/assigned_variants.csv"
  
  # Original unfiltered variant calls
  variant.data = "../../results/variants/variants.csv"
    
  # Annotations for MeV genes
  annotations.filepath = "../../config/annotations.csv"
 
  # Counts of all reads bridging *all* pairs of SNP positions
  bridging.data = "../../results/bridging/bridging_reads.csv"
  
  # 'Ancestral' mutation included in the reference
  ancestral.data = "../../config/ref/annotated_SSPE_consensus_snps.csv"
  
}

```

```{r Outputs, echo=T}

## ==== File paths output data ==== ##

if (exists("snakemake")) {
  
  # Path to save the validated variants
  output.path = snakemake@params[["outcsv"]]
  
  # Path to save figures
  figure.dir = snakemake@params[['figures']]
  
} else {

  # Path to save the polished assigned variants
  output.path = "../../results/variants/validated_variants.csv"
  
  # Path to save figures
  figure.dir = "../../results/figures/"
  
}

# Make the figure directory if it doesn't exist
if (!file.exists(figure.dir)) {
  dir.create(figure.dir, recursive = TRUE)
  print("Directory created.")
} else {
  print("Directory already exists.")
}

```

## Current Haplotype Clusters

Let's take a look at the re-named and background annotated clusters/haplotypes from the previous notebook `04-assign-haplotype-backgrounds.Rmd`.

```{r Load in Data, echo=F}

# Read in the variants with G1, G2, and G1-1 labeled
labeled.df = read_csv(updated.data, show_col_types = FALSE) %>% 
  mutate(Tissue = case_when(
    Tissue == "UBS" ~ "Upper Brain Stem",
    TRUE ~ Tissue
  ))

# Get the haplotype names
haplotypes.label = labeled.df %>% 
  select(SNP, Haplotype, Background) %>% 
  distinct()

# Expand the mutations to have a frequency for every tissue.
expanded.df = labeled.df %>% 
  select(SNP, Tissue, AF) %>% 
  pivot_wider(names_from = "Tissue", values_from = "AF", values_fill = 0) %>% 
  pivot_longer(cols = !SNP, names_to = "Tissue", values_to = "AF") %>% 
  left_join(., select(labeled.df, c("SNP", "Tissue", "DP")), by = c("SNP", "Tissue")) %>% 
  mutate(DP = if_else(is.na(DP), 0, DP)) %>% 
  left_join(., haplotypes.label, by = "SNP") 

```

```{r Plot Unvalidated Clusters, echo=F, message=F, warning=F, fig.align='center', fig.width=19, fig.height=10}

# Tissue order - roughly by the location in the brain 
tissue.order = c(
  "SSPE 1", 
  "SSPE 2",
  "Frontal Cortex 2", 
  "Frontal Cortex 1", 
  "Frontal Cortex 3", 
  "Parietal Lobe",
  "Temporal Lobe", 
  "Occipital Lobe",
  "Hippocampus",
  "Internal Capsule", 
  "Midbrain", 
  "Upper Brain Stem",
  "Brain Stem",
  "Cerebellum",
  "Cerebellum Nucleus"
)

# Haplotype order - sequential by frequency adn genotype
haplotype.order = c("cluster 1",
                    "cluster 2",
                    "cluster 3",
                    "cluster 4",
                    "cluster 5",
                    "cluster 6",
                    "cluster 7",
                    "cluster 8",
                    "cluster 9",
                    "cluster 10",
                    "cluster 11",
                    "cluster 12",
                    "cluster 13")

expanded.df %>% 
    filter(Haplotype %in% haplotype.order) %>% 
    mutate(Genotype = case_when(
      Background == "genome-1" ~ "G1",
      Background == "genome-2" ~ "G2"
    )) %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue.order), y = AF, col = (Genotype))) +
      geom_line(aes(group = SNP)) +
      facet_wrap(~factor(Haplotype, levels = haplotype.order)) + 
      scale_color_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill = "Genotype", col = "Genotype") +
      theme_bw(20) +
      theme(strip.text.x = element_text(size = 12)) +
      theme(axis.title.x=element_blank()) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(legend.position="bottom")

```

The main question of this notebook is how well our clustering approach has done at identifying 'true' haplotypes. It's possible that some of these clusters contain mutations that are linked through spurious correlations. The hope is that we can resolve this using: 

1. Reads that bridge the position of variants within and between clusters. 
2. The identity of mutations (i.e are they in highly polymorphic genes like Matrix).
3. Violations in our assumptions about the sum of frequencies in each specimen (i.e the 'pigeon-hole' priciple). 

## Bridging Reads

First, I'll use reads that bridge all pairs of positions where there are reads covering both position simultaneously. The question is whether these reads can help confirm the halplotypes identities and tell us whether haplotypes are linked to one another. 

```{r Import Bridging Reads, echo=F}
# Bridging reads
bridging.df = read_csv(bridging.data, show_col_types = FALSE) %>% 
  mutate(Tissue = case_when(
      Tissue == "UBS" ~ "Upper Brain Stem",
      TRUE ~ Tissue
    ))

head(bridging.df) %>% 
  kable(., caption = "Bridging SNPs")
  
```

The columns are as follows: `snp_1` is the position in the genome of one SNP, `snp_2` is the position of the second SNP, `00` are reads that overlap both positions but have neither SNP, `10` are reads that have only `snp_1`, `01` are reads that have only `snp_2`, and `11` are reads with both SNPs. 

Below, I'll calculate the probability that any two SNPs are either linked or forbidden over a log-spaced range of `taus`. 

```{r Calculate Linked and Forbidden SNPs, echo=T}

taus = lseq(0.005, .5, length.out = 50) # Log spaced values for tau
min_obsv = 100 # Minimum number of bridging reads (of any type)
min_snp_obsv = 25 # Minimum number of bridging reads with SNPs
min.sig = 0.05 # Minimum significance before multiple testing correction
  
# Assign to genome for each value of tau
SNPassignmentsByTau = foreach(tau = taus, .combine = "rbind")%do%{

  bridging.df %>% 
    mutate(total = `00` + `10` + `01` + `11`) %>% 
    filter(total >= min_obsv) %>% 
    filter(((`01` + `11`) >= min_snp_obsv) & ((`10` + `11`) >= min_snp_obsv)) %>% 
    # Multiple testing done by tissue, not polymorphic position
    group_by(snp_1, snp_2) %>% 
    # Calculate the probability of being linked
    mutate(linked = pbinom(`11`, total, tau, lower.tail = F), 
           linked_sig = linked < min.sig/n()) %>% 
      # Calculate the probability of being forbidden
    mutate(forbidden = pbinom(`11`, total, tau, lower.tail = T), 
           forbidden_sig = forbidden < min.sig/n()) %>% 
    mutate(linked_sig = ifelse(total == 0, NA, linked_sig),
           forbidden_sig = ifelse(total == 0, NA, forbidden_sig)) %>% 
    ungroup() %>% 
    mutate(tau = tau)

}

SNPassignmentsByTau %>% 
  select(!c("total", "linked", "forbidden")) %>% 
  head() %>% 
  kable(., caption = "Linkage Info")

```

Then, I'll make a function that takes in the name of a cluster, the data frame of SNPs and their frequencies, the data frame of linked/forbidden SNPs as determined by bridging reads, and a minimum frequency at which both of the SNPs must be present in a given tissue's bridging reads. 

```{r Function to Summarise Bridging Reads, echo=T}

rank_bridging_reads = function(cluster.ID, snp.df, tau.df, min.freq = 0.1) {
    
  tau.df.by.tissue = tau.df %>% 
    mutate(snp_1_freq = (`10` + `11`) / total, 
           snp_2_freq = (`01` + `11`) / total) %>% 
    filter((snp_1_freq >= min.freq) & (snp_2_freq >= min.freq)) %>% 
    select(snp_1, snp_2, linked_sig, forbidden_sig, tau, Tissue) %>% 
    group_by(snp_1, snp_2, tau) %>% 
    mutate(linked = sum(linked_sig),
           forbidden = sum(forbidden_sig)) %>% 
    select(!c("Tissue", "linked_sig", "forbidden_sig")) %>% 
    distinct() %>% 
    ungroup()
  
  cluster.pos = snp.df %>% 
    filter(Haplotype == cluster.ID) %>% 
    pull(POS) %>% 
    unique()
  
  snp.1.df = tau.df.by.tissue %>% 
    filter(snp_1 %in% cluster.pos) %>% 
    dplyr::rename("cluster_snp" = snp_1, "query_snp" = snp_2)
  
  snp.2.df = tau.df.by.tissue %>% 
    filter(snp_2 %in% cluster.pos) %>% 
    dplyr::rename("cluster_snp" = snp_2, "query_snp" = snp_1)
  
  query.df = rbind(snp.1.df, snp.2.df) %>% 
    left_join(., distinct(select(snp.df, POS, Haplotype)), by = c("query_snp" = "POS")) %>% 
    filter(!Haplotype %in% c("subclonal", "both")) %>% 
    arrange(Haplotype, cluster_snp, tau) %>% 
    filter((linked > 0) | (forbidden > 0))
  
  return(query.df)
  
}

```

I'll use this later when cleaning up individual clusters. 

## Investigate Individual Clusters

There are a few haplotypes that have a significant amount of internal variation even though our method clustered all of these SNPs together. I'll go through each of these haplotypes/clusters independently. 

### Cluster 6

Cluster 6 appears to be a set of SNPs on the background of `Genome 1` in the Frontal Cortex 2 specimen. However, since these SNPs are only present at appreciable frequency in a single tissue, we can't as clearly link them by their correlation in frequency alone.

```{r Cluster 6, fig.align='center', fig.width=8, fig.height=5, echo=F}

expanded.df %>% 
    filter(Haplotype %in% c("cluster 6")) %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue.order), y = AF,)) +
      geom_line(aes(group = SNP)) +
      facet_wrap(~factor(Haplotype, levels = haplotype.order)) + 
      scale_color_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))

```

Another issue with this cluster is that many of the mutations are in the Matrix gene. This gene has a stop codon fairly near the start of the protein and has been allowed to neutrally accumulate a large number of mutations over time. This is exacerbated by the mutagenic activity of ADAR. The issue with this is that mutations can arise multiple times on multiple haplotype backgrounds. Controlling for this is beyond the scope of our haplotyping approach. 

```{r Percentage of Matrix SNPs, echo = F}

labeled.df %>% 
  filter (Haplotype == "cluster 6") %>% 
  select(SNP, Gene) %>% 
  distinct() %>% 
  group_by(Gene) %>% 
  count() %>% 
  arrange(-n) %>% 
  kable(., caption = "Cluster 6 SNPs per Gene")


```

We can see that if we drop only the Matrix mutations from this cluster, the peaks in frequency in tissues where cluster 6 should presumably be missing are mostly gone.

```{r Cluster 6 Non-matrix, fig.align='center', fig.width=8, fig.height=5, echo=F}

expanded.df %>% 
    filter(Haplotype %in% c("cluster 6")) %>%
    left_join(., distinct(select(labeled.df, SNP, Gene))) %>% 
    filter(Gene != "M") %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue.order), y = AF,)) +
      geom_line(aes(group = SNP)) +
      facet_wrap(~factor(Haplotype, levels = haplotype.order)) + 
      scale_color_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))


```

One solution to this issue, short of removing all Matrix mutations, is to remove the matrix mutations that are *also* linked to genome-1-1 or genome-2. This makes the correlation much better and is a more principled approach to refining this haplotype.

```{r Exclude Reccurent SNPs from Cluster 6, fig.align='center', fig.width=8, fig.height=5, echo=F}

exclude.from.cluster.6 = rank_bridging_reads("cluster 6", labeled.df, SNPassignmentsByTau, min.freq = 0.1) %>% 
  filter(linked >= 1) %>% 
  filter(Haplotype %in% c("genome-1-1", "genome-2")) %>% 
  select(cluster_snp) %>% 
  pull() %>% 
  unique()

expanded.df %>% 
    filter(Haplotype %in% c("cluster 6")) %>%
    left_join(., distinct(select(labeled.df, SNP, Gene, POS))) %>% 
    filter(!POS %in% exclude.from.cluster.6) %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue.order), y = AF,)) +
      geom_line(aes(group = SNP)) +
      facet_wrap(~factor(Haplotype, levels = haplotype.order)) + 
      scale_color_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))

```

```{r Excluding SNPs from Cluster 6, echo = T}

print(paste("Excluding", length(exclude.from.cluster.6), "SNPs from cluster 6"))

labeled.df = labeled.df %>% 
  mutate(Background = case_when(POS %in% exclude.from.cluster.6 ~ "subclonal",
                                TRUE ~ Background)) %>% 
  mutate(Haplotype = case_when(POS %in% exclude.from.cluster.6 ~ "subclonal",
                                TRUE ~ Haplotype))

```

### Cluster 12

Now let's see about cluster 12 These are SNPs present in the lower brain where the read depth is pretty low, and the correlation is somewhat loose. 

```{r Cluster 12, fig.align='center', fig.width=8, fig.height=5, echo=F}

expanded.df %>% 
    filter(Haplotype %in% c("cluster 12")) %>%
    left_join(., distinct(select(labeled.df, SNP, Gene, POS))) %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue.order), y = AF,)) +
      geom_line(aes(group = SNP)) +
      facet_wrap(~factor(Haplotype, levels = haplotype.order)) + 
      scale_color_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))

```

This cluster is half matrix mutations and half non-matrix mutations

```{r Percentage of Matrix SNPs Cluster 12, echo=F}

labeled.df %>% 
  filter (Haplotype == "cluster 12") %>% 
  select(SNP, Gene) %>%
  distinct() %>% 
  group_by(Gene) %>% 
  count() %>% 
  arrange(-n) %>% 
  kable(., caption = "Cluster 12 SNPs per Gene")

```

Below, I'll clean up cluster 12 by removing SNPs that are linked to genome-1 (since cluster 12 is a genome-2 cluster) and removing Matrix mutations unless they're linked to other cluster 12 mutations or genome-2. Finally, there is a very low depth F mutation that visibly looks like an outlier. I'll remove this as well.

```{r Clean Up Cluster 12, echo=T}

linked.to.cluster.12 = rank_bridging_reads("cluster 12", labeled.df, SNPassignmentsByTau, min.freq = 0.1) %>% 
  filter(linked >= 1) %>% 
  filter(Haplotype %in% c("cluster 12", "genome-2")) %>% 
  select(cluster_snp) %>% 
  pull() %>% 
  unique()

unlinked.from.cluster.12 = rank_bridging_reads("cluster 12", labeled.df, SNPassignmentsByTau, min.freq = 0.1) %>% 
  filter(linked >= 1) %>% 
  filter(Haplotype %in% c("genome-1-1", "genome-1")) %>% 
  select(cluster_snp) %>% 
  pull() %>% 
  unique()

keep.in.cluster.12 = linked.to.cluster.12[which(!linked.to.cluster.12 %in% unlinked.from.cluster.12)]

non.matrix.cluster.12 = labeled.df %>% 
  filter(Haplotype == "cluster 12") %>% 
  filter(Gene != "M") %>% 
  pull(POS) %>% 
  unique()

keep.in.cluster.12 = c(keep.in.cluster.12, non.matrix.cluster.12)

low.depth.outlier = 4923

keep.in.cluster.12 = keep.in.cluster.12[which(!keep.in.cluster.12 %in% low.depth.outlier)]

cluster.12.snps = labeled.df %>% 
  filter(Haplotype %in% c("cluster 12")) %>%
  pull(POS) %>% 
  unique()  

exclude.from.cluster.12 = cluster.12.snps[which(!cluster.12.snps %in% keep.in.cluster.12)]

labeled.df = labeled.df %>% 
  mutate(Background = case_when((POS %in% exclude.from.cluster.12) ~ "subclonal",
                                TRUE ~ Background)) %>% 
  mutate(Haplotype = case_when((POS %in% exclude.from.cluster.12) ~ "subclonal",
                                TRUE ~ Haplotype))

```

```{r Cluster 12 Clean, fig.align='center', fig.width=8, fig.height=5, echo=F}

expanded.df %>% 
    filter(Haplotype %in% c("cluster 12")) %>%
    left_join(., distinct(select(labeled.df, SNP, Gene, POS))) %>% 
    filter((!POS %in% exclude.from.cluster.12)) %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue.order), y = AF,)) +
      geom_line(aes(group = SNP)) +
      facet_wrap(~factor(Haplotype, levels = haplotype.order)) + 
      scale_color_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))
```

### Cluster 2 

Finally, Cluster 2 contains a set of mutations that are missing from the Cerebellum due to coverage. These appear as a dip in the mean Cluster 2 frequency. Let's amend this using the original variant calls. 

```{r Missing Cluster 2 SNPs, echo=T}

# Get the raw original variants not filtered on depth
variant.df = read_csv(variant.data, show_col_types = F)

all.c2.snps = labeled.df %>%
  filter(Haplotype == "cluster 2") %>% 
  pull(SNP) %>%
  unique()

cerebellum.c2.snps = labeled.df %>%
  filter(Haplotype == "cluster 2" & Tissue == "Cerebellum") %>% 
  pull(SNP) %>%
  unique()

# These SNPS are not found in the Cerebellum, likely due to depth
missing.cerebellum.c2.snps = all.c2.snps[!(all.c2.snps %in% cerebellum.c2.snps)]

# Are they actually missing? 
variant.df %>% 
  mutate(SNP = paste0(REF, POS, ALT)) %>% 
  filter(SNP %in% missing.cerebellum.c2.snps & Tissue == "Cerebellum") %>% 
  select(SNP, AF, DP, Tissue, Caller) %>% 
  kable(., caption = "Missing SNPS")

```

These SNPs aren't missing, they are just low depth and were filtered out due to this. 

```{r Amend Cluster 2, echo = T}

# Get the SNPs from the original data frame
cluster.2.snps = variant.df %>%
  mutate(SNP = paste0(REF, POS, ALT)) %>% 
  filter(SNP %in% missing.cerebellum.c2.snps & Tissue == "Cerebellum" & Caller == "lofreq") %>% 
  select(!c(Caller, Virus, Host, R1, R2, Sample)) %>% 
  mutate(Cluster = "cluster 2",
         Background = "genome-1",
         Haplotype = "cluster 2", 
         Gene = Gene_Name)
  
# Add them back to the data
labeled.df = rbind(labeled.df, cluster.2.snps)

```

What do these finalized clusters look like? 

```{r Ext Figure 4, message=F, warning=F, fig.align='center', fig.width=19, fig.height=10, echo=F}

# Get the haplotype names
haplotypes.label = labeled.df %>% 
  select(SNP, Haplotype, Background) %>% 
  distinct()

# Expand the mutations to have a frequency for every tissue.
expanded.df = labeled.df %>% 
  select(SNP, Tissue, AF) %>% 
  pivot_wider(names_from = "Tissue", values_from = "AF", values_fill = 0) %>% 
  pivot_longer(cols = !SNP, names_to = "Tissue", values_to = "AF") %>% 
  left_join(., select(labeled.df, c("SNP", "Tissue", "DP")), by = c("SNP", "Tissue")) %>% 
  mutate(DP = if_else(is.na(DP), 0, DP)) %>% 
  left_join(., haplotypes.label, by = "SNP") 

expanded.df %>% 
    filter(!Haplotype %in% c("fixed", "both", "subclonal", "genome-1", "genome-2", "genome-1-1")) %>% 
    mutate(Genotype = case_when(
      Background == "genome-1" ~ "G1",
      Background == "genome-2" ~ "G2"
    )) %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue.order), y = AF, col = (Genotype))) +
      geom_line(aes(group = SNP)) +
      facet_wrap(~factor(Haplotype, levels = haplotype.order)) + 
      scale_color_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Genotype") +
      theme_bw(20) +
      theme(strip.text.x = element_text(size = 12)) +
      theme(axis.title.x=element_blank()) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(legend.position="bottom")

ggsave(paste(figure.dir, "final-cluster-structures.png", sep=""), width = 19, height = 10)

```

## Mutation distribution

What are the distribution of these haplotypes in the genome? 

```{r Annotations, message=FALSE, warning=FALSE, echo=F}

# Import genome annotations
annotations.df = read_csv(annotations.filepath, show_col_types = FALSE)

# Get the interval positions for these genes 
N = annotations.df %>% dplyr::filter(Locus == "N")
P.V.C = annotations.df %>% dplyr::filter(`Protein Name` == "phosphoprotein")
M = annotations.df %>% dplyr::filter(Locus == "M")
F. = annotations.df %>% dplyr::filter(Locus == "F")
H = annotations.df %>% dplyr::filter(Locus == "H")
L = annotations.df %>% dplyr::filter(Locus == "L")

gene_order = c("N", "P", "M", "F", "H", "L")

```

```{r Figure 6, message=F, warnings=F, fig.align='center', fig.width=15, fig.height=10}

# Define the order of haplotypes on the y-axis
figure.haplotype.order = c("SSPE Ancestor",
                            "G-01",
                            "G-FC2",
                            "G-1",
                            "cluster 1",
                            "cluster 1a",
                            "cluster 2",
                            "cluster 3",
                            "cluster 4",
                            "cluster 5",
                            "cluster 6",
                            "G-2",
                            "cluster 7",
                            "cluster 8",
                            "cluster 9",
                            "cluster 10",
                            "cluster 11")

  
# Load in the 'ancestral' SNPs that are included in the consensus patient-specific reference
ancestral.snps = read_csv(ancestral.data, show_col_types = FALSE) %>% 
  mutate(ADAR = case_when(
    REF == 'T' & ALT == 'C' ~ TRUE, 
    REF == 'A' & ALT == 'G' ~ TRUE,
    TRUE ~ FALSE
  )) %>% 
  mutate(Effect = case_when(
    is.na(WT_AA) ~ "Synonymous",
    WT_AA == MUT_AA ~ "Synonymous",
    WT_AA != MUT_AA ~ "Missense"
  )) %>% 
  select(POS, REF, ALT, Gene, ADAR, Effect) %>% 
  mutate(Haplotype = "SSPE Ancestor",
         Background = "SSPE Ancestor")

# Prepare the data for plotting
distribution.df = labeled.df %>% 
  select(Background, Haplotype, Gene, Effect, POS, REF, ALT) %>% 
  mutate(Gene = if_else(Gene == "P/V/C", "P", Gene)) %>% 
  mutate(Effect = if_else(is.na(Effect), "Synonymous", Effect)) %>% 
  mutate(ADAR = case_when(
    REF == 'T' & ALT == 'C' ~ TRUE, 
    REF == 'A' & ALT == 'G' ~ TRUE,
    TRUE ~ FALSE
  )) %>%  
  mutate(Haplotype = case_when(
    Haplotype == "genome-1" ~ "G-01",
    Haplotype == "genome-1-1" ~ "G-1",
    Haplotype == "cluster 6" ~ "G-FC2",
    Haplotype == "genome-2" ~ "G-2",
    Haplotype == "cluster 1" ~  "cluster 1",
    Haplotype == "cluster 2" ~  "cluster 2",
    Haplotype == "cluster 3" ~  "cluster 3",
    Haplotype == "cluster 4" ~  "cluster 4",
    Haplotype == "cluster 5" ~  "cluster 1a",
    Haplotype == "cluster 7" ~  "cluster 5",
    Haplotype == "cluster 8" ~  "cluster 6",
    Haplotype == "cluster 9" ~  "cluster 7",
    Haplotype == "cluster 10" ~  "cluster 8",
    Haplotype == "cluster 11" ~  "cluster 9",
    Haplotype == "cluster 12" ~  "cluster 10",
    Haplotype == "cluster 13" ~  "cluster 11",
    TRUE ~ Haplotype
  )) %>% 
  distinct()  %>% 
  filter(!Haplotype %in% c('subclonal', 'both')) %>% 
  rbind(., ancestral.snps)


# Haplotype Colors
haplotype.colors =  c("#000000", "dark grey", "dark grey", "dark grey", "dark grey", "dark grey", "dark grey", "dark grey",  "dark grey", "dark grey", "dark grey", "dark grey", "dark grey", "dark grey" ,"dark grey", "dark grey", "dark grey", "dark grey", "dark grey")
names(haplotype.colors) = c("SSPE Ancestor", "G-01", "G-FC2", "G-1", "G-2", "cluster 1", "cluster 1a", "cluster 2", "cluster 3", "cluster 4", "cluster 5", "cluster 6", "cluster 7", "cluster 8", "cluster 9", "cluster 10", "cluster 11")

# Plot the distribution of mutations
distribution.df %>% 
  ggplot(aes(x = POS, y = factor(Haplotype, levels = rev(figure.haplotype.order)), col = Haplotype)) +
    geom_point(aes(shape = ADAR), size = 8) +
    xlab("Position") +
  
  
    scale_x_continuous(limits=c(-180, 16180), breaks=c(0, 5000, 10000, 15000)) +
    scale_color_manual(values = haplotype.colors, guide = 'none') +
    scale_shape_manual(name="A to G / U to C", values = c(108, 4)) +
    
    # == Annotations of genes == #
    
    # 3' Start
    annotate(geom = "text", x = (N$Start - 180), y = .05, label = "3'", size = 8) +

    # Nucleoprotein
    annotate("rect", xmin = N$Start-1, xmax = N$Stop, ymin = -.5, ymax = .5 ,
               alpha = .5, fill = "#d97634", col = "black", size = .1) +
    annotate(geom = "text", x = (N$Stop + N$Start)/2, y = .05, label = "N", size = 8) +
    annotate("rect", xmin = N$Start-1, xmax = N$Stop, ymin = -.5, ymax = 18.5  ,
                   alpha = .05, fill = "#d97634", size = .1) +  
    # P/V/C
    annotate("rect", xmin = P.V.C$Start, xmax = P.V.C$Stop, ymin = -.5, ymax = .5 ,
               alpha = .5, fill = "#fcff3d", col = "black", size = .1) +
    annotate(geom = "text", x = (P.V.C$Stop + P.V.C$Start)/2, y = .05, label = "P/V/C", size = 8) +
    annotate("rect", xmin = P.V.C$Start, xmax = P.V.C$Stop, ymin = -.5, ymax = 18.5  ,
               alpha = .05, fill = "#fcff3d", size = .1) +
    # Matrix protein
    annotate("rect", xmin = M$Start, xmax = M$Stop, ymin = -.5, ymax = .5 ,
               alpha = .5, fill = "#0db51b", col = "black", size = .1) +
    annotate(geom = "text", x = (M$Stop + M$Start)/2, y = .05, label = "M", size = 8)  +
    annotate("rect", xmin = M$Start, xmax = M$Stop, ymin = -.5, ymax = 18.5  ,
               alpha = .05, fill = "#0db51b", size = .1) +
    # Fusion protein
    annotate("rect", xmin = F.$Start, xmax = F.$Stop, ymin = -.5, ymax = .5 ,
               alpha = .5, fill = "#6A3D9A", col = "black", size = .1) +
    annotate(geom = "text", x = (F.$Stop + F.$Start)/2, y = .05, label = "F", size = 8) +
    annotate("rect", xmin = F.$Start, xmax = F.$Stop, ymin = -.5, ymax = 18.5  ,
               alpha = .05, fill = "#6A3D9A", size = .1) +
    # Hemagluttinin
    annotate("rect", xmin = H$Start, xmax = H$Stop, ymin = -.5, ymax = .5 ,
               alpha = .5, fill = "#0251d9", col = "black", size = .1) +
    annotate(geom = "text", x = (H$Stop + H$Start)/2, y = .05, label = "H", size = 8) +
    annotate("rect", xmin = H$Start, xmax = H$Stop, ymin = -.5, ymax = 18.5 ,
               alpha = .05, fill = "#0251d9", size = .1) +
    # Large protein
    annotate("rect", xmin = L$Start, xmax = L$Stop, ymin = -.5, ymax = .5 ,
               alpha = .5, fill = "#bf1515", col = "black", size = .1) +
    annotate(geom = "text", x = (L$Stop + L$Start)/2, y = .05, label = "L", size = 8) +
    annotate("rect", xmin = L$Start, xmax = L$Stop, ymin = -.5, ymax = 18.5 ,
               alpha = .05, fill = "#bf1515", size = .1) +
  
    # 5' Start
    annotate(geom = "text", x = (L$Stop + 180), y =  .05, label = "5'", size = 8) +

  
    theme_classic() +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.background = element_rect(color = NA)) +
    theme(text=element_text(size=24)) + 
    theme(axis.text.x = element_text(size=28)) + 
    theme(axis.title.x = element_text(size=30)) + 
    theme(plot.title = element_text(hjust = 0.5))  +
    theme(axis.title.y = element_blank()) +
    theme(legend.position = "bottom")  

ggsave(paste(figure.dir, "figure-7a.png", sep=""), width = 15, height = 10, dpi = 300)


```

Finally, I'll write out these validated and renamed clusters to a new file. 

```{r Write Output, echo=F}

# Write the validated labeled data
print(paste("Writing out the variants to", output.path))
write_csv(labeled.df, output.path)

```

