---
title: "5. Validate Haplotype Assignments"
author:
    - "Will Hannon"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
  highlight: tango
  number_sections: no
  theme: default
  toc: yes
  toc_depth: 3
  toc_float:
    collapsed: no
    smooth_scroll: yes
---

The goal of this notebook is to validate and organize the nomenclature for the subclonal haplotypes that were identified and assigned to `Genome 1` or `Genome 2` in the previous notebooks. I'll use the bridging reads over any given pair of SNPs to do this validation along with checking for violations in our assumptions about the sum of allele frequencies in each tissue specimen. **The output of this notebook is a final set of phased variants with cluster/haplotype assignments.**

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```

```{r Required Packages, message=FALSE, warning=FALSE, echo=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed
packages = c("tidyverse", "foreach", "emdbook", "RColorBrewer", "data.table", "ggnetwork", "network")

## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```

```{r Inputs, echo=T}

## ==== File paths input data ==== ##

if (exists("snakemake")) {

  # Data from lofreq and varscan labeled with subclonal haplotypes
  updated.data = snakemake@params[["incsv"]]
  
  # Data from the count of bridging reads
  bridging.data = snakemake@input[[1]]
    
  # Annotations 
  annotations.filepath = snakemake@params[["annotations"]]
  
  # 'Ancestral' mutation included in the reference
  ancestral.data = snakemake@params[['ancestral']]
  
  # G1 or G2 linkage information for all phased SNPs including disagreements 
  background.data = snakemake@params[['backgrounds']]

} else {

  # Data from lofreq and varscan labeled with subclonal haplotypes
  updated.data = "../../results/variants/assigned_variants.csv"
  
  # Data from the count of bridging reads
  bridging.data = "../../results/bridging/bridging_reads.csv"
  
  # Annotations 
  annotations.filepath = "../../config/annotations.csv"
  
  # 'Ancestral' mutation included in the reference
  ancestral.data = "../../config/ref/annotated_SSPE_consensus_snps.csv"
  
  # G1 or G2 linkage information for all phased SNPs including disagreements 
  background.data = "../../results/bridging/cluster_snp_backgrounds.csv"
  
}

```

```{r Outputs, echo=T}

## ==== File paths output data ==== ##

if (exists("snakemake")) {
  
  # Path to save the polished assigned variants
  output.path = snakemake@params[["outcsv"]]
  
} else {

  # Path to save the polished assigned variants
  output.path = "../../results/variants/validated_variants.csv"
}


```

## Current Haplotype Clusters

First, using the previous notebook (`genotype-subclonal-snps.Rmd`), let's assign a putative background lineage (`Genome 1` or `Genome 2`) to each of the major clusters.

```{r Assign Genotypes, echo=T}

# Read in the variants with G1, G2, and G1-1 labeled
labeled.df = read_csv(updated.data, show_col_types = FALSE) 

```

Now, I'll plot all of the current haplotypes colored by whether they belong to either lineage.

```{r All Polished Haplotypes, message=F, warning=F, fig.align='center', fig.width=19, fig.height=10, echo=T}

haplotypes.label = labeled.df %>% 
  select(SNP, Haplotype, Background) %>% 
  distinct()

# Expand the mutations to have a frequency for every tissue.
expanded.df = labeled.df %>% 
  select(SNP, Tissue, AF) %>% 
  pivot_wider(names_from = "Tissue", values_from = "AF", values_fill = 0) %>% 
  pivot_longer(cols = !SNP, names_to = "Tissue", values_to = "AF") %>% 
  left_join(., select(labeled.df, c("SNP", "Tissue", "DP")), by = c("SNP", "Tissue")) %>% 
  mutate(DP = if_else(is.na(DP), 0, DP)) %>% 
  left_join(., haplotypes.label, by = "SNP") 

# Get the mean frequency of the major genomes 
tissue.mean = labeled.df %>% 
  filter(Haplotype %in% c("genome-1", "genome-2")) %>% 
  group_by(Tissue, Haplotype) %>% 
  summarize(AF.mean = mean(AF, na.rm = TRUE), 
            SD = sd(AF, na.rm = TRUE),
            N = n()) %>% 
  mutate(SE = SD / sqrt(N),
         Lower.CI = qt(1 - (0.05 / 2), N - 1) * SE,
         Upper.CI = qt(1 - (0.05 / 2), N - 1) * SE) %>% 
  rename("AF" = AF.mean)

#  Mean frequency in each haplotype
haplotype.mean = expanded.df %>% 
  group_by(Tissue, Haplotype) %>% 
  summarize(AF = mean(AF))

# Tissue order - roughly by the location in the brain 
tissue_order = c(
  "SSPE 1", 
  "SSPE 2",
  "Frontal Cortex 2", 
  "Frontal Cortex 1", 
  "Frontal Cortex 3", 
  "Parietal Lobe",
  "Temporal Lobe", 
  "Occipital Lobe",
  "Hippocampus",
  "Internal Capsule", 
  "Midbrain", 
  "UBS",
  "Brain Stem",
  "Cerebellum",
  "Cerebellum Nucleus"
)

haplotype.order = c("cluster 1",
                    "cluster 2",
                    "cluster 3",
                    "cluster 4",
                    "cluster 5",
                    "cluster 6",
                    "cluster 7",
                    "cluster 8",
                    "cluster 9",
                    "cluster 10",
                    "cluster 11",
                    "cluster 12",
                    "cluster 13",
                    "cluster 14", 
                    "cluster 15",
                    "cluster 16",
                    "genome-1-1")

expanded.df %>% 
    filter(!Haplotype %in% c("fixed", "both", "subclonal", "genome-1", "genome-2")) %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue_order), y = AF, col = (Background))) +
      geom_line(aes(group = SNP)) +
      facet_wrap(~factor(Haplotype, levels = haplotype.order)) + 
      scale_color_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))


```

The main question of this notebook is how well our clustering approach has done at identifying 'true' haplotypes. It's possible that some of these clusters contain mutations that are linked through spurious correlations. The hope is that we can resolve some of this using: 

1. Reads that bridge the position of variants within and between clusters. 
2. The identity of mutations (i.e are they in highly polymorphic genes like Matrix).
3. Violations in our assumptions about the sum of frequencies in each specimen (i.e the 'pigeon-hole' priciple). 

## Bridging Reads

First, I'll use reads that bridge all pairs of positions where there are reads covering both position simultaneously. The question is whether these reads can help confirm the halplotypes identities and tell us whether haplotypes are linked to one another. 

```{r Upload Bridging Reads, echo=T}
# Bridging reads
bridging.df = read_csv(bridging.data, show_col_types = FALSE)
head(bridging.df)
```

The columns are as follows: `snp_1` is the position in the genome of one SNP, `snp_2` is the position of the second SNP, `00` are reads that overlap both positions but have neither SNP, `10` are reads that have only `snp_1`, `01` are reads that have only `snp_2`, and `11` are reads with both SNPs. 

Below, I'll calculate the probability that any two SNPs are either linked or forbidden over a log-spaced range of `taus`. 

```{r Calculate Linked and Forbidden SNPs, echo=T}

taus = lseq(0.005, .5, length.out = 50) # Log spaced values for tau
min_obsv = 100 # Minimum number of bridging reads (of any type)
min_snp_obsv = 25 # Minimum number of bridging reads with SNPs
min.sig = 0.05 # Minimum significance before multiple testing correction
  
# Assign to genome for each value of tau
SNPassignmentsByTau = foreach(tau = taus, .combine = "rbind")%do%{

  bridging.df %>% 
    mutate(total = `00` + `10` + `01` + `11`) %>% 
    filter(total >= min_obsv) %>% 
    filter(((`01` + `11`) >= min_snp_obsv) & ((`10` + `11`) >= min_snp_obsv)) %>% 
    # Multiple testing done by tissue, not polymorphic position
    group_by(snp_1, snp_2) %>% 
    # Calculate the probability of being linked
    mutate(linked = pbinom(`11`, total, tau, lower.tail = F), 
           linked_sig = linked < min.sig/n()) %>% 
      # Calculate the probability of being forbidden
    mutate(forbidden = pbinom(`11`, total, tau, lower.tail = T), 
           forbidden_sig = forbidden < min.sig/n()) %>% 
    mutate(linked_sig = ifelse(total == 0, NA, linked_sig),
           forbidden_sig = ifelse(total == 0, NA, forbidden_sig)) %>% 
    ungroup() %>% 
    mutate(tau = tau)

}

SNPassignmentsByTau %>% 
  select(!c("total", "linked", "forbidden")) %>% 
  head()

```

Then, I'll make a function that takes in the name of a cluster, the data frame of SNPs and their frequencies, the data frame of linked/forbidden SNPs as determined by bridging reads, and a minimum frequency at which both of the SNPs must be present in a given tissue's bridging reads. 

```{r Function to Summarise Bridging Reads, echo=T}

rank_bridging_reads = function(cluster.ID, snp.df, tau.df, min.freq = 0.1) {
    
  tau.df.by.tissue = tau.df %>% 
    mutate(snp_1_freq = (`10` + `11`) / total, 
           snp_2_freq = (`01` + `11`) / total) %>% 
    filter((snp_1_freq >= min.freq) & (snp_2_freq >= min.freq)) %>% 
    select(snp_1, snp_2, linked_sig, forbidden_sig, tau, Tissue) %>% 
    group_by(snp_1, snp_2, tau) %>% 
    mutate(linked = sum(linked_sig),
           forbidden = sum(forbidden_sig)) %>% 
    select(!c("Tissue", "linked_sig", "forbidden_sig")) %>% 
    distinct() %>% 
    ungroup()
  
  cluster.pos = snp.df %>% 
    filter(Haplotype == cluster.ID) %>% 
    pull(POS) %>% 
    unique()
  
  snp.1.df = tau.df.by.tissue %>% 
    filter(snp_1 %in% cluster.pos) %>% 
    dplyr::rename("cluster_snp" = snp_1, "query_snp" = snp_2)
  
  snp.2.df = tau.df.by.tissue %>% 
    filter(snp_2 %in% cluster.pos) %>% 
    dplyr::rename("cluster_snp" = snp_2, "query_snp" = snp_1)
  
  query.df = rbind(snp.1.df, snp.2.df) %>% 
    left_join(., distinct(select(snp.df, POS, Haplotype)), by = c("query_snp" = "POS")) %>% 
    filter(!Haplotype %in% c("subclonal", "both")) %>% 
    arrange(Haplotype, cluster_snp, tau) %>% 
    filter((linked > 0) | (forbidden > 0))
  
  return(query.df)
  
}

```

For all of the clusters plotted above, I'll print every known haplotype (including SNPs in G1 and G2) that is possibly linked to that cluster. As long as a SNP with a known haplotype is linked to the cluster's SNP in more than one tissue where both SNPs are present in greater than 10% of bridging reads, it will get printed below.

```{r Print Linked Matches, echo=T}

clusters = labeled.df %>% 
  filter(!Haplotype %in% c("genome-1", "genome-2", "both", "fixed", "subclonal")) %>% 
  pull(Haplotype) %>% 
  unique()

for (cluster.id in clusters) {
  
  linked.matches = rank_bridging_reads(cluster.id, labeled.df, SNPassignmentsByTau, min.freq = 0.1) %>% 
    filter(linked >= 1) %>% 
    pull(Haplotype) %>% 
    unique()
  
  print(paste(cluster.id, "is linked to:"))
  print(paste(linked.matches, collapse=", "))
  print("----")
  
}

```

Most of these conform with our expectations regarding which backgrounds we think each haplotype cluster belongs to. The exceptions are cluster 10, 11, 13, 15, and 6. Let's investigate each of these independently. 

## Investigate Individual Clusters

I'll go through each cluster that violates our assumptions about which background it's on independently. 

### Cluster 15

Cluster 15 is a *very* low frequency cluster, there is good amount of internal variation, and there are only two SNPs. 

```{r Cluster 15, fig.align='center', fig.width=8, fig.height=5, echo=T}

expanded.df %>% 
    filter(Haplotype %in% c("cluster 15")) %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue_order), y = AF,)) +
      geom_line(aes(group = SNP)) +
      facet_wrap(~factor(Haplotype, levels = haplotype.order)) + 
      scale_color_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))

```

I don't know if we can really be that confident in this cluster, so let's remove it. It's only 2 SNPs, so we're not going to lose much information either way.

```{r Remove cluster 15, echo=T}

labeled.df = labeled.df %>% 
  mutate(Background = case_when(Haplotype == "cluster 15" ~ "subclonal",
                                TRUE ~ Background)) %>% 
  mutate(Haplotype = case_when(Haplotype == "cluster 15" ~ "subclonal",
                                TRUE ~ Haplotype))

```

### Cluster 10

Cluster 10 appears to be a set of SNPs on the background of genome-1 in the Frontal Cortex 2 specimen. However, since these SNPs are only present at appreciable frequency in a single tissue, we can't as clearly link them by their correlation in frequency alone.

```{r Cluster 10, fig.align='center', fig.width=8, fig.height=5, echo=T}

expanded.df %>% 
    filter(Haplotype %in% c("cluster 10")) %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue_order), y = AF,)) +
      geom_line(aes(group = SNP)) +
      facet_wrap(~factor(Haplotype, levels = haplotype.order)) + 
      scale_color_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))

```

Another issue with this cluster is that many of the mutations are in the Matrix gene. This gene has a stop codon fairly near the start of the protein and has been allowed to neutrally accumulate a large number of mutations over time. This is exacerbated by the mutagenic activity of ADAR. The issue with this is that mutations can arise multiple times on multiple haplotype backgrounds. Controlling for this is beyond the scope of our haplotyping approach. 

```{r Percentage of Matrix SNPs}

labeled.df %>% 
  filter (Haplotype == "cluster 10") %>% 
  select(SNP, Gene) %>% 
  distinct() %>% 
  group_by(Gene) %>% 
  count() %>% 
  arrange(-n)


```

We can see that if we drop only the Matrix mutations from this cluster, the peaks in frequency in tissues where cluster 10 should presumably be missing are mostly gone.

```{r Cluster 10 Non-matrix, fig.align='center', fig.width=8, fig.height=5, echo=T}

expanded.df %>% 
    filter(Haplotype %in% c("cluster 10")) %>%
    left_join(., distinct(select(labeled.df, SNP, Gene))) %>% 
    filter(Gene != "M") %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue_order), y = AF,)) +
      geom_line(aes(group = SNP)) +
      facet_wrap(~factor(Haplotype, levels = haplotype.order)) + 
      scale_color_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))


```

One solution to this issue, short of removing all Matrix mutations, is to remove the matrix mutations that are *also* linked to genome-1-1 or genome-2. This makes the correlation much better and is a more principled approach to refining this haplotype.

```{r Exclude Reccurent SNPs from Cluster 10, fig.align='center', fig.width=8, fig.height=5, echo=T}

exclude.from.cluster.10 = rank_bridging_reads("cluster 10", labeled.df, SNPassignmentsByTau, min.freq = 0.1) %>% 
  filter(linked >= 1) %>% 
  filter(Haplotype %in% c("genome-1-1", "genome-2")) %>% 
  select(cluster_snp) %>% 
  pull() %>% 
  unique()

expanded.df %>% 
    filter(Haplotype %in% c("cluster 10")) %>%
    left_join(., distinct(select(labeled.df, SNP, Gene, POS))) %>% 
    filter(!POS %in% exclude.from.cluster.10) %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue_order), y = AF,)) +
      geom_line(aes(group = SNP)) +
      facet_wrap(~factor(Haplotype, levels = haplotype.order)) + 
      scale_color_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))

print(paste("Excluding", length(exclude.from.cluster.10), "SNPs from cluster 10."))

labeled.df = labeled.df %>% 
  mutate(Background = case_when(POS %in% exclude.from.cluster.10 ~ "subclonal",
                                TRUE ~ Background)) %>% 
  mutate(Haplotype = case_when(POS %in% exclude.from.cluster.10 ~ "subclonal",
                                TRUE ~ Haplotype))
```

## Cluster 13

Cluster 13 is interesting, because from frequency alone, we know it's not possible that this exists as a single haplotype. This is because we believe that it's linked to genome-1 and therefore descended from cluster 10, yet, it's found at a higher frequency than cluster 10 in the Midbrain. This violates the 'pigeon-hole' principle. This cluster can't simultaneously be one both Cluster 10 and Genome 1-1 (Genome 01b). 

```{r Cluster 13, fig.align='center', fig.width=8, fig.height=5, echo=T}

expanded.df %>% 
    filter(Haplotype %in% c("cluster 13")) %>%
    left_join(., distinct(select(labeled.df, SNP, Gene, POS))) %>% 
    filter(!POS %in% exclude.from.cluster.10) %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue_order), y = AF,)) +
      geom_line(aes(group = SNP)) +
      facet_wrap(~factor(Haplotype, levels = haplotype.order)) + 
      scale_color_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))

```

This cluster suffers from several probable issues. For one, it's pretty low frequency so error is proportionately large. Second, it's only high frequency in two tissues - the frontal cortex and mid brain - so the correlation could be spurious Finally, it's only Matrix mutations, so the likelihood that they're recurrent is high. 

```{r Percentage of Matrix SNPs C13, echo=F}

labeled.df %>% 
  filter (Haplotype == "cluster 13") %>% 
  select(SNP, Gene) %>% 
  distinct() %>% 
  group_by(Gene) %>% 
  count() %>% 
  arrange(-n)

```

This is likely two or more haplotypes, and unsurprisingly, all found are linked to both genome-1-1 and cluster 10. This suggests that these are recurrent SNPs and shouldn't be considered a haplotype. 

```{r Remove Cluster 13, echo=T}

rank_bridging_reads("cluster 13", labeled.df, SNPassignmentsByTau, min.freq = 0.1) %>% 
  filter(linked >= 1) %>% 
  filter(Haplotype %in% c("genome-1-1")) %>% 
  select(cluster_snp) %>% 
  pull() %>% 
  unique()

labeled.df = labeled.df %>% 
  mutate(Background = case_when(Haplotype == "cluster 13" ~ "subclonal",
                                TRUE ~ Background)) %>% 
  mutate(Haplotype = case_when(Haplotype == "cluster 13" ~ "subclonal",
                                TRUE ~ Haplotype))

```

### Cluster 11

Now let's see about cluster 11. These are SNPs present in the lower brain where the read depth is pretty low, and the correlation is somewhat loose. 

```{r Cluster 11, fig.align='center', fig.width=8, fig.height=5, echo=T}

expanded.df %>% 
    filter(Haplotype %in% c("cluster 11")) %>%
    left_join(., distinct(select(labeled.df, SNP, Gene, POS))) %>% 
    filter(!POS %in% exclude.from.cluster.10) %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue_order), y = AF,)) +
      geom_line(aes(group = SNP)) +
      facet_wrap(~factor(Haplotype, levels = haplotype.order)) + 
      scale_color_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))

```

This cluster is half matrix mutations and half non-matrix mutations

```{r Percentage of Matrix SNPs C11, echo=F}

labeled.df %>% 
  filter (Haplotype == "cluster 11") %>% 
  select(SNP, Gene) %>%
  distinct() %>% 
  group_by(Gene) %>% 
  count() %>% 
  arrange(-n)

```

Below, I'll clean up cluster 11 by removing SNPs that are linked to genome-1 (since cluster 11 is a genome-2 cluster) and removing Matrix mutations unless they're linked to other cluster 11 mutations or genome-2. Finally, there is a very low depth F mutation that visibly looks like an outlier. I'll remove this as well.

```{r Clean Up Cluster 11, fig.align='center', fig.width=8, fig.height=5, echo=T}

linked.to.cluster11 = rank_bridging_reads("cluster 11", labeled.df, SNPassignmentsByTau, min.freq = 0.1) %>% 
  filter(linked >= 1) %>% 
  filter(Haplotype %in% c("cluster 11", "genome-2")) %>% 
  select(cluster_snp) %>% 
  pull() %>% 
  unique()

unlinked.from.cluster.11 = rank_bridging_reads("cluster 11", labeled.df, SNPassignmentsByTau, min.freq = 0.1) %>% 
  filter(linked >= 1) %>% 
  filter(Haplotype %in% c("genome-1-1", "genome-1")) %>% 
  select(cluster_snp) %>% 
  pull() %>% 
  unique()

keep.in.cluster.11 = linked.to.cluster11[which(!linked.to.cluster11 %in% unlinked.from.cluster.11)]

non.matrix.cluster.11 = labeled.df %>% 
  filter(Haplotype == "cluster 11") %>% 
  filter(Gene != "M") %>% 
  pull(POS) %>% 
  unique()

keep.in.cluster.11 = c(keep.in.cluster.11, non.matrix.cluster.11)

low.depth.outlier = 4923

keep.in.cluster.11 = keep.in.cluster.11[which(!keep.in.cluster.11 %in% low.depth.outlier)]

cluster.11.snps = labeled.df %>% 
  filter(Haplotype %in% c("cluster 11")) %>%
  pull(POS) %>% 
  unique()  

exclude.from.cluster.11 = cluster.11.snps[which(!cluster.11.snps %in% keep.in.cluster.11)]


expanded.df %>% 
    filter(Haplotype %in% c("cluster 11")) %>%
    left_join(., distinct(select(labeled.df, SNP, Gene, POS))) %>% 
    filter((!POS %in% exclude.from.cluster.11)) %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue_order), y = AF,)) +
      geom_line(aes(group = SNP)) +
      facet_wrap(~factor(Haplotype, levels = haplotype.order)) + 
      scale_color_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))


labeled.df = labeled.df %>% 
  mutate(Background = case_when((POS %in% exclude.from.cluster.11) ~ "subclonal",
                                TRUE ~ Background)) %>% 
  mutate(Haplotype = case_when((POS %in% exclude.from.cluster.11) ~ "subclonal",
                                TRUE ~ Haplotype))

```


### Cluster 2 and Cluster 9 

Finally, Cluster 2 and Cluster 9 are the same cluster but Cerebellum has very low depth. The SNPs in cluster 2 are present in Cerebellum at a similar frequency to cluster 9, but they fail the depth filters. Below, I'll amend that by setting the frequency to the mean frequency of cluster 9 and merging them.

```{r Merge Cluster 2 into 9, echo=T}

c9.mean.cerebellum.freq = labeled.df %>% 
  filter(Haplotype == "cluster 9" & Tissue == "Cerebellum") %>% 
  pull(AF) %>% 
  mean()

cluster.2.cerebellum.dummy.alleles = labeled.df %>% 
  filter(Haplotype == "cluster 2") %>% 
  filter(Tissue == "Cerebellum Nucleus") %>% 
  mutate(AF = c9.mean.cerebellum.freq, 
         Tissue = "Cerebellum", 
         Accession = "Cerebellum",
         DP = 100)


labeled.df = rbind(labeled.df, cluster.2.cerebellum.dummy.alleles) %>% 
  mutate(Haplotype = case_when(
    Haplotype == "cluster 2" ~ "cluster 9",
    TRUE ~ Haplotype)) 


```

Okay, so what do the clusters look like now after cleaning them up using the bridging reads? Before I proceed, I'll rename clusters sequentially according to whether they belong to the `Genome 1` or `Genome 2` lineage. 

```{r Cattaneo Nameing Scheme, echo=T}


labeled.df = labeled.df %>% 
  mutate(Haplotype_Name = case_when(
    Haplotype == "cluster 1" ~ "cluster 1",
    Haplotype == "cluster 3" ~ "cluster 2",
    Haplotype == "cluster 4" ~ "cluster 3",
    Haplotype == "cluster 8" ~ "cluster 4",
    Haplotype == "cluster 9" ~ "cluster 5",
    Haplotype == "cluster 10" ~ "cluster 6",
    Haplotype == "cluster 12" ~ "cluster 7",
    Haplotype == "cluster 14" ~ "cluster 8",
    Haplotype == "cluster 5" ~ "cluster 9",
    Haplotype == "cluster 6" ~ "cluster 10",
    Haplotype == "cluster 7" ~ "cluster 11",
    Haplotype == "cluster 11" ~ "cluster 12",
    Haplotype == "cluster 16" ~ "cluster 13",
    Haplotype == "cluster 14" ~ "cluster 14",
    Haplotype == "genome-1-1" ~ "genome 1",
    Haplotype == "genome-1" ~ "genome 01",
    Haplotype == "genome-2" ~ "genome 2",
    TRUE ~ Haplotype)) 


```

What do these validated clusters look like? 

```{r Ext Figure 4, message=F, warning=F, fig.align='center', fig.width=19, fig.height=10, echo=T}

# Get the haplotype names
haplotypes.label = labeled.df %>% 
  select(SNP, Haplotype, Haplotype_Name, Background) %>% 
  distinct()

# Expand the mutations to have a frequency for every tissue.
expanded.df = labeled.df %>% 
  select(SNP, Tissue, AF) %>% 
  pivot_wider(names_from = "Tissue", values_from = "AF", values_fill = 0) %>% 
  pivot_longer(cols = !SNP, names_to = "Tissue", values_to = "AF") %>% 
  left_join(., select(labeled.df, c("SNP", "Tissue", "DP")), by = c("SNP", "Tissue")) %>% 
  mutate(DP = if_else(is.na(DP), 0, DP)) %>% 
  left_join(., haplotypes.label, by = "SNP") 

# Tissue order - roughly by the location in the brain 
expanded_tissue_order = c(
  "SSPE 1", 
  "SSPE 2",
  "Frontal Cortex 2", 
  "Frontal Cortex 1", 
  "Frontal Cortex 3", 
  "Parietal Lobe",
  "Temporal Lobe", 
  "Occipital Lobe",
  "Hippocampus",
  "Internal Capsule", 
  "Midbrain", 
  "Upper Brain Stem",
  "Brain Stem",
  "Cerebellum",
  "Cerebellum Nucleus"
)

haplotype.order = c("cluster 1",
                    "cluster 2",
                    "cluster 3",
                    "cluster 4",
                    "cluster 5",
                    "cluster 6",
                    "cluster 7",
                    "cluster 8",
                    "cluster 9",
                    "cluster 10",
                    "cluster 11",
                    "cluster 12",
                    "cluster 13",
                    "cluster 14")

expanded.df %>% 
    filter(!Haplotype %in% c("fixed", "both", "subclonal", "genome-1", "genome-2", "genome-1-1")) %>% 
    mutate(Tissue = case_when(
      Tissue == "UBS" ~ "Upper Brain Stem",
      TRUE ~ Tissue
    )) %>% 
    mutate(Genotype = case_when(
      Background == "genome-1" ~ "G1",
      Background == "genome-2" ~ "G2"
    )) %>% 
    ggplot(aes(x = factor(Tissue, levels = expanded_tissue_order), y = AF, col = (Genotype))) +
      geom_line(aes(group = SNP)) +
      facet_wrap(~factor(Haplotype_Name, levels = haplotype.order)) + 
      scale_color_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Genotype") +
      theme_bw(20) +
      theme(strip.text.x = element_text(size = 12)) +
      theme(axis.title.x=element_blank()) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(legend.position="bottom")

ggsave("./ext-figure-4.png", width = 19, height = 10)

```

## Mutation Distribution 

Now that we have all of these haplotypes, we can see how these mutations are distributed throughout the MeV genome.

## Mutation distribution

What are the distribution of these haplotypes in the genome? 

```{r Annotations, message=FALSE, warning=FALSE, echo=F}

# Import genome annotations
annotations.df = read_csv(annotations.filepath, show_col_types = FALSE)

# Get the interval positions for these genes 
N = annotations.df %>% dplyr::filter(Locus == "N")
P.V.C = annotations.df %>% dplyr::filter(`Protein Name` == "phosphoprotein")
M = annotations.df %>% dplyr::filter(Locus == "M")
F. = annotations.df %>% dplyr::filter(Locus == "F")
H = annotations.df %>% dplyr::filter(Locus == "H")
L = annotations.df %>% dplyr::filter(Locus == "L")

gene_order = c("N", "P", "M", "F", "H", "L")

```

```{r Figure 5a, fig.align='center', fig.width=15, fig.height=10}

# Define the order of haplotypes on the y-axis
figure.haplotype.order = c("SSPE Ancestor",
                            "genome 01a",
                            "genome 01b",
                            "genome 1",
                            "cluster 1",
                            "cluster 2",
                            "cluster 3",
                            "cluster 4",
                            "cluster 5",
                            "cluster 6",
                            "cluster 7",
                            "cluster 8",
                            "genome 2",
                            "cluster 9",
                            "cluster 10",
                            "cluster 11",
                            "cluster 12",
                            "cluster 13",
                            "cluster 14", 
                            "cluster 15",
                            "cluster 16")


# Load in the 'ancestral' SNPs that are included in the consensus patient-specific reference
ancestral.snps = read_csv(ancestral.data) %>% 
  mutate(ADAR = case_when(
    REF == 'T' & ALT == 'C' ~ TRUE, 
    REF == 'A' & ALT == 'G' ~ TRUE,
    TRUE ~ FALSE
  )) %>% 
  mutate(Effect = case_when(
    is.na(WT_AA) ~ "Synonymous",
    WT_AA == MUT_AA ~ "Synonymous",
    WT_AA != MUT_AA ~ "Missense"
  )) %>% 
  select(POS, REF, ALT, Gene, ADAR, Effect) %>% 
  mutate(Haplotype = "SSPE Ancestor",
         Background = "SSPE Ancestor")

# Genome-1 SNPS
g1.snps = labeled.df %>% 
  filter(Haplotype %in% c("genome-1", "genome-1-1")) %>% 
  select(Background, Haplotype = Haplotype_Name, Gene, Effect, POS, REF, ALT) %>% 
  mutate(Gene = if_else(Gene == "P/V/C", "P", Gene)) %>% 
  mutate(Effect = if_else(is.na(Effect), "Synonymous", Effect)) %>% 
  mutate(ADAR = case_when(
    REF == 'T' & ALT == 'C' ~ TRUE, 
    REF == 'A' & ALT == 'G' ~ TRUE,
    TRUE ~ FALSE
  )) %>%  
  mutate(Haplotype = case_when(
    Haplotype == "genome 01" ~ "genome 1",
    Haplotype == "genome 1" ~ "genome 1",
    TRUE ~ Haplotype
  )) %>% 
  distinct() 

# Prepare the data for plotting
distribution.df = labeled.df %>% 
  select(Background, Haplotype = Haplotype_Name, Gene, Effect, POS, REF, ALT) %>% 
  mutate(Gene = if_else(Gene == "P/V/C", "P", Gene)) %>% 
  mutate(Effect = if_else(is.na(Effect), "Synonymous", Effect)) %>% 
  mutate(ADAR = case_when(
    REF == 'T' & ALT == 'C' ~ TRUE, 
    REF == 'A' & ALT == 'G' ~ TRUE,
    TRUE ~ FALSE
  )) %>%  
  mutate(Haplotype = case_when(
    Haplotype == "genome 01" ~ "genome 01a",
    Haplotype == "genome 1" ~ "genome 01b",
    TRUE ~ Haplotype
  )) %>% 
  distinct()  %>% 
  filter(!Haplotype %in% c('subclonal', 'both')) %>% 
  rbind(., ancestral.snps) %>%
  rbind(., g1.snps)


# Plot the distribution of mutations
distribution.df %>% 
  ggplot(aes(x = POS, y = factor(Haplotype, levels = rev(figure.haplotype.order)), col = Background)) +
    geom_point(aes(shape = ADAR), size = 8) +
    xlab("Position") +
  
  
    scale_x_continuous(limits=c(-180,16180), breaks=c(0, 5000, 10000, 15000)) +
    scale_color_manual(values=c("#858585", "#858585", "black"), guide = 'none') +
    scale_shape_manual(name="A to G / U to C", values = c(108, 4)) +
    
    # == Annotations of genes == #
    
  # 3' Start
    annotate(geom = "text", x = (N$Start - 180), y = .05, label = "3'", size = 8) +

    # Nucleoprotein
    annotate("rect", xmin = N$Start-1, xmax = N$Stop, ymin = -.5, ymax = .5 ,
               alpha = .5, fill = "#d97634", col = "black", size = .1) +
    annotate(geom = "text", x = (N$Stop + N$Start)/2, y = .05, label = "N", size = 8) +
    annotate("rect", xmin = N$Start-1, xmax = N$Stop, ymin = -.5, ymax = 18.5  ,
                   alpha = .05, fill = "#d97634", size = .1) +  
    # P/V/C
    annotate("rect", xmin = P.V.C$Start, xmax = P.V.C$Stop, ymin = -.5, ymax = .5 ,
               alpha = .5, fill = "#fcff3d", col = "black", size = .1) +
    annotate(geom = "text", x = (P.V.C$Stop + P.V.C$Start)/2, y = .05, label = "P/V/C", size = 8) +
    annotate("rect", xmin = P.V.C$Start, xmax = P.V.C$Stop, ymin = -.5, ymax = 18.5  ,
               alpha = .05, fill = "#fcff3d", size = .1) +
    # Matrix protein
    annotate("rect", xmin = M$Start, xmax = M$Stop, ymin = -.5, ymax = .5 ,
               alpha = .5, fill = "#0db51b", col = "black", size = .1) +
    annotate(geom = "text", x = (M$Stop + M$Start)/2, y = .05, label = "M", size = 8)  +
    annotate("rect", xmin = M$Start, xmax = M$Stop, ymin = -.5, ymax = 18.5  ,
               alpha = .05, fill = "#0db51b", size = .1) +
    # Fusion protein
    annotate("rect", xmin = F.$Start, xmax = F.$Stop, ymin = -.5, ymax = .5 ,
               alpha = .5, fill = "#6A3D9A", col = "black", size = .1) +
    annotate(geom = "text", x = (F.$Stop + F.$Start)/2, y = .05, label = "F", size = 8) +
    annotate("rect", xmin = F.$Start, xmax = F.$Stop, ymin = -.5, ymax = 18.5  ,
               alpha = .05, fill = "#6A3D9A", size = .1) +
    # Hemagluttinin
    annotate("rect", xmin = H$Start, xmax = H$Stop, ymin = -.5, ymax = .5 ,
               alpha = .5, fill = "#0251d9", col = "black", size = .1) +
    annotate(geom = "text", x = (H$Stop + H$Start)/2, y = .05, label = "H", size = 8) +
    annotate("rect", xmin = H$Start, xmax = H$Stop, ymin = -.5, ymax = 18.5 ,
               alpha = .05, fill = "#0251d9", size = .1) +
    # Large protein
    annotate("rect", xmin = L$Start, xmax = L$Stop, ymin = -.5, ymax = .5 ,
               alpha = .5, fill = "#bf1515", col = "black", size = .1) +
    annotate(geom = "text", x = (L$Stop + L$Start)/2, y = .05, label = "L", size = 8) +
    annotate("rect", xmin = L$Start, xmax = L$Stop, ymin = -.5, ymax = 18.5 ,
               alpha = .05, fill = "#bf1515", size = .1) +
  
    # 5' Start
    annotate(geom = "text", x = (L$Stop + 180), y =  .05, label = "5'", size = 8) +

  
    theme_classic() +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.background = element_rect(color = NA)) +
    theme(text=element_text(size=24)) + 
    theme(axis.text.x = element_text(size=28)) + 
    theme(axis.title.x = element_text(size=30)) + 
    theme(plot.title = element_text(hjust = 0.5))  +
    theme(axis.title.y = element_blank()) +
    theme(legend.position = "bottom")  

ggsave("./figure-5a.png", width = 15, height = 10, dpi = 300)

```

Finally, I'll write out these validated and renamed clusters to a new file. 

```{r echo=T}
# Write the validated labeled data
write_csv(labeled.df, output.path)
```


