{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1a759fd5",
   "metadata": {},
   "source": [
    "# Phylogenetic Analysis\n",
    "\n",
    "The goal of this notebook is to perform phylogenetic inference on the haplotypes identified in the brain. \n",
    "\n",
    "**Requirements:**\n",
    "Make sure you have `Biopython`, `MAFFT`, and `ETE3` installed in your environment. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b20b6636",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "from subprocess import call \n",
    "from Bio import SeqIO, AlignIO, SeqRecord, Seq\n",
    "from Bio.Align.Applications import MafftCommandline \n",
    "\n",
    "# from ete3 import Tree, TreeStyle, NodeStyle, TextFace, SequenceFace, CircleFace, faces, AttrFace "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1541c3cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path to output directory\n",
    "outpath = \"../../results/phylogeny\"\n",
    "\n",
    "if not os.path.exists(outpath):\n",
    "    os.mkdir(outpath)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d1f1ac8",
   "metadata": {},
   "source": [
    "## Make haplotypes from mutations\n",
    "\n",
    "I'll take the mutations from the variant calling analysis and make the haplotypes using the reference genome. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b8fa5014",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AF</th>\n",
       "      <th>DP</th>\n",
       "      <th>Effect</th>\n",
       "      <th>Gene_Name</th>\n",
       "      <th>AA_Change</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>SNP</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Cluster</th>\n",
       "      <th>Background</th>\n",
       "      <th>Haplotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>537</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.3265</td>\n",
       "      <td>132284</td>\n",
       "      <td>Missense</td>\n",
       "      <td>N</td>\n",
       "      <td>Ser144Pro</td>\n",
       "      <td>SSPE_1</td>\n",
       "      <td>SSPE 1</td>\n",
       "      <td>T537C</td>\n",
       "      <td>N</td>\n",
       "      <td>genome-2</td>\n",
       "      <td>genome-2</td>\n",
       "      <td>genome 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>684</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1140</td>\n",
       "      <td>127945</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>N</td>\n",
       "      <td>Leu193Leu</td>\n",
       "      <td>SSPE_1</td>\n",
       "      <td>SSPE 1</td>\n",
       "      <td>C684T</td>\n",
       "      <td>N</td>\n",
       "      <td>cluster 1</td>\n",
       "      <td>genome-1</td>\n",
       "      <td>cluster 4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1328</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.6809</td>\n",
       "      <td>131283</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>N</td>\n",
       "      <td>Ser407Ser</td>\n",
       "      <td>SSPE_1</td>\n",
       "      <td>SSPE 1</td>\n",
       "      <td>T1328C</td>\n",
       "      <td>N</td>\n",
       "      <td>genome-1-1</td>\n",
       "      <td>genome-1</td>\n",
       "      <td>genome 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1632</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.3126</td>\n",
       "      <td>121322</td>\n",
       "      <td>Missense</td>\n",
       "      <td>N</td>\n",
       "      <td>Gly509Ser</td>\n",
       "      <td>SSPE_1</td>\n",
       "      <td>SSPE 1</td>\n",
       "      <td>G1632A</td>\n",
       "      <td>N</td>\n",
       "      <td>cluster 7</td>\n",
       "      <td>genome-2</td>\n",
       "      <td>cluster 10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2139</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.6473</td>\n",
       "      <td>62808</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>P/V/C</td>\n",
       "      <td>His111His</td>\n",
       "      <td>SSPE_1</td>\n",
       "      <td>SSPE 1</td>\n",
       "      <td>T2139C</td>\n",
       "      <td>P/V/C</td>\n",
       "      <td>genome-1</td>\n",
       "      <td>genome-1</td>\n",
       "      <td>genome 01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    POS REF ALT      AF      DP      Effect Gene_Name  AA_Change Accession  \\\n",
       "0   537   T   C  0.3265  132284    Missense         N  Ser144Pro    SSPE_1   \n",
       "1   684   C   T  0.1140  127945  Synonymous         N  Leu193Leu    SSPE_1   \n",
       "5  1328   T   C  0.6809  131283  Synonymous         N  Ser407Ser    SSPE_1   \n",
       "7  1632   G   A  0.3126  121322    Missense         N  Gly509Ser    SSPE_1   \n",
       "9  2139   T   C  0.6473   62808  Synonymous     P/V/C  His111His    SSPE_1   \n",
       "\n",
       "   Tissue     SNP   Gene     Cluster Background   Haplotype  \n",
       "0  SSPE 1   T537C      N    genome-2   genome-2    genome 2  \n",
       "1  SSPE 1   C684T      N   cluster 1   genome-1   cluster 4  \n",
       "5  SSPE 1  T1328C      N  genome-1-1   genome-1    genome 1  \n",
       "7  SSPE 1  G1632A      N   cluster 7   genome-2  cluster 10  \n",
       "9  SSPE 1  T2139C  P/V/C    genome-1   genome-1   genome 01  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Final tree from SPRUCE/MACHINA\n",
    "cluster_relationships = pd.read_csv(\"../../results/phylogeny/filtered_spruce_tree.csv\")\n",
    "cluster_relationships['from'] = cluster_relationships['from'].replace('_', ' ', regex=True)\n",
    "cluster_relationships['to'] = cluster_relationships['to'].replace('_', ' ', regex=True)\n",
    "rename_clusters = {\"Anc\": \"SSPE ancestor\"}\n",
    "cluster_relationships['from'] = cluster_relationships['from'].replace(rename_clusters)\n",
    "\n",
    "# SSPE reference\n",
    "reference = [base for record in SeqIO.parse(\"../../config/ref/MeVChiTok-SSPE.fa\", \"fasta\") for base in record.seq]\n",
    "\n",
    "# Import the mutations\n",
    "haplotypes_df = pd.read_csv(\"../../results/variants/validated_variants.csv\")\n",
    "rename_haplotypes = {\"genome-1\": \"genome 01\", \n",
    "                     \"genome-1-1\": \"genome 1\",\n",
    "                     \"genome-2\": \"genome 2\"}\n",
    "haplotypes_df['Haplotype'] = haplotypes_df['Haplotype'].replace(rename_haplotypes)\n",
    "\n",
    "# We only need the mutations that have been assigned to a haplotype\n",
    "haplotypes_df = haplotypes_df[~haplotypes_df.Haplotype.isin(['subclonal', 'both'])]\n",
    "\n",
    "haplotypes_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "457d67a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_haplotype(to_node, from_seq):\n",
    "    \"\"\"\n",
    "    Add the SNPs the ancestor to build a haplotype.\n",
    "    \"\"\"\n",
    "    \n",
    "    # SNPs for a given haplotype\n",
    "    snps_df = haplotypes_df[haplotypes_df.Haplotype == to_node]\n",
    "    \n",
    "    # get a list of tuples for the SNPs\n",
    "    snps = {(pos, alt) for pos, alt in zip(snps_df.POS, snps_df.ALT)}\n",
    "    \n",
    "    # Copy the reference \n",
    "    haplotype = deepcopy(from_seq)\n",
    "    \n",
    "    # Change the mutation in the list.\n",
    "    for pos, alt in snps:\n",
    "        from_seq[pos - 1] = alt\n",
    "        \n",
    "    return haplotype\n",
    "    \n",
    "\n",
    "def traverse_tree(node, tree_df, haplotypes):\n",
    "    \"\"\"\n",
    "    Traverse the finalized SPRUCE tree and build the haplotypes into a \n",
    "    dictionary before alignment.\n",
    "    \"\"\"\n",
    "    children = tree_df[tree_df['from'] == node]\n",
    "\n",
    "    if children.empty:\n",
    "        return haplotypes\n",
    "\n",
    "    for _, row in children.iterrows():\n",
    "        from_node = row['from']\n",
    "        to_node = row['to']\n",
    "        from_seq = haplotypes[from_node]\n",
    "        print(f\"Makeing haplotype for {to_node} using background {from_node}\")\n",
    "        new_haplotype = make_haplotype(to_node, from_seq)\n",
    "        haplotypes[to_node] = new_haplotype\n",
    "        traverse_tree(to_node, tree_df, haplotypes)\n",
    "\n",
    "    return haplotypes\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8f95e06",
   "metadata": {},
   "source": [
    "### Haplotype Relationships"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b0286d5",
   "metadata": {},
   "source": [
    "I'll make the fasta files for each haplotype by cumulativley building them along the tree that we deteremine using `SPRUCE` filtered down using briding reads. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0b090fda",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>from</th>\n",
       "      <th>to</th>\n",
       "      <th>tree</th>\n",
       "      <th>treenum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>genome 2</td>\n",
       "      <td>cluster 10</td>\n",
       "      <td>tree 1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>genome 2</td>\n",
       "      <td>cluster 12</td>\n",
       "      <td>tree 1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>genome 2</td>\n",
       "      <td>cluster 11</td>\n",
       "      <td>tree 1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>genome 2</td>\n",
       "      <td>cluster 13</td>\n",
       "      <td>tree 1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>genome 2</td>\n",
       "      <td>cluster 9</td>\n",
       "      <td>tree 1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>genome 1</td>\n",
       "      <td>cluster 4</td>\n",
       "      <td>tree 1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>genome 1</td>\n",
       "      <td>cluster 2</td>\n",
       "      <td>tree 1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>genome 1</td>\n",
       "      <td>cluster 1</td>\n",
       "      <td>tree 1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>genome 1</td>\n",
       "      <td>cluster 8</td>\n",
       "      <td>tree 1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>genome 1</td>\n",
       "      <td>cluster 7</td>\n",
       "      <td>tree 1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>genome 1</td>\n",
       "      <td>cluster 3</td>\n",
       "      <td>tree 1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>genome 01</td>\n",
       "      <td>genome 1</td>\n",
       "      <td>tree 1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>genome 01</td>\n",
       "      <td>cluster 6</td>\n",
       "      <td>tree 1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>cluster 1</td>\n",
       "      <td>cluster 5</td>\n",
       "      <td>tree 1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SSPE ancestor</td>\n",
       "      <td>genome 2</td>\n",
       "      <td>tree 1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SSPE ancestor</td>\n",
       "      <td>genome 01</td>\n",
       "      <td>tree 1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             from          to    tree  treenum\n",
       "0        genome 2  cluster 10  tree 1        1\n",
       "1        genome 2  cluster 12  tree 1        1\n",
       "2        genome 2  cluster 11  tree 1        1\n",
       "3        genome 2  cluster 13  tree 1        1\n",
       "4        genome 2   cluster 9  tree 1        1\n",
       "5        genome 1   cluster 4  tree 1        1\n",
       "6        genome 1   cluster 2  tree 1        1\n",
       "7        genome 1   cluster 1  tree 1        1\n",
       "8        genome 1   cluster 8  tree 1        1\n",
       "9        genome 1   cluster 7  tree 1        1\n",
       "10       genome 1   cluster 3  tree 1        1\n",
       "11      genome 01    genome 1  tree 1        1\n",
       "12      genome 01   cluster 6  tree 1        1\n",
       "13      cluster 1   cluster 5  tree 1        1\n",
       "14  SSPE ancestor    genome 2  tree 1        1\n",
       "15  SSPE ancestor   genome 01  tree 1        1"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cluster_relationships"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "0525ac4d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Makeing haplotype for genome 2 using background SSPE ancestor\n",
      "Makeing haplotype for cluster 10 using background genome 2\n",
      "Makeing haplotype for cluster 12 using background genome 2\n",
      "Makeing haplotype for cluster 11 using background genome 2\n",
      "Makeing haplotype for cluster 13 using background genome 2\n",
      "Makeing haplotype for cluster 9 using background genome 2\n",
      "Makeing haplotype for genome 01 using background SSPE ancestor\n",
      "Makeing haplotype for genome 1 using background genome 01\n",
      "Makeing haplotype for cluster 4 using background genome 1\n",
      "Makeing haplotype for cluster 2 using background genome 1\n",
      "Makeing haplotype for cluster 1 using background genome 1\n",
      "Makeing haplotype for cluster 5 using background cluster 1\n",
      "Makeing haplotype for cluster 8 using background genome 1\n",
      "Makeing haplotype for cluster 7 using background genome 1\n",
      "Makeing haplotype for cluster 3 using background genome 1\n",
      "Makeing haplotype for cluster 6 using background genome 01\n"
     ]
    }
   ],
   "source": [
    "# Root is the SSPE Reference Sequence\n",
    "root = 'SSPE ancestor'\n",
    "haplotypes = {}\n",
    "haplotypes[root] = reference\n",
    "# Build the haplotypes \n",
    "resulting_haplotypes = traverse_tree(root, cluster_relationships, haplotypes)\n",
    "haplotype_seqs = {hap: \"\".join(nt for nt in seq) for hap, seq in resulting_haplotypes.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6cd331cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "haplotype_seqs['cluster 9'] == haplotype_seqs['genome 2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "546c4094",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Write out to a fasta file\n",
    "haplotype_records = [SeqRecord.SeqRecord(Seq.Seq(seq), id = hap) for hap, seq in haplotype_seqs.items()]\n",
    "SeqIO.write(haplotype_records, os.path.join(outpath, \"haplotype-sequences.fa\"), \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5767103c",
   "metadata": {},
   "source": [
    "## Visualize the Tree"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50ad79b0",
   "metadata": {},
   "source": [
    "Make a tree with the Fasta file of each terminal haplotype. These are alredady aligned by the nature of how I made them. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3a8cb936",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IQ-TREE multicore version 2.1.4-beta COVID-edition for Linux 64-bit built Jun 24 2021\n",
      "Developed by Bui Quang Minh, James Barbetti, Nguyen Lam Tung,\n",
      "Olga Chernomor, Heiko Schmidt, Dominik Schrempf, Michael Woodhams.\n",
      "\n",
      "Host:    rhino02 (AVX512, FMA3, 754 GB RAM)\n",
      "Command: iqtree -s ../../results/phylogeny/haplotype-sequences.fa -m GTR+I+G -bb 1000 -redo\n",
      "Seed:    55499 (Using SPRNG - Scalable Parallel Random Number Generator)\n",
      "Time:    Fri Mar 31 15:04:29 2023\n",
      "Kernel:  AVX+FMA - 1 threads (72 CPU cores detected)\n",
      "\n",
      "HINT: Use -nt option to specify number of threads because your CPU has 72 cores!\n",
      "HINT: -nt AUTO will automatically determine the best number of threads to use.\n",
      "\n",
      "Reading alignment file ../../results/phylogeny/haplotype-sequences.fa ... Fasta format detected\n",
      "NOTE: Change sequence name 'SSPE ancestor <unknown description>' -> SSPE_ancestor\n",
      "NOTE: Change sequence name 'genome 2 <unknown description>' -> genome_2\n",
      "NOTE: Change sequence name 'cluster 10 <unknown description>' -> cluster_10\n",
      "NOTE: Change sequence name 'cluster 12 <unknown description>' -> cluster_12\n",
      "NOTE: Change sequence name 'cluster 11 <unknown description>' -> cluster_11\n",
      "NOTE: Change sequence name 'cluster 13 <unknown description>' -> cluster_13\n",
      "NOTE: Change sequence name 'cluster 9 <unknown description>' -> cluster_9\n",
      "NOTE: Change sequence name 'genome 01 <unknown description>' -> genome_01\n",
      "NOTE: Change sequence name 'genome 1 <unknown description>' -> genome_1\n",
      "NOTE: Change sequence name 'cluster 4 <unknown description>' -> cluster_4\n",
      "NOTE: Change sequence name 'cluster 2 <unknown description>' -> cluster_2\n",
      "NOTE: Change sequence name 'cluster 1 <unknown description>' -> cluster_1\n",
      "NOTE: Change sequence name 'cluster 5 <unknown description>' -> cluster_5\n",
      "NOTE: Change sequence name 'cluster 8 <unknown description>' -> cluster_8\n",
      "NOTE: Change sequence name 'cluster 7 <unknown description>' -> cluster_7\n",
      "NOTE: Change sequence name 'cluster 3 <unknown description>' -> cluster_3\n",
      "NOTE: Change sequence name 'cluster 6 <unknown description>' -> cluster_6\n",
      "Alignment most likely contains DNA/RNA sequences\n",
      "Alignment has 17 sequences with 15894 columns, 46 distinct patterns\n",
      "55 parsimony-informative, 23 singleton sites, 15816 constant sites\n",
      "               Gap/Ambiguity  Composition  p-value\n",
      "   1  SSPE_ancestor    0.00%    passed    100.00%\n",
      "   2  genome_2         0.00%    passed     99.92%\n",
      "   3  cluster_10       0.00%    passed    100.00%\n",
      "   4  cluster_12       0.00%    passed    100.00%\n",
      "   5  cluster_11       0.00%    passed     99.99%\n",
      "   6  cluster_13       0.00%    passed     99.97%\n",
      "   7  cluster_9        0.00%    passed     99.92%\n",
      "   8  genome_01        0.00%    passed     99.93%\n",
      "   9  genome_1         0.00%    passed    100.00%\n",
      "  10  cluster_4        0.00%    passed    100.00%\n",
      "  11  cluster_2        0.00%    passed    100.00%\n",
      "  12  cluster_1        0.00%    passed     99.97%\n",
      "  13  cluster_5        0.00%    passed     99.99%\n",
      "  14  cluster_8        0.00%    passed    100.00%\n",
      "  15  cluster_7        0.00%    passed    100.00%\n",
      "  16  cluster_3        0.00%    passed    100.00%\n",
      "  17  cluster_6        0.00%    passed     99.86%\n",
      "****  TOTAL            0.00%  0 sequences failed composition chi2 test (p-value<5%; df=3)\n",
      "NOTE: cluster_10 is identical to SSPE_ancestor but kept for subsequent analysis\n",
      "NOTE: 1 identical sequences (see below) will be ignored for subsequent analysis\n",
      "NOTE: cluster_4 (identical to SSPE_ancestor) is ignored but added at the end\n",
      "Alignment was printed to ../../results/phylogeny/haplotype-sequences.fa.uniqueseq.phy\n",
      "\n",
      "For your convenience alignment with unique sequences printed to ../../results/phylogeny/haplotype-sequences.fa.uniqueseq.phy\n",
      "\n",
      "Create initial parsimony tree by phylogenetic likelihood library (PLL)... 0.000 seconds\n",
      "Generating 1000 samples for ultrafast bootstrap (seed: 55499)...\n",
      "\n",
      "NOTE: 0 MB RAM (0 GB) is required!\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "Thoroughly optimizing +I+G parameters from 10 start values...\n",
      "Init pinv, alpha: 0.000, 1.000 / Estimate: 0.000, 0.020 / LogL: -22690.462\n",
      "Init pinv, alpha: 0.111, 1.000 / Estimate: 0.111, 0.020 / LogL: -22690.390\n",
      "Init pinv, alpha: 0.221, 1.000 / Estimate: 0.221, 0.054 / LogL: -22690.369\n",
      "Init pinv, alpha: 0.332, 1.000 / Estimate: 0.332, 0.167 / LogL: -22690.338\n",
      "Init pinv, alpha: 0.442, 1.000 / Estimate: 0.442, 0.284 / LogL: -22690.309\n",
      "Init pinv, alpha: 0.553, 1.000 / Estimate: 0.553, 0.490 / LogL: -22690.290\n",
      "Init pinv, alpha: 0.663, 1.000 / Estimate: 0.663, 1.004 / LogL: -22690.281\n",
      "Init pinv, alpha: 0.774, 1.000 / Estimate: 0.774, 5.143 / LogL: -22690.255\n",
      "Init pinv, alpha: 0.885, 1.000 / Estimate: 0.884, 998.441 / LogL: -22690.400\n",
      "Init pinv, alpha: 0.995, 1.000 / Estimate: 0.975, 998.447 / LogL: -22695.945\n",
      "Optimal pinv,alpha: 0.774, 5.143 / LogL: -22690.255\n",
      "\n",
      "Parameters optimization took 0.067 sec\n",
      "Computing ML distances based on estimated model parameters...\n",
      "Computing ML distances took 0.002608 sec (of wall-clock time) 0.002614 sec(of CPU time)\n",
      "Computing RapidNJ tree took 0.000210 sec (of wall-clock time) 0.000174 sec (of CPU time)\n",
      "Log-likelihood of RapidNJ tree: -22757.247\n",
      "--------------------------------------------------------------------\n",
      "|             INITIALIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Generating 98 parsimony trees... 0.045 second\n",
      "Computing log-likelihood of 98 initial trees ... 0.031 seconds\n",
      "Current best score: -22690.255\n",
      "\n",
      "Do NNI search on 20 best initial trees\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 1: -22690.250\n",
      "Iteration 10 / LogL: -22690.251 / Time: 0h:0m:0s\n",
      "Iteration 20 / LogL: -22690.251 / Time: 0h:0m:0s\n",
      "Finish initializing candidate tree set (6)\n",
      "Current best tree score: -22690.250 / CPU time: 0.130\n",
      "Number of iterations: 20\n",
      "--------------------------------------------------------------------\n",
      "|               OPTIMIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Iteration 30 / LogL: -22690.273 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 40 / LogL: -22697.986 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 50 / LogL: -22690.331 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -22717.638\n",
      "Iteration 60 / LogL: -22690.273 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 70 / LogL: -22690.418 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 80 / LogL: -22690.443 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 90 / LogL: -22690.374 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 100 / LogL: -22690.295 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -22717.832\n",
      "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 0.997\n",
      "TREE SEARCH COMPLETED AFTER 102 ITERATIONS / Time: 0h:0m:0s\n",
      "\n",
      "--------------------------------------------------------------------\n",
      "|                    FINALIZING TREE SEARCH                        |\n",
      "--------------------------------------------------------------------\n",
      "Performs final model parameters optimization\n",
      "Estimate model parameters (epsilon = 0.010)\n",
      "1. Initial log-likelihood: -22690.250\n",
      "2. Current log-likelihood: -22690.228\n",
      "3. Current log-likelihood: -22690.211\n",
      "Optimal log-likelihood: -22690.205\n",
      "Rate parameters:  A-C: 1.49906  A-G: 6.36554  A-T: 0.78613  C-G: 0.31203  C-T: 13.11669  G-T: 1.00000\n",
      "Base frequencies:  A: 0.291  C: 0.241  G: 0.236  T: 0.232\n",
      "Proportion of invariable sites: 0.774\n",
      "Gamma shape alpha: 5.138\n",
      "Parameters optimization took 3 rounds (0.003 sec)\n",
      "BEST SCORE FOUND : -22690.205\n",
      "Creating bootstrap support values...\n",
      "Split supports printed to NEXUS file ../../results/phylogeny/haplotype-sequences.fa.splits.nex\n",
      "Total tree length: 0.005\n",
      "\n",
      "Total number of iterations: 102\n",
      "CPU time used for tree search: 0.812 sec (0h:0m:0s)\n",
      "Wall-clock time used for tree search: 0.837 sec (0h:0m:0s)\n",
      "Total CPU time used: 0.922 sec (0h:0m:0s)\n",
      "Total wall-clock time used: 0.979 sec (0h:0m:0s)\n",
      "\n",
      "Computing bootstrap consensus tree...\n",
      "Reading input file ../../results/phylogeny/haplotype-sequences.fa.splits.nex...\n",
      "16 taxa and 58 splits.\n",
      "Consensus tree written to ../../results/phylogeny/haplotype-sequences.fa.contree\n",
      "Reading input trees file ../../results/phylogeny/haplotype-sequences.fa.contree\n",
      "Log-likelihood of consensus tree: -22690.666\n",
      "\n",
      "Analysis results written to: \n",
      "  IQ-TREE report:                ../../results/phylogeny/haplotype-sequences.fa.iqtree\n",
      "  Maximum-likelihood tree:       ../../results/phylogeny/haplotype-sequences.fa.treefile\n",
      "  Likelihood distances:          ../../results/phylogeny/haplotype-sequences.fa.mldist\n",
      "\n",
      "Ultrafast bootstrap approximation results written to:\n",
      "  Split support values:          ../../results/phylogeny/haplotype-sequences.fa.splits.nex\n",
      "  Consensus tree:                ../../results/phylogeny/haplotype-sequences.fa.contree\n",
      "  Screen log file:               ../../results/phylogeny/haplotype-sequences.fa.log\n",
      "\n",
      "Date and Time: Fri Mar 31 15:04:30 2023\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Building the phylogeny with IQtree with 1000 bootstrap iterations with GTR+I+G (Invariable site plus discrete Gamma model) with asr\n",
    "alignfasta = os.path.join(outpath, \"haplotype-sequences.fa\")\n",
    "call(f\"iqtree -s {alignfasta} -m GTR+I+G -bb 1000 -redo\", shell=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd2374d4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
