{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "5dd6c7e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import pandas as pd\n",
    "from subprocess import call \n",
    "from collections import defaultdict\n",
    "from Bio import SeqIO, AlignIO, SeqRecord, Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "2cbe93ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path to output directory\n",
    "outpath = \"../../results/phylogeny\"\n",
    "\n",
    "if not os.path.exists(outpath):\n",
    "    os.mkdir(outpath)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ce4b6a0",
   "metadata": {},
   "source": [
    "## Spruce Tree\n",
    "\n",
    "This data below represent the phylogenetic relationship of the haplotypes to one another. This is the final tree validated using bridging reads. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "38f7eace",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>from</th>\n",
       "      <th>to</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>genome 2</td>\n",
       "      <td>cluster 10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>genome 2</td>\n",
       "      <td>cluster 12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>genome 2</td>\n",
       "      <td>cluster 11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>genome 2</td>\n",
       "      <td>cluster 13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>genome 2</td>\n",
       "      <td>cluster 9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>genome 1</td>\n",
       "      <td>cluster 4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>genome 1</td>\n",
       "      <td>cluster 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>genome 1</td>\n",
       "      <td>cluster 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>genome 1</td>\n",
       "      <td>cluster 8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>genome 1</td>\n",
       "      <td>cluster 7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>genome 1</td>\n",
       "      <td>cluster 3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>genome 01</td>\n",
       "      <td>genome 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>genome 01</td>\n",
       "      <td>cluster 6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>cluster 1</td>\n",
       "      <td>cluster 5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SSPE ancestor</td>\n",
       "      <td>genome 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SSPE ancestor</td>\n",
       "      <td>genome 01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             from          to\n",
       "0        genome 2  cluster 10\n",
       "1        genome 2  cluster 12\n",
       "2        genome 2  cluster 11\n",
       "3        genome 2  cluster 13\n",
       "4        genome 2   cluster 9\n",
       "5        genome 1   cluster 4\n",
       "6        genome 1   cluster 2\n",
       "7        genome 1   cluster 1\n",
       "8        genome 1   cluster 8\n",
       "9        genome 1   cluster 7\n",
       "10       genome 1   cluster 3\n",
       "11      genome 01    genome 1\n",
       "12      genome 01   cluster 6\n",
       "13      cluster 1   cluster 5\n",
       "14  SSPE ancestor    genome 2\n",
       "15  SSPE ancestor   genome 01"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Final tree from SPRUCE/MACHINA\n",
    "cluster_relationships = pd.read_csv(\"../../results/phylogeny/filtered_spruce_tree.csv\")\n",
    "cluster_relationships['from'] = cluster_relationships['from'].replace('_', ' ', regex=True)\n",
    "cluster_relationships['to'] = cluster_relationships['to'].replace('_', ' ', regex=True)\n",
    "rename_clusters = {\"Anc\": \"SSPE ancestor\"}\n",
    "cluster_relationships['from'] = cluster_relationships['from'].replace(rename_clusters)\n",
    "cluster_relationships.drop(['tree', 'treenum'], axis=1, inplace=True)\n",
    "cluster_relationships"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "102e50f9",
   "metadata": {},
   "source": [
    "## Haplotype Mutations\n",
    "\n",
    "The following dataframe contains all of the variants in the 15 SSPE samples annotated by their haplotype."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "6188a0e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>Haplotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>537</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>genome 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>684</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>cluster 4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1328</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>genome 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1632</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>cluster 10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2139</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>genome 01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    POS REF ALT   Haplotype\n",
       "0   537   T   C    genome 2\n",
       "1   684   C   T   cluster 4\n",
       "5  1328   T   C    genome 1\n",
       "7  1632   G   A  cluster 10\n",
       "9  2139   T   C   genome 01"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import the mutations\n",
    "haplotypes_df = pd.read_csv(\"../../results/variants/validated_variants.csv\")\n",
    "\n",
    "# Rename to standardize the naming scheme\n",
    "rename_haplotypes = {\"genome-1\": \"genome 01\", \n",
    "                     \"genome-1-1\": \"genome 1\",\n",
    "                     \"genome-2\": \"genome 2\"}\n",
    "haplotypes_df['Haplotype'] = haplotypes_df['Haplotype'].replace(rename_haplotypes)\n",
    "\n",
    "# We only need the mutations that have been assigned to a haplotype\n",
    "haplotypes_df = haplotypes_df[~haplotypes_df.Haplotype.isin(['subclonal', 'both', 'fixed'])]\n",
    "haplotypes_df = haplotypes_df.loc[:, ['POS', 'REF', 'ALT', 'Haplotype']].drop_duplicates()\n",
    "\n",
    "haplotypes_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f33634fb",
   "metadata": {},
   "source": [
    "## Make Haplotypes \n",
    "\n",
    "Iterate through the tree structure and add in the mutations to generate a dictionary of haplotype sequences. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "5ec0de45",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to recursively apply mutations\n",
    "def apply_mutations(haplotype, tree, haplotype_df, haplotype_dict, reference):\n",
    "    \n",
    "    if haplotype in haplotype_dict:\n",
    "        return haplotype_dict[haplotype]\n",
    "    \n",
    "    ancestor = tree[haplotype]\n",
    "    ancestor_sequence = apply_mutations(ancestor, tree, haplotype_df, haplotype_dict, reference)\n",
    "    \n",
    "    haplotype_sequence = list(ancestor_sequence)  # Create a copy of the ancestor's sequence\n",
    "    \n",
    "    haplotype_mutations = haplotype_df[haplotype_df['Haplotype'] == haplotype]\n",
    "    \n",
    "    for _, mutation in haplotype_mutations.iterrows():\n",
    "        pos, ref, alt = mutation['POS'], mutation['REF'], mutation['ALT']\n",
    "        if haplotype_sequence[pos-1] == ref:\n",
    "            haplotype_sequence[pos-1] = alt\n",
    "        else:\n",
    "            raise ValueError(f\"Reference allele mismatch at position {pos} for {haplotype}\")\n",
    "    \n",
    "    haplotype_sequence = ''.join(haplotype_sequence)\n",
    "    haplotype_dict[haplotype] = haplotype_sequence\n",
    "    \n",
    "    return haplotype_sequence\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "df0dbadc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import the reference sequence \n",
    "reference = [base for record in SeqIO.parse(\"../../config/ref/MeVChiTok-SSPE.fa\", \"fasta\") for base in record.seq]\n",
    "\n",
    "# Build the tree-like structure\n",
    "tree = {}\n",
    "for _, row in cluster_relationships.iterrows():\n",
    "    tree[row['to']] = row['from']\n",
    "\n",
    "# Initialize the haplotype dictionary with the known ancestor sequence\n",
    "haplotype_dict = {'SSPE ancestor': ''.join(reference)}\n",
    "\n",
    "# Create the DNA haplotypes_df for each haplotype based on the tree structure\n",
    "for haplotype in haplotypes_df['Haplotype'].unique():\n",
    "    if haplotype not in haplotype_dict:\n",
    "        apply_mutations(haplotype, tree, haplotypes_df, haplotype_dict, reference)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e87309f",
   "metadata": {},
   "source": [
    "## Build a Tree\n",
    "\n",
    "Now, I'm going to take the haplotype sequences and build a tree using `iqtree`. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "7deed2d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Write out to a fasta file\n",
    "haplotype_records = [SeqRecord.SeqRecord(Seq.Seq(seq), id = hap) for hap, seq in haplotype_dict.items()]\n",
    "SeqIO.write(haplotype_records, os.path.join(outpath, \"haplotype-sequences.fa\"), \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "fb42bcf8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IQ-TREE multicore version 2.1.4-beta COVID-edition for Linux 64-bit built Jun 24 2021\n",
      "Developed by Bui Quang Minh, James Barbetti, Nguyen Lam Tung,\n",
      "Olga Chernomor, Heiko Schmidt, Dominik Schrempf, Michael Woodhams.\n",
      "\n",
      "Host:    rhino02 (AVX512, FMA3, 754 GB RAM)\n",
      "Command: iqtree -s ../../results/phylogeny/haplotype-sequences.fa -m GTR+I+G -bb 1000 -redo\n",
      "Seed:    941483 (Using SPRNG - Scalable Parallel Random Number Generator)\n",
      "Time:    Mon Apr  3 11:22:01 2023\n",
      "Kernel:  AVX+FMA - 1 threads (72 CPU cores detected)\n",
      "\n",
      "HINT: Use -nt option to specify number of threads because your CPU has 72 cores!\n",
      "HINT: -nt AUTO will automatically determine the best number of threads to use.\n",
      "\n",
      "Reading alignment file ../../results/phylogeny/haplotype-sequences.fa ... Fasta format detected\n",
      "NOTE: Change sequence name 'SSPE ancestor <unknown description>' -> SSPE_ancestor\n",
      "NOTE: Change sequence name 'genome 2 <unknown description>' -> genome_2\n",
      "NOTE: Change sequence name 'genome 01 <unknown description>' -> genome_01\n",
      "NOTE: Change sequence name 'genome 1 <unknown description>' -> genome_1\n",
      "NOTE: Change sequence name 'cluster 4 <unknown description>' -> cluster_4\n",
      "NOTE: Change sequence name 'cluster 10 <unknown description>' -> cluster_10\n",
      "NOTE: Change sequence name 'cluster 2 <unknown description>' -> cluster_2\n",
      "NOTE: Change sequence name 'cluster 6 <unknown description>' -> cluster_6\n",
      "NOTE: Change sequence name 'cluster 1 <unknown description>' -> cluster_1\n",
      "NOTE: Change sequence name 'cluster 5 <unknown description>' -> cluster_5\n",
      "NOTE: Change sequence name 'cluster 8 <unknown description>' -> cluster_8\n",
      "NOTE: Change sequence name 'cluster 7 <unknown description>' -> cluster_7\n",
      "NOTE: Change sequence name 'cluster 3 <unknown description>' -> cluster_3\n",
      "NOTE: Change sequence name 'cluster 12 <unknown description>' -> cluster_12\n",
      "NOTE: Change sequence name 'cluster 11 <unknown description>' -> cluster_11\n",
      "NOTE: Change sequence name 'cluster 13 <unknown description>' -> cluster_13\n",
      "NOTE: Change sequence name 'cluster 9 <unknown description>' -> cluster_9\n",
      "Alignment most likely contains DNA/RNA sequences\n",
      "Alignment has 17 sequences with 15894 columns, 56 distinct patterns\n",
      "58 parsimony-informative, 66 singleton sites, 15770 constant sites\n",
      "               Gap/Ambiguity  Composition  p-value\n",
      "   1  SSPE_ancestor    0.00%    passed     99.00%\n",
      "   2  genome_2         0.00%    passed     99.81%\n",
      "   3  genome_01        0.00%    passed    100.00%\n",
      "   4  genome_1         0.00%    passed     99.82%\n",
      "   5  cluster_4        0.00%    passed     99.71%\n",
      "   6  cluster_10       0.00%    passed     99.77%\n",
      "   7  cluster_2        0.00%    passed     99.90%\n",
      "   8  cluster_6        0.00%    passed    100.00%\n",
      "   9  cluster_1        0.00%    passed     99.82%\n",
      "  10  cluster_5        0.00%    passed     99.55%\n",
      "  11  cluster_8        0.00%    passed     99.77%\n",
      "  12  cluster_7        0.00%    passed     99.78%\n",
      "  13  cluster_3        0.00%    passed     99.80%\n",
      "  14  cluster_12       0.00%    passed     99.81%\n",
      "  15  cluster_11       0.00%    passed     99.63%\n",
      "  16  cluster_13       0.00%    passed     99.67%\n",
      "  17  cluster_9        0.00%    passed     99.81%\n",
      "****  TOTAL            0.00%  0 sequences failed composition chi2 test (p-value<5%; df=3)\n",
      "\n",
      "Create initial parsimony tree by phylogenetic likelihood library (PLL)... 0.001 seconds\n",
      "Generating 1000 samples for ultrafast bootstrap (seed: 941483)...\n",
      "\n",
      "NOTE: 0 MB RAM (0 GB) is required!\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "Thoroughly optimizing +I+G parameters from 10 start values...\n",
      "Init pinv, alpha: 0.000, 1.000 / Estimate: 0.000, 0.735 / LogL: -23042.247\n",
      "Init pinv, alpha: 0.110, 1.000 / Estimate: 0.110, 1.000 / LogL: -23042.191\n",
      "Init pinv, alpha: 0.220, 1.000 / Estimate: 0.220, 1.481 / LogL: -23042.163\n",
      "Init pinv, alpha: 0.331, 1.000 / Estimate: 0.331, 2.639 / LogL: -23042.151\n",
      "Init pinv, alpha: 0.441, 1.000 / Estimate: 0.441, 9.278 / LogL: -23042.143\n",
      "Init pinv, alpha: 0.551, 1.000 / Estimate: 0.551, 998.422 / LogL: -23042.150\n",
      "Init pinv, alpha: 0.661, 1.000 / Estimate: 0.661, 998.439 / LogL: -23042.241\n",
      "Init pinv, alpha: 0.772, 1.000 / Estimate: 0.772, 998.444 / LogL: -23042.588\n",
      "Init pinv, alpha: 0.882, 1.000 / Estimate: 0.882, 998.446 / LogL: -23044.090\n",
      "Init pinv, alpha: 0.992, 1.000 / Estimate: 0.950, 998.447 / LogL: -23051.101\n",
      "Optimal pinv,alpha: 0.441, 9.278 / LogL: -23042.143\n",
      "\n",
      "Parameters optimization took 0.088 sec\n",
      "\n",
      "WARNING: Estimated model parameters are at boundary that can cause numerical instability!\n",
      "\n",
      "Computing ML distances based on estimated model parameters...\n",
      "Computing ML distances took 0.001761 sec (of wall-clock time) 0.001652 sec(of CPU time)\n",
      "Computing RapidNJ tree took 0.000250 sec (of wall-clock time) 0.000000 sec (of CPU time)\n",
      "Log-likelihood of RapidNJ tree: -23042.531\n",
      "--------------------------------------------------------------------\n",
      "|             INITIALIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Generating 98 parsimony trees... 0.028 second\n",
      "Computing log-likelihood of 98 initial trees ... 0.030 seconds\n",
      "Current best score: -23042.143\n",
      "\n",
      "Do NNI search on 20 best initial trees\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "\n",
      "WARNING: Estimated model parameters are at boundary that can cause numerical instability!\n",
      "\n",
      "BETTER TREE FOUND at iteration 1: -23042.142\n",
      "BETTER TREE FOUND at iteration 6: -23042.142\n",
      "BETTER TREE FOUND at iteration 10: -23042.142\n",
      "Iteration 10 / LogL: -23042.142 / Time: 0h:0m:0s\n",
      "Iteration 20 / LogL: -23042.142 / Time: 0h:0m:0s\n",
      "Finish initializing candidate tree set (20)\n",
      "Current best tree score: -23042.142 / CPU time: 0.112\n",
      "Number of iterations: 20\n",
      "--------------------------------------------------------------------\n",
      "|               OPTIMIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Iteration 30 / LogL: -23042.142 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "BETTER TREE FOUND at iteration 35: -23042.142\n",
      "Iteration 40 / LogL: -23042.142 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "BETTER TREE FOUND at iteration 50: -23042.142\n",
      "Iteration 50 / LogL: -23042.142 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -23066.071\n",
      "BETTER TREE FOUND at iteration 56: -23042.142\n",
      "Iteration 60 / LogL: -23042.204 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 70 / LogL: -23042.142 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 80 / LogL: -23042.142 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 90 / LogL: -23042.160 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "BETTER TREE FOUND at iteration 93: -23042.142\n",
      "Iteration 100 / LogL: -23042.180 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -23066.071\n",
      "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 0.916\n",
      "NOTE: UFBoot does not converge, continue at least 100 more iterations\n",
      "BETTER TREE FOUND at iteration 105: -23042.142\n",
      "BETTER TREE FOUND at iteration 108: -23042.142\n",
      "Iteration 110 / LogL: -23042.158 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 120 / LogL: -23042.230 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "Iteration 130 / LogL: -23206.293 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "BETTER TREE FOUND at iteration 139: -23042.142\n",
      "Iteration 140 / LogL: -23042.191 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "Iteration 150 / LogL: -23042.216 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -23066.071\n",
      "Iteration 160 / LogL: -23042.172 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "Iteration 170 / LogL: -23049.894 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "Iteration 180 / LogL: -23042.172 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "Iteration 190 / LogL: -23042.202 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "BETTER TREE FOUND at iteration 192: -23042.142\n",
      "Iteration 200 / LogL: -23042.185 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -23066.226\n",
      "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 0.830\n",
      "NOTE: UFBoot does not converge, continue at least 100 more iterations\n",
      "BETTER TREE FOUND at iteration 201: -23042.142\n",
      "Iteration 210 / LogL: -23042.173 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "Iteration 220 / LogL: -23042.190 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "Iteration 230 / LogL: -23240.834 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "Iteration 240 / LogL: -23042.168 / Time: 0h:0m:2s (0h:0m:0s left)\n",
      "Iteration 250 / LogL: -23042.142 / Time: 0h:0m:2s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -23066.308\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Iteration 260 / LogL: -23042.212 / Time: 0h:0m:2s (0h:0m:0s left)\n",
      "Iteration 270 / LogL: -23057.952 / Time: 0h:0m:2s (0h:0m:0s left)\n",
      "Iteration 280 / LogL: -23042.207 / Time: 0h:0m:2s (0h:0m:0s left)\n",
      "Iteration 290 / LogL: -23049.855 / Time: 0h:0m:2s (0h:0m:0s left)\n",
      "Iteration 300 / LogL: -23042.142 / Time: 0h:0m:2s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -23066.308\n",
      "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 0.943\n",
      "NOTE: UFBoot does not converge, continue at least 100 more iterations\n",
      "Iteration 310 / LogL: -23042.158 / Time: 0h:0m:2s (0h:0m:0s left)\n",
      "Iteration 320 / LogL: -23042.142 / Time: 0h:0m:2s (0h:0m:0s left)\n",
      "Iteration 330 / LogL: -23042.192 / Time: 0h:0m:2s (0h:0m:0s left)\n",
      "UPDATE BEST LOG-LIKELIHOOD: -23042.142\n",
      "Iteration 340 / LogL: -23042.182 / Time: 0h:0m:2s (0h:0m:0s left)\n",
      "Iteration 350 / LogL: -23042.185 / Time: 0h:0m:2s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -23066.308\n",
      "Iteration 360 / LogL: -23156.159 / Time: 0h:0m:3s (0h:0m:0s left)\n",
      "Iteration 370 / LogL: -23042.233 / Time: 0h:0m:3s (0h:0m:0s left)\n",
      "Iteration 380 / LogL: -23156.226 / Time: 0h:0m:3s (0h:0m:0s left)\n",
      "Iteration 390 / LogL: -23042.142 / Time: 0h:0m:3s (0h:0m:0s left)\n",
      "Iteration 400 / LogL: -23042.142 / Time: 0h:0m:3s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -23066.308\n",
      "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 0.930\n",
      "NOTE: UFBoot does not converge, continue at least 100 more iterations\n",
      "Iteration 410 / LogL: -23042.142 / Time: 0h:0m:3s (0h:0m:0s left)\n",
      "Iteration 420 / LogL: -23042.188 / Time: 0h:0m:3s (0h:0m:0s left)\n",
      "Iteration 430 / LogL: -23049.858 / Time: 0h:0m:3s (0h:0m:0s left)\n",
      "Iteration 440 / LogL: -23042.177 / Time: 0h:0m:3s (0h:0m:0s left)\n",
      "Iteration 450 / LogL: -23042.195 / Time: 0h:0m:3s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -23066.308\n",
      "Iteration 460 / LogL: -23042.166 / Time: 0h:0m:4s (0h:0m:0s left)\n",
      "Iteration 470 / LogL: -23042.158 / Time: 0h:0m:4s (0h:0m:0s left)\n",
      "Iteration 480 / LogL: -23042.142 / Time: 0h:0m:4s (0h:0m:0s left)\n",
      "Iteration 490 / LogL: -23042.142 / Time: 0h:0m:4s (0h:0m:0s left)\n",
      "Iteration 500 / LogL: -23156.158 / Time: 0h:0m:4s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -23066.308\n",
      "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 0.914\n",
      "NOTE: UFBoot does not converge, continue at least 100 more iterations\n",
      "Iteration 510 / LogL: -23042.200 / Time: 0h:0m:4s (0h:0m:0s left)\n",
      "Iteration 520 / LogL: -23042.160 / Time: 0h:0m:4s (0h:0m:0s left)\n",
      "Iteration 530 / LogL: -23042.182 / Time: 0h:0m:4s (0h:0m:0s left)\n",
      "Iteration 540 / LogL: -23042.166 / Time: 0h:0m:4s (0h:0m:0s left)\n",
      "UPDATE BEST LOG-LIKELIHOOD: -23042.142\n",
      "Iteration 550 / LogL: -23042.142 / Time: 0h:0m:4s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -23066.308\n",
      "Iteration 560 / LogL: -23042.180 / Time: 0h:0m:4s (0h:0m:0s left)\n",
      "Iteration 570 / LogL: -23042.195 / Time: 0h:0m:5s (0h:0m:0s left)\n",
      "Iteration 580 / LogL: -23097.317 / Time: 0h:0m:5s (0h:0m:0s left)\n",
      "Iteration 590 / LogL: -23042.142 / Time: 0h:0m:5s (0h:0m:0s left)\n",
      "Iteration 600 / LogL: -23042.158 / Time: 0h:0m:5s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -23066.308\n",
      "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 0.919\n",
      "NOTE: UFBoot does not converge, continue at least 100 more iterations\n",
      "Iteration 610 / LogL: -23042.142 / Time: 0h:0m:5s (0h:0m:0s left)\n",
      "Iteration 620 / LogL: -23042.172 / Time: 0h:0m:5s (0h:0m:0s left)\n",
      "Iteration 630 / LogL: -23042.211 / Time: 0h:0m:5s (0h:0m:0s left)\n",
      "Iteration 640 / LogL: -23042.191 / Time: 0h:0m:5s (0h:0m:0s left)\n",
      "Iteration 650 / LogL: -23065.637 / Time: 0h:0m:5s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -23066.308\n",
      "Iteration 660 / LogL: -23042.182 / Time: 0h:0m:6s (0h:0m:0s left)\n",
      "Iteration 670 / LogL: -23042.142 / Time: 0h:0m:6s (0h:0m:0s left)\n",
      "Iteration 680 / LogL: -23042.142 / Time: 0h:0m:6s (0h:0m:0s left)\n",
      "Iteration 690 / LogL: -23042.161 / Time: 0h:0m:6s (0h:0m:0s left)\n",
      "Iteration 700 / LogL: -23042.235 / Time: 0h:0m:6s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -23066.308\n",
      "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 0.953\n",
      "NOTE: UFBoot does not converge, continue at least 100 more iterations\n",
      "Iteration 710 / LogL: -23042.177 / Time: 0h:0m:6s (0h:0m:0s left)\n",
      "Iteration 720 / LogL: -23042.142 / Time: 0h:0m:6s (0h:0m:0s left)\n",
      "Iteration 730 / LogL: -23042.142 / Time: 0h:0m:6s (0h:0m:0s left)\n",
      "Iteration 740 / LogL: -23042.222 / Time: 0h:0m:6s (0h:0m:0s left)\n",
      "Iteration 750 / LogL: -23042.182 / Time: 0h:0m:6s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -23066.330\n",
      "Iteration 760 / LogL: -23042.210 / Time: 0h:0m:7s (0h:0m:0s left)\n",
      "Iteration 770 / LogL: -23042.208 / Time: 0h:0m:7s (0h:0m:0s left)\n",
      "Iteration 780 / LogL: -23042.180 / Time: 0h:0m:7s (0h:0m:0s left)\n",
      "Iteration 790 / LogL: -23042.180 / Time: 0h:0m:7s (0h:0m:0s left)\n",
      "Iteration 800 / LogL: -23042.142 / Time: 0h:0m:7s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -23066.330\n",
      "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 0.970\n",
      "NOTE: UFBoot does not converge, continue at least 100 more iterations\n",
      "Iteration 810 / LogL: -23042.142 / Time: 0h:0m:7s (0h:0m:0s left)\n",
      "Iteration 820 / LogL: -23042.180 / Time: 0h:0m:7s (0h:0m:0s left)\n",
      "Iteration 830 / LogL: -23042.209 / Time: 0h:0m:7s (0h:0m:0s left)\n",
      "Iteration 840 / LogL: -23042.213 / Time: 0h:0m:8s (0h:0m:0s left)\n",
      "Iteration 850 / LogL: -23042.180 / Time: 0h:0m:8s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -23066.330\n",
      "Iteration 860 / LogL: -23042.235 / Time: 0h:0m:8s (0h:0m:0s left)\n",
      "Iteration 870 / LogL: -23042.158 / Time: 0h:0m:8s (0h:0m:0s left)\n",
      "Iteration 880 / LogL: -23042.210 / Time: 0h:0m:8s (0h:0m:0s left)\n",
      "Iteration 890 / LogL: -23042.158 / Time: 0h:0m:8s (0h:0m:0s left)\n",
      "Iteration 900 / LogL: -23042.142 / Time: 0h:0m:8s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -23066.372\n",
      "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 0.969\n",
      "NOTE: UFBoot does not converge, continue at least 100 more iterations\n",
      "Iteration 910 / LogL: -23042.217 / Time: 0h:0m:8s (0h:0m:0s left)\n",
      "Iteration 920 / LogL: -23042.195 / Time: 0h:0m:9s (0h:0m:0s left)\n",
      "Iteration 930 / LogL: -23042.211 / Time: 0h:0m:9s (0h:0m:0s left)\n",
      "Iteration 940 / LogL: -23042.186 / Time: 0h:0m:9s (0h:0m:0s left)\n",
      "Iteration 950 / LogL: -23049.949 / Time: 0h:0m:9s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -23066.372\n",
      "Iteration 960 / LogL: -23049.914 / Time: 0h:0m:9s (0h:0m:0s left)\n",
      "Iteration 970 / LogL: -23042.209 / Time: 0h:0m:9s (0h:0m:0s left)\n",
      "Iteration 980 / LogL: -23042.142 / Time: 0h:0m:9s (0h:0m:0s left)\n",
      "Iteration 990 / LogL: -23049.917 / Time: 0h:0m:9s (0h:0m:0s left)\n",
      "Iteration 1000 / LogL: -23042.142 / Time: 0h:0m:9s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -23066.372\n",
      "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 0.980\n",
      "NOTE: UFBoot does not converge, continue at least 100 more iterations\n",
      "WARNING: bootstrap analysis did not converge. You should rerun with higher number of iterations (-nm option)\n",
      "TREE SEARCH COMPLETED AFTER 1001 ITERATIONS / Time: 0h:0m:10s\n",
      "\n",
      "--------------------------------------------------------------------\n",
      "|                    FINALIZING TREE SEARCH                        |\n",
      "--------------------------------------------------------------------\n",
      "Performs final model parameters optimization\n",
      "Estimate model parameters (epsilon = 0.010)\n",
      "1. Initial log-likelihood: -23042.142\n",
      "Optimal log-likelihood: -23042.140\n",
      "Rate parameters:  A-C: 7.46234  A-G: 39.24238  A-T: 5.62208  C-G: 1.32173  C-T: 100.00000  G-T: 1.00000\n",
      "Base frequencies:  A: 0.292  C: 0.241  G: 0.235  T: 0.232\n",
      "Proportion of invariable sites: 0.441\n",
      "Gamma shape alpha: 9.260\n",
      "Parameters optimization took 1 rounds (0.002 sec)\n",
      "\n",
      "WARNING: Estimated model parameters are at boundary that can cause numerical instability!\n",
      "\n",
      "BEST SCORE FOUND : -23042.140\n",
      "Creating bootstrap support values...\n",
      "Split supports printed to NEXUS file ../../results/phylogeny/haplotype-sequences.fa.splits.nex\n",
      "Total tree length: 0.008\n",
      "\n",
      "Total number of iterations: 1001\n",
      "CPU time used for tree search: 9.553 sec (0h:0m:9s)\n",
      "Wall-clock time used for tree search: 9.891 sec (0h:0m:9s)\n",
      "Total CPU time used: 9.692 sec (0h:0m:9s)\n",
      "Total wall-clock time used: 10.082 sec (0h:0m:10s)\n",
      "\n",
      "Computing bootstrap consensus tree...\n",
      "Reading input file ../../results/phylogeny/haplotype-sequences.fa.splits.nex...\n",
      "17 taxa and 170 splits.\n",
      "Consensus tree written to ../../results/phylogeny/haplotype-sequences.fa.contree\n",
      "Reading input trees file ../../results/phylogeny/haplotype-sequences.fa.contree\n",
      "Log-likelihood of consensus tree: -23042.601\n",
      "\n",
      "Analysis results written to: \n",
      "  IQ-TREE report:                ../../results/phylogeny/haplotype-sequences.fa.iqtree\n",
      "  Maximum-likelihood tree:       ../../results/phylogeny/haplotype-sequences.fa.treefile\n",
      "  Likelihood distances:          ../../results/phylogeny/haplotype-sequences.fa.mldist\n",
      "\n",
      "Ultrafast bootstrap approximation results written to:\n",
      "  Split support values:          ../../results/phylogeny/haplotype-sequences.fa.splits.nex\n",
      "  Consensus tree:                ../../results/phylogeny/haplotype-sequences.fa.contree\n",
      "  Screen log file:               ../../results/phylogeny/haplotype-sequences.fa.log\n",
      "\n",
      "Date and Time: Mon Apr  3 11:22:12 2023\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Building the phylogeny with IQtree with 1000 bootstrap iterations with GTR+I+G (Invariable site plus discrete Gamma model) with asr\n",
    "alignfasta = os.path.join(outpath, \"haplotype-sequences.fa\")\n",
    "call(f\"iqtree -s {alignfasta} -m GTR+I+G -bb 1000 -redo\", shell=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b917138",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
