---
title: "10. Cluster Tissues Spatially"
author:
    - "Will Hannon"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
  highlight: tango
  number_sections: no
  theme: default
  toc: yes
  toc_depth: 3
  toc_float:
    collapsed: no
    smooth_scroll: yes
---

The goal of this notebook is visualize how alleles delineate spatial regions in the brain. I'll see how minor alleles cluster tissue samples in the brain and which sets of alleles relate samples to one another and separate samples from one another. 

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```

```{r Required Packages, message=FALSE, warning=FALSE, echo=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed

packages = c("tidyverse", "UpSetR", "ComplexHeatmap", "cluster", "ggrepel", "gridExtra", 'grid', 'uwot')
## Check that packages are installed, if not, install them
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}
## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```

```{r Inputs, echo = T}

## ==== File paths input data ==== ##

# Data from lofreq and varscan and ivar
variant.data = "../../results/variants/variants.csv"

# Data from lofreq and varscan labeled with G1 and G2 mutations
labeled.data = "../../results/variants/genotyped_variants.csv"

# Data from lofreq and varscan labeled with subclonal haplotypes
updated.data = "../../results/variants/clustered_variants.csv"

# Haplotypes parsed from CliqueSNV
haplotype.data = "../../results/haplotypes/haplotypes.csv"

# Import the sample data
sample.data = "../../config/samples.csv"

# Annotations 
annotations.filepath = "../../config/annotations.csv"

```


```{r Tissue Order, echo = F}

# Tissue order for plotting 
tissue_order = c(
  "Frontal Cortex 2", 
  "Frontal Cortex 1", 
  "Frontal Cortex 3", 
  "Temporal Lobe", 
  "Parietal Lobe",
  "Occipital Lobe",
  "Hippocampus",
  "Internal Capsule", 
  "Cerebellum",
  "Cerebellum Nucleus",
  "Midbrain", 
  "UBS",
  "Brain Stem")

```

## Clustering Tissues by SNPs

First, I do a PCA analysis to see which sets of SNPs explain the variance in allele frequencies between tissues.

```{r Process Data, echo=T, warning=F, message=F}
# Using the maximally labled data
updated.haplotype.df = read_csv(updated.data, show_col_types = F)

haplotypes.label = updated.haplotype.df %>% 
  select(SNP, Haplotype, Background) %>% 
  distinct()

# Expand the mutations to have a frequency for every tissue.
expanded.df = updated.haplotype.df %>% 
  select(SNP, Tissue, AF) %>% 
  pivot_wider(names_from = "Tissue", values_from = "AF", values_fill = 0) %>% 
  pivot_longer(cols = !SNP, names_to = "Tissue", values_to = "AF") %>% 
  left_join(., select(updated.haplotype.df, c("SNP", "Tissue", "DP")), by = c("SNP", "Tissue")) %>% 
  mutate(DP = if_else(is.na(DP), 0, DP)) %>% 
  left_join(., haplotypes.label, by = "SNP") 

# Get the mean frequency of the major genomes 
tissue.mean = updated.haplotype.df %>% 
  filter(Haplotype %in% c("genome-1", "genome-2")) %>% 
  group_by(Tissue, Haplotype) %>% 
  summarize(AF.mean = mean(AF, na.rm = TRUE), 
            SD = sd(AF, na.rm = TRUE),
            N = n()) %>% 
  mutate(SE = SD / sqrt(N),
         Lower.CI = qt(1 - (0.05 / 2), N - 1) * SE,
         Upper.CI = qt(1 - (0.05 / 2), N - 1) * SE) %>% 
  rename("AF" = AF.mean)

```

```{r fig.width=25, fig.height=15}

# Prepare the SNPs as column and Allele frequency as row
snp.by.freq = updated.haplotype.df %>% 
  filter(AF >= 0.05) %>% 
  filter(!Haplotype %in% c("fixed", "both", "genome-1", "genome-2", "genome-1-1")) %>%
  select(AF, SNP, Tissue) %>% 
  pivot_wider(names_from = SNP, values_from = AF, values_fill = 0)  %>% 
  remove_rownames %>% 
  column_to_rownames(var = "Tissue") %>% 
  t()

snp.pca = prcomp(snp.by.freq, center = TRUE, scale. = TRUE)

clusters = kmeans(snp.pca$x, centers = 25)

cluster.df = clusters$cluster %>% 
  as.list() %>% 
  as.tibble() %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "SNP") %>% 
  rename(cluster = "V1")

kmeans.cluster.haplotypes = expanded.df %>% 
  left_join(., cluster.df, by = "SNP") %>% 
  filter(!is.na(cluster)) %>% 
  mutate(cluster = as.character(cluster))

n.clusters.per.snp = kmeans.cluster.haplotypes %>% 
  select(SNP, cluster) %>%
  distinct() %>% 
  group_by(cluster) %>% 
  count() %>% 
  mutate(cluster_size = paste0(cluster, " (", n, " snps in cluster)"))

kmeans.cluster.haplotypes = left_join(kmeans.cluster.haplotypes, n.clusters.per.snp, by = "cluster")

# Mean of each tissue
kmeans.cluster.haplotypes.mean = kmeans.cluster.haplotypes %>% 
  group_by(Tissue, cluster, cluster_size) %>% 
  summarize(AF = mean(AF))

# Plot the clusters
kmeans.cluster.haplotypes %>% 
    ggplot(aes(x = Tissue, y = AF)) +
      geom_line(aes(group = SNP)) +
      geom_line(data = kmeans.cluster.haplotypes.mean, aes(x = Tissue, y = AF, group = 1, col = (cluster)), size = 1) +
      geom_ribbon(data = tissue.mean, aes(x=Tissue, ymin=AF-SD, ymax=AF+SD, group = Haplotype, fill = Haplotype), alpha=0.2, colour = NA) +
      facet_wrap(~cluster_size) + 
      xlim((tissue_order)) +
      scale_fill_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Haplotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))

ggsave(filename = "../../clusters.png", dpi = 300, width = 25, height = 15)


```


```{r PCA analysis, fig.align='center', fig.width = 12, fig.height = 8}

# Prepare the SNPs as column and Allele frequency as row
frequency.by.snp = updated.haplotype.df %>% 
  filter(!Haplotype %in% c("fixed", "both")) %>%
  select(AF, SNP, Tissue) %>% 
  pivot_wider(names_from = SNP, values_from = AF, values_fill = 0)  %>% 
  remove_rownames %>% 
  column_to_rownames(var = "Tissue")

# Do the PCA 
pcobj = prcomp(frequency.by.snp, center = TRUE, scale. = TRUE)

# PCs to plot
choices = 1:2

# Scaling factors 
scale = 1
var.scale = scale 
obs.scale = 1 - scale

# Variables for plotting
circle.prob = 0.69
varname.adjust = 1.5
varname.size = 3

# ----------

# Pull the data from the PCA object
nobs.factor = sqrt(nrow(pcobj$x) - 1)
d = pcobj$sdev
u = sweep(pcobj$x, 2, 1 / (d * nobs.factor), FUN = '*')
v = pcobj$rotation
 
# Scores
choices = pmin(choices, ncol(u))
df.u = as.data.frame(sweep(u[,choices], 2, d[choices]^obs.scale, FUN='*'))

# Directions
v = sweep(v, 2, d^var.scale, FUN='*')
df.v = as.data.frame(v[, choices])

names(df.u)  = c('xvar', 'yvar')
names(df.v) = names(df.u)

df.u = df.u * nobs.factor

# Scale directions
r = sqrt(qchisq(circle.prob, df = 2)) * prod(colMeans(df.u^2))^(1/4)
v.scale = rowSums(v^2)
df.v = r * df.v / sqrt(max(v.scale))

# Append the proportion of explained variance to the axis labels
u.axis.labs = paste('PC', choices, sep='')
u.axis.labs = paste(u.axis.labs, 
                    sprintf('(%0.1f%% explained var.)', 
                            100 * pcobj$sdev[choices]^2/sum(pcobj$sdev^2)))

# Variables for text label placement
df.v$varname = rownames(v)
df.v$angle = with(df.v, (180/pi) * atan(yvar / xvar))
df.v$hjust = with(df.v, (1 - varname.adjust * sign(xvar)) / 2)


# Add the data from the haplotypes
df.u = df.u %>% 
  rownames_to_column(var = "Tissue") 

df.v = df.v %>% 
  rownames_to_column(var = "SNP") %>% 
  left_join(., select(updated.haplotype.df, SNP, Background, Haplotype, Gene)) %>% 
  distinct()
  
# Plot the PCA
no.loadings = ggplot(data = df.u, aes(x = xvar, y = yvar)) + 
  xlab(u.axis.labs[1]) + 
  ylab(u.axis.labs[2]) +
  coord_equal() +
  geom_segment(data = df.v,
               aes(x = 0, y = 0, xend = xvar, yend = yvar),
               arrow = arrow(length = unit(1/2, 'picas')), col = "#FFFFFF") +
  geom_point(size = 1.5) + 
  geom_text_repel(aes(label = Tissue)) +
  # geom_text(data = df.v, 
  #           aes(label = varname, x = xvar, y = yvar, 
  #               angle = angle, hjust = hjust), 
  #           color = 'darkred', size = varname.size) + 
  # scale_color_manual(values = c("#307EC2", "#B63F3E", "#bdbdbd")) + 
  theme_bw(18) + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())


with.loadings.no.color = ggplot(data = df.u, aes(x = xvar, y = yvar)) + 
  xlab(u.axis.labs[1]) + 
  ylab(u.axis.labs[2]) +
  coord_equal() +
  geom_segment(data = df.v,
               aes(x = 0, y = 0, xend = xvar, yend = yvar),
               arrow = arrow(length = unit(1/2, 'picas')), col = "grey") + 
  geom_point(size = 1.5) + 
  geom_text_repel(aes(label = Tissue)) +
  # geom_text(data = df.v, 
  #           aes(label = varname, x = xvar, y = yvar, 
  #               angle = angle, hjust = hjust), 
  #           color = 'darkred', size = varname.size) + 
  # scale_color_manual(values = c("#307EC2", "#B63F3E", "#bdbdbd")) + 
  theme_bw(18) + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + 
  theme(legend.position = "none")

with.loadings.and.color = ggplot(data = df.u, aes(x = xvar, y = yvar)) + 
  xlab(u.axis.labs[1]) + 
  ylab(u.axis.labs[2]) +
  coord_equal() +
  geom_segment(data = df.v,
               aes(x = 0, y = 0, xend = xvar, yend = yvar, col = Haplotype),
               arrow = arrow(length = unit(1/2, 'picas'))) + 
  geom_point(size = 1.5) + 
  geom_text_repel(aes(label = Tissue)) +
  # geom_text(data = df.v, 
  #           aes(label = varname, x = xvar, y = yvar, 
  #               angle = angle, hjust = hjust), 
  #           color = 'darkred', size = varname.size) + 
  # scale_color_manual(values = c("#307EC2", "#B63F3E", "#bdbdbd")) + 
  theme_bw(18) + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + 
  theme(legend.position = "none")


ggsave(no.loadings, filename = "../../pca-no-loadings.png", dpi = 300, width = 12, height = 8)
ggsave(with.loadings.no.color, filename = "../../pca-with-loadings.png", dpi = 300, width = 12, height = 8)
ggsave(with.loadings.and.color, filename = "../../pca-with-color.png", dpi = 300, width = 12, height = 8)


```

What is the set of SNPs distinguishing OL and TL? 

```{r}

snps.ol.tl = df.v %>% 
  filter(xvar <= -.7 & yvar <= -.4) %>% 
  pull(SNP)

haplotypes.label = updated.haplotype.df %>% 
  select(SNP, Haplotype, Background) %>% 
  distinct()

# Expand the mutations to have a frequency for every tissue.
expanded.df = updated.haplotype.df %>% 
  select(SNP, Tissue, AF) %>% 
  pivot_wider(names_from = "Tissue", values_from = "AF", values_fill = 0) %>% 
  pivot_longer(cols = !SNP, names_to = "Tissue", values_to = "AF") %>% 
  left_join(., select(updated.haplotype.df, c("SNP", "Tissue", "DP")), by = c("SNP", "Tissue")) %>% 
  mutate(DP = if_else(is.na(DP), 0, DP)) %>% 
  left_join(., haplotypes.label, by = "SNP") 

# Get the mean frequency of the major genomes 
tissue.mean = updated.haplotype.df %>% 
  filter(Haplotype %in% c("genome-1", "genome-2")) %>% 
  group_by(Tissue, Haplotype) %>% 
  summarize(AF.mean = mean(AF, na.rm = TRUE), 
            SD = sd(AF, na.rm = TRUE),
            N = n()) %>% 
  mutate(SE = SD / sqrt(N),
         Lower.CI = qt(1 - (0.05 / 2), N - 1) * SE,
         Upper.CI = qt(1 - (0.05 / 2), N - 1) * SE) %>% 
  rename("AF" = AF.mean)

# Plot the clusters
expanded.df %>% 
  filter(SNP %in% snps.ol.tl) %>% 
    ggplot(aes(x = reorder(Tissue, -AF), y = AF)) +
      geom_line(aes(group = SNP, col = Haplotype)) +
      geom_ribbon(data = tissue.mean, aes(x=Tissue, ymin=AF-SD, ymax=AF+SD, group = Haplotype, fill = Haplotype), alpha=0.2, colour = NA) +
      scale_fill_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Haplotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))
```


What is the set of SNPs distinguishing the FL2 from everything else? 

```{r fig.align='center', fig.width = 10, fig.height = 8}

# Using the maximally labled data
updated.haplotype.df = read_csv(updated.data)

# SNP as column and Allele frequency as row
frequency.by.snp = updated.haplotype.df %>% 
  select(AF, SNP, Tissue) %>% 
  pivot_wider(names_from = SNP, values_from = AF, values_fill = 0)  %>% 
  remove_rownames %>% 
  column_to_rownames(var = "Tissue")

# With centering and scaling. 
brain.pca = prcomp(frequency.by.snp, center = TRUE, scale. = TRUE)

scores = brain.pca$x %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "Tissue")

loadings = as.data.frame(brain.pca$rotation) %>% 
  rownames_to_column(var = "SNP") %>% 
  left_join(., select(updated.haplotype.df, SNP, Background)) %>% 
  mutate(Background = case_when(Background == "genome-1" ~ "genome-1", 
                                Background == "genome-2" ~ "genome-2",
                                TRUE ~ "subclonal"))

# make biplot
scores %>% 
  ggplot(aes(x = PC1, y = PC2)) +
    geom_segment(data = loadings, aes(x = 0, y = 0, xend = PC1 * 175, yend = PC2 * 175, col = Background),
          arrow = arrow(length = unit(0.05, "cm"))) +
    geom_point(size = 1.5) + 
    geom_text_repel(aes(label = Tissue)) +
    scale_color_manual(values = c("#307EC2", "#B63F3E", "#bdbdbd")) + 
    theme_bw(18)

ggsave(filename = "../../pca.png", dpi = 300, width = 10, height = 8)


```


```{r fig.align='center', fig.width = 25, fig.height = 15}

# Colors for plotting
n = length(tissue_order)
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
cols = sample(col_vector, n)

# SNPs to plot 
snps.to.compare.df = updated.haplotype.df %>% 
  # Remove the SNPs on the background of most haplotypes
  filter(!Haplotype %in% c("genome-1", "genome-2", "genome-1-1", "both", "fixed"))

tissue.plots = list()
for (tissue.of.interest in tissue_order) {
  
  # Get the color of the tissue for plotting
  tissue.color = cols[which(tissue_order == tissue.of.interest)]
  
  # SNPs from the tissue to analyze
  snps.of.interest = snps.to.compare.df %>% 
    filter(Tissue == tissue.of.interest) %>% 
    pull(SNP)

  # Get the SNPs of interest in every sample
  tissue.of.interest.snps = snps.to.compare.df %>% 
    filter(SNP %in% snps.of.interest) %>% 
    mutate(TI = tissue.of.interest)
  
  # Get the SNPs that aren't SNPs of interest
  non.tissue.of.interest.snps = snps.to.compare.df %>% 
    filter(!SNP %in% snps.of.interest)

  # Make the plot
  plt = tissue.of.interest.snps %>% 
    ggplot(aes(x = Tissue, y = AF)) + 
      geom_jitter(data = non.tissue.of.interest.snps, aes(x = Tissue, y = AF), col = "#bdbdbd", alpha = 0.75, size = .75, width = 0.25) + 
      geom_jitter(col = "#B63F3E", alpha = 1, size = .75, width = 0.25) + 
      xlab("") + 
      ylab("") + 
      facet_wrap(~TI) + 
      coord_fixed(ratio = 10) + 
      xlim((tissue_order)) + 
      theme_bw() + 
      theme(axis.text.x = element_text(angle = 90, hjust=1, vjust = 0.5)) +
      theme(axis.text=element_text(size=8)) + 
      theme(plot.margin=unit(c(-.15, 0.2, -.15, -.25), "cm"))
  
  # Save the plot
  tissue.plots[[which(tissue_order == tissue.of.interest)]] = plt

}

final_plot = gridExtra::grid.arrange(arrangeGrob(grobs = tissue.plots, 
                          left = textGrob("Allele Frequency", rot = 90, vjust = 0.5,
                                          gp = gpar(col = "black", fontsize = 24)), 
                          bottom = textGrob("Tissue", hjust = .5,
                                            gp = gpar(col = "black", fontsize = 24))))


ggsave(final_plot, filename = "../../all-tissues.png", dpi = 300, width = 25, height = 15)

```

```{r fig.align='center', fig.width = 25, fig.height = 15}

tissue.plots = list()
for (tissue.of.interest in tissue_order) {
  
  # Get the color of the tissue for plotting
  tissue.color = cols[which(tissue_order == tissue.of.interest)]
  
  # SNPs from the tissue to analyze
  snps.in.tissue.of.interest = snps.to.compare.df %>% 
    filter(Tissue == tissue.of.interest) %>% 
    pull(SNP)
  
  # SNPs in the tissue of interest also in other samples
  snps.of.interest.in.other.tissues = snps.to.compare.df %>% 
    filter(Tissue != tissue.of.interest) %>% 
    filter(SNP %in% snps.in.tissue.of.interest) %>% 
    pull(SNP)
  
  freq.of.snps.df = snps.to.compare.df %>% 
    filter(SNP %in% snps.of.interest.in.other.tissues) %>% 
    filter(Tissue == tissue.of.interest) %>% 
    select(SNP, tissueAF = "AF")

  # Get the SNPs of interest in every sample
  tissue.of.interest.snps = snps.to.compare.df %>% 
    filter(SNP %in% snps.of.interest.in.other.tissues) %>% 
    mutate(TI = tissue.of.interest) %>% 
    left_join(., freq.of.snps.df, by = "SNP")
    
  
  # Get the SNPs that aren't SNPs of interest
  non.tissue.of.interest.snps = snps.to.compare.df %>% 
    filter(!SNP %in% snps.of.interest.in.other.tissues)

  # Make the plot
  plt = tissue.of.interest.snps %>% 
    ggplot(aes(x = Tissue, y = AF)) + 
      geom_jitter(data = non.tissue.of.interest.snps, aes(x = Tissue, y = AF), col = "#bdbdbd", alpha = 0.75, size = .75, width = 0.25) + 
      geom_jitter(aes(col = tissueAF), alpha = 1, size = .75, width = 0.25) + 
      xlab("") + 
      ylab("") + 
      facet_wrap(~TI) + 
      coord_fixed(ratio = 10) + 
      xlim((tissue_order)) +
      scale_color_gradient(low="blue", high="red") + 
      theme_bw() + 
      theme(axis.text.x = element_text(angle = 90, hjust=1, vjust = 0.5)) +
      theme(axis.text=element_text(size=8)) + 
      theme(plot.margin=unit(c(0, 0.2, -.15, -.25), "cm")) 
      # theme(legend.position = "none")
  
  # Save the plot
  tissue.plots[[which(tissue_order == tissue.of.interest)]] = plt

}

final_plot = gridExtra::grid.arrange(arrangeGrob(grobs = tissue.plots, 
                          left = textGrob("Allele Frequency", rot = 90, vjust = 0.5,
                                          gp = gpar(col = "black", fontsize = 24)), 
                          bottom = textGrob("Tissue", hjust = .5,
                                            gp = gpar(col = "black", fontsize = 24))))

ggsave(final_plot, filename = "../../all-tissues.png", dpi = 300, width = 25, height = 15)

```


```{r fig.align='center', fig.width = 10, fig.height = 8}

# Get the mean frequency of the major genomes 
no.color = updated.haplotype.df %>% 
    filter(!Haplotype %in% c("both", "fixed")) %>% 
    ggplot(aes(x = Tissue, y = AF)) + 
      geom_jitter(alpha = 1, size = 1.5, width = 0.25) + 
      xlab("") + 
      ylab("Allele Frequency") + 
      coord_fixed(ratio = 5) + 
      xlim((tissue_order)) +
      theme_bw(18) + 
      theme(axis.text.x = element_text(angle = 45, hjust=1)) +
      # theme(axis.text=element_text(size=8)) + 
      theme(legend.position = "none")
      
# Get the mean frequency of the major genomes 
color = updated.haplotype.df %>% 
    filter(!Haplotype %in% c("both", "fixed")) %>% 
    mutate(Haplotype = case_when(Haplotype %in% c("genome-1", "genome-2") ~ Haplotype,
                                  TRUE ~ "subclonal")) %>% 
    ggplot(aes(x = Tissue, y = AF)) + 
      geom_jitter(aes(col = Haplotype), alpha = 1, size = 1.5, width = 0.25) + 
      xlab("") + 
      ylab("Allele Frequency") + 
      coord_fixed(ratio = 5) + 
      xlim((tissue_order)) +
      scale_color_manual(values = c("#307EC2", "#B63F3E", "darkgrey")) + 
      theme_bw(18) + 
      theme(axis.text.x = element_text(angle = 45, hjust=1)) +
      # theme(axis.text=element_text(size=8)) + 
      theme(legend.position = "none")

ggsave(no.color, filename = "../../g1-g2-no-color-all-tissues.png", dpi = 300, width = 10, height = 8)
ggsave(color, filename = "../../g1-g2-color-all-tissues.png", dpi = 300, width = 10, height = 8)

```

```{r fig.align='center', fig.width = 10, fig.height = 8}


# Get the mean frequency of the major genomes 
updated.haplotype.df %>% 
  filter(!Haplotype %in% c("subclonal", "both", "fixed")) %>% 
  group_by(Tissue, Haplotype) %>% 
  summarize(AF.mean = mean(AF, na.rm = TRUE), 
            SD = sd(AF, na.rm = TRUE),
            N = n()) %>% 
  mutate(SE = SD / sqrt(N),
         Lower.CI = qt(1 - (0.05 / 2), N - 1) * SE,
         Upper.CI = qt(1 - (0.05 / 2), N - 1) * SE) %>% 
  rename("AF" = AF.mean) %>% 
  ggplot(aes(x = Tissue, y = AF, col = Haplotype)) + 
    geom_point(alpha = 1, size = 4,
                   position=position_dodge(0.5)) + 
    geom_errorbar(aes(ymin=AF-SD, ymax=AF+SD), width=1,
                   position=position_dodge(0.5)) +
    xlab("") + 
    ylab("") + 
    coord_fixed(ratio = 5) + 
    xlim((tissue_order)) +
    theme_bw() + 
    theme(axis.text.x = element_text(angle = 90, hjust=1, vjust = 0.5))

ggsave(filename = "../../all-tissues.png", dpi = 300, width = 10, height = 8)

```

```{r fig.align='center', fig.width = 10, fig.height = 8}

haplotypes.label = updated.haplotype.df %>% 
  select(SNP, Haplotype, Background) %>% 
  distinct()

# Expand the mutations to have a frequency for every tissue.
expanded.df = updated.haplotype.df %>% 
  select(SNP, Tissue, AF) %>% 
  pivot_wider(names_from = "Tissue", values_from = "AF", values_fill = 0) %>% 
  pivot_longer(cols = !SNP, names_to = "Tissue", values_to = "AF") %>% 
  left_join(., select(updated.haplotype.df, c("SNP", "Tissue", "DP")), by = c("SNP", "Tissue")) %>% 
  mutate(DP = if_else(is.na(DP), 0, DP)) %>% 
  left_join(., haplotypes.label, by = "SNP") 

# Get the mean frequency of the major genomes 
tissue.mean = updated.haplotype.df %>% 
  filter(Haplotype %in% c("genome-1", "genome-2")) %>% 
  group_by(Tissue, Haplotype) %>% 
  summarize(AF.mean = mean(AF, na.rm = TRUE), 
            SD = sd(AF, na.rm = TRUE),
            N = n()) %>% 
  mutate(SE = SD / sqrt(N),
         Lower.CI = qt(1 - (0.05 / 2), N - 1) * SE,
         Upper.CI = qt(1 - (0.05 / 2), N - 1) * SE) %>% 
  rename("AF" = AF.mean)

# Mean of each tissue
g1.1.mean = expanded.df %>% 
  filter(Haplotype == "genome-1-1") %>% 
  group_by(Tissue, Haplotype) %>% 
  summarize(AF = mean(AF))

# Plot the clusters
expanded.df %>% 
  filter(Haplotype == "genome-1-1") %>% 
    ggplot(aes(x = (Tissue), y = AF)) +
      geom_line(aes(group = SNP), col = "darkgrey") +
      geom_ribbon(data = tissue.mean, aes(x=Tissue, ymin=AF-SD, ymax=AF+SD, group = Haplotype, fill = Haplotype), alpha=0.2, colour = NA) +
      geom_line(data = g1.1.mean, aes(x = Tissue, y = AF, group = 1), size = 1.5, col = "darkgreen") +
      scale_fill_manual(values=c("#424ef5", "#cf1919")) + 
      xlim((tissue_order)) +
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Haplotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 45, hjust=1))  + 
      theme(legend.title = element_text(size = 0))

ggsave(filename = "../../genome-1-1.png", dpi = 300, width = 12, height = 8)

```


## UMAP

I made a UMAP plot of the PCA loading of every allele over 5% and colored the points by a kmeans cluster with 15 centers. It would be interesting to see how this stacks up to the other ways we've tried to cluster these points. 

```{r fig.width=15, fig.height=8}

# Prepare the SNPs as column and Allele frequency as row
snp.by.freq = updated.haplotype.df %>% 
  filter(AF >= 0.05) %>% 
  filter(!Haplotype %in% c("fixed", "both")) %>%
  select(AF, SNP, Tissue) %>% 
  pivot_wider(names_from = SNP, values_from = AF, values_fill = 0)  %>% 
  remove_rownames %>% 
  column_to_rownames(var = "Tissue") %>% 
  t()

snp.pca = prcomp(snp.by.freq, center = TRUE, scale. = TRUE)

clusters = kmeans(snp.pca$x, centers = 15)

cluster.df = clusters$cluster %>% 
  as.list() %>% 
  as.tibble() %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "SNP") %>% 
  rename(cluster = "V1")


umap(snp.pca$x) %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "SNP") %>% 
  left_join(., select(updated.haplotype.df, SNP, Haplotype)) %>% 
  distinct() %>% 
  left_join(., cluster.df, by = "SNP") %>% 
  ggplot(aes(x = V1, y = V2, col = as.factor(cluster))) + 
    geom_point() +
    coord_fixed() + 
    theme_bw()

```

