---
title: "11. Cluster Tissues Spatially"
author:
    - "Will Hannon"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
  highlight: tango
  number_sections: no
  theme: default
  toc: yes
  toc_depth: 3
  toc_float:
    collapsed: no
    smooth_scroll: yes
---

The goal of this notebook is to take an unbiased approach to show how similar tissue compartments have similar viral populations. We'll do this by clustering tissues based on the frequency of SNPs in each sample.

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```

```{r Required Packages, message=FALSE, warning=FALSE, echo=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed

packages = c("tidyverse", "UpSetR", "ComplexHeatmap", "cluster", "ggrepel", "gridExtra", 'grid', 'RColorBrewer')
## Check that packages are installed, if not, install them
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}
## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```

```{r Inputs, echo=T}

## ==== File paths input data ==== ##

if (exists("snakemake")) {

  # Data from lofreq and varscan labeled with subclonal haplotypes
  updated.data = snakemake@input[["1"]]

} else {

  # Data from lofreq and varscan labeled with subclonal haplotypes
  updated.data = "../../results/variants/validated_variants.csv"

}

```

```{r Color Pallet, echo=F}

haplotype.colors = c("#989891", "#000000", "#22488F", "#B02B23", "#E1EAF6",  "#A6C9DF", "#7AADD2","#5691C1", "#3871B0", "#67649F", "#8484B1", "#B6B7D1", "#FFFCBA", "#F9D984", "#F4B460", "#EE9B4F", "#DE4B32")

names(haplotype.colors) = c("genome 01", "cluster 6", "genome 1", "genome 2", "cluster 1", "cluster 5", "cluster 2", "cluster 3", "cluster 4", "cluster 6", "cluster 7", "cluster 8", "cluster 9", "cluster 10", "cluster 11", "cluster 12", "cluster 13")

```

## Clustering Tissue Samples

First, I do a PCA analysis on the frequency of mutations in each compartment to see if variance in allele frequency between tissues captures any level of spatial information. 

```{r Process Data, echo=F, warning=F, message=F}

# Final set of haplotype assignments
updated.df = read_csv(updated.data, show_col_types = F) %>% 
  # For consistency, change the names of G1, G2, and G-1-1
  mutate(Haplotype = case_when(
    Haplotype == "genome-1" ~ "genome 01",
    Haplotype == "genome-1-1" ~ "genome 1",
    Haplotype == "genome-2" ~ "genome 2",
    T ~ Haplotype
  )) 

# Prepare the SNPs as column and Allele frequency as row
frequency.by.snp = updated.df %>% 
  # filter(!Haplotype %in% c("fixed", "both")) %>%
  select(AF, SNP, Tissue) %>% 
  pivot_wider(names_from = SNP, values_from = AF, values_fill = 0)  %>% 
  remove_rownames %>% 
  column_to_rownames(var = "Tissue")

```

```{r PCA analysis, message=F, echo=T}

# Do the PCA 
pcobj = prcomp(frequency.by.snp, center = TRUE, scale. = TRUE)

# PCs to plot
choices = 1:2

# Scaling factors 
scale = 1
var.scale = scale 
obs.scale = 1 - scale

# Variables for plotting
circle.prob = 0.69
varname.adjust = 1.5
varname.size = 3

# ----------

# Pull the data from the PCA object
nobs.factor = sqrt(nrow(pcobj$x) - 1)
d = pcobj$sdev
u = sweep(pcobj$x, 2, 1 / (d * nobs.factor), FUN = '*')
v = pcobj$rotation
 
# Scores
choices = pmin(choices, ncol(u))
df.u = as.data.frame(sweep(u[,choices], 2, d[choices]^obs.scale, FUN='*'))

# Directions
v = sweep(v, 2, d^var.scale, FUN='*')
df.v = as.data.frame(v[, choices])

names(df.u)  = c('xvar', 'yvar')
names(df.v) = names(df.u)

df.u = df.u * nobs.factor

# Scale directions
r = sqrt(qchisq(circle.prob, df = 2)) * prod(colMeans(df.u^2))^(1/4)
v.scale = rowSums(v^2)
df.v = r * df.v / sqrt(max(v.scale))

# Append the proportion of explained variance to the axis labels
u.axis.labs = paste('PC', choices, sep='')
u.axis.labs = paste(u.axis.labs, 
                    sprintf('(%0.1f%% explained var.)', 
                            100 * pcobj$sdev[choices]^2/sum(pcobj$sdev^2)))

# Variables for text label placement
df.v$varname = rownames(v)
df.v$angle = with(df.v, (180/pi) * atan(yvar / xvar))
df.v$hjust = with(df.v, (1 - varname.adjust * sign(xvar)) / 2)

# Add the data from the haplotypes
df.u = df.u %>% 
  rownames_to_column(var = "Tissue") 

df.v = df.v %>% 
  rownames_to_column(var = "SNP") %>% 
  left_join(., select(updated.df, SNP, Background, Haplotype, Gene)) %>% 
  distinct()

```

First, let's see what the PCA plot looks like with no loadings superimposed on the plot. The closer the tissues are to each other, the more similar the variant populations.  

```{r PCA no loadings, echo = F, fig.align='center', fig.width = 12, fig.height = 8}

# Plot the PCA
ggplot(data = df.u, aes(x = xvar, y = yvar)) + 
  xlab(u.axis.labs[1]) + 
  ylab(u.axis.labs[2]) +
  coord_equal() +
  geom_segment(data = df.v,
               aes(x = 0, y = 0, xend = xvar, yend = yvar),
               arrow = arrow(length = unit(1/2, 'picas')), col = "#FFFFFF") +
  geom_point(size = 1.5) + 
  geom_text_repel(aes(label = Tissue)) +
  # geom_text(data = df.v, 
  #           aes(label = varname, x = xvar, y = yvar, 
  #               angle = angle, hjust = hjust), 
  #           color = 'darkred', size = varname.size) + 
  # scale_color_manual(values = c("#307EC2", "#B63F3E", "#bdbdbd")) + 
  theme_bw(18) + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())

```

Now, let's add the loadings to the plot. These are the individual SNPs. 

```{r PCA with loadings, echo = F, fig.align='center', fig.width = 12, fig.height = 8}

ggplot(data = df.u, aes(x = xvar, y = yvar)) + 
  xlab(u.axis.labs[1]) + 
  ylab(u.axis.labs[2]) +
  coord_equal() +
  geom_segment(data = df.v,
               aes(x = 0, y = 0, xend = xvar, yend = yvar),
               arrow = arrow(length = unit(1/2, 'picas')), col = "grey") + 
  geom_point(size = 1.5) + 
  geom_text_repel(aes(label = Tissue)) +
  # geom_text(data = df.v, 
  #           aes(label = varname, x = xvar, y = yvar, 
  #               angle = angle, hjust = hjust), 
  #           color = 'darkred', size = varname.size) + 
  # scale_color_manual(values = c("#307EC2", "#B63F3E", "#bdbdbd")) + 
  theme_bw(18) + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + 
  theme(legend.position = "none")

```

We can see that there are clear groups of SNPs that separate tissues from one another. Let's see if some of these are explained by the haplotype labels. 

```{r PCA with labeled loadings, echo = F, fig.align='center', fig.width = 12, fig.height = 10}

ggplot(data = df.u, aes(x = xvar, y = yvar)) + 
  xlab(u.axis.labs[1]) + 
  ylab(u.axis.labs[2]) +
  coord_equal() +
  geom_segment(data = df.v,
               aes(x = 0, y = 0, xend = xvar, yend = yvar, col = Haplotype),
               arrow = arrow(length = unit(1/2, 'picas'))) + 
  geom_point(size = 1.5) + 
  geom_text_repel(aes(label = Tissue)) +
  scale_color_manual(values = haplotype.colors) +
  # geom_text(data = df.v, 
  #           aes(label = varname, x = xvar, y = yvar, 
  #               angle = angle, hjust = hjust), 
  #           color = 'darkred', size = varname.size) + 
  # scale_color_manual(values = c("#307EC2", "#B63F3E", "#bdbdbd")) + 
  theme_bw(18) + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + 
  theme(legend.position = "bottom")

```

Let's look at only the haplotype loadings - i.e. the SNPs that we've been able to haplotype.

```{r PCA with haplotype loadings, echo = F, fig.align='center', fig.width = 12, fig.height = 10}

ggplot(data = df.u, aes(x = xvar, y = yvar)) + 
  xlab(u.axis.labs[1]) + 
  ylab(u.axis.labs[2]) +
  coord_equal() +
  geom_segment(data = filter(df.v, !Haplotype %in% c('subclonal', 'both')),
               aes(x = 0, y = 0, xend = xvar, yend = yvar, col = Haplotype),
               arrow = arrow(length = unit(1/2, 'picas'))) + 
  geom_point(size = 1.5) + 
  geom_text_repel(aes(label = Tissue)) +
  scale_color_manual(values = haplotype.colors) +
  # geom_text(data = df.v, 
  #           aes(label = varname, x = xvar, y = yvar, 
  #               angle = angle, hjust = hjust), 
  #           color = 'darkred', size = varname.size) + 
  # scale_color_manual(values = c("#307EC2", "#B63F3E", "#bdbdbd")) + 
  theme_bw(18) + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + 
  theme(legend.position = "bottom")

```

Finally, let's look at the loadings that are subclonal. Is there some structure that we've missed? 

```{r PCA with subclonal loadings, echo = F, fig.align='center', fig.width = 12, fig.height = 10}

ggplot(data = df.u, aes(x = xvar, y = yvar)) + 
  xlab(u.axis.labs[1]) + 
  ylab(u.axis.labs[2]) +
  coord_equal() +
  geom_segment(data = filter(df.v, Haplotype == 'subclonal'),
               aes(x = 0, y = 0, xend = xvar, yend = yvar, col = Haplotype),
               arrow = arrow(length = unit(1/2, 'picas'))) + 
  geom_point(size = 1.5) + 
  geom_text_repel(aes(label = Tissue)) +
  scale_color_manual(values = haplotype.colors) +
  # geom_text(data = df.v, 
  #           aes(label = varname, x = xvar, y = yvar, 
  #               angle = angle, hjust = hjust), 
  #           color = 'darkred', size = varname.size) + 
  # scale_color_manual(values = c("#307EC2", "#B63F3E", "#bdbdbd")) + 
  theme_bw(18) + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + 
  theme(legend.position = "bottom")

```
