---
title: "7. Analyze Spruce Output"
author:
    - "Will Hannon"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
  highlight: tango
  number_sections: no
  theme: default
  toc: yes
  toc_depth: 3
  toc_float:
    collapsed: no
    smooth_scroll: yes
---

The goal of this notebook is to visualize the phylogenetic relationship of the haplotypes identified in the previous notebooks. I'll read in the tree's from `SPRUCE` as implemented by `MACHINA` and make a nicely formatted tree where branch lenght corresponds to mutations. 

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```

```{r Required Packages, message=FALSE, warning=FALSE, echo=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed
packages = c("tidyverse", "foreach", "emdbook", "RColorBrewer", "data.table", "ggnetwork", "network", "tidytree", "phytools", "ggtree", "treeio")

## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```

```{r Inputs, echo=T}

## ==== File paths input data ==== ##

if (exists("snakemake")) {

  # Data from lofreq and varscan labeled with subclonal haplotypes
  updated.data = snakemake@input[["1"]]
  # Phylogeny 
  phylo.data =  snakemake@input[["2"]]


} else {

  # Data from lofreq and varscan labeled with subclonal haplotypes
  updated.data = "../../results/variants/validated_variants.csv"

  # Spruce 
  spruce.data = "../../results/variants/spruce_output.tsv"
  
  # Phylogeny 
  phylo.data = "../../results/phylogeny/haplotype-sequences.fa.treefile"

}

```

## SPRUCE Trees

A set of phylogenetic relationships were fit to the data using the algorithm `SPRUCE`. We'll visualize these possible trees below. Ultimately, the goal is filter these based on bridging reads. 

```{r Read Spruce, echo=T}

# Code adapted from Alison Feder #

# Read in the output from Machina
machina.output = readLines( "../../results/variants/spruce_output.tsv")

# Process the output trees into edges
treenum = NA
tree.df = foreach(i = 1:length(machina.output), .combine = "rbind") %do% {
    toParse = machina.output[i]
    if(grepl("^#", toParse)){ return(NULL) }
    if(grepl("#trees", toParse)){ return(NULL) }
    if(grepl("#edges", toParse)){
        treenum = gsub("[0-9]+ \\#edges, ", "", toParse)
        return(NULL)
    }
    tibble(toParse) %>%
      separate(toParse, into = c("from", "to"), sep = " ") %>%
        mutate(tree = treenum)
}
tree.df = tree.df %>%
  mutate(treenum = as.numeric(gsub("tree ", "", tree)))

head(tree.df)

```

Let's visualize these trees below. 

```{r Color Palletes, echo=F}

# Make a color scheme for the haplotypes
## Genome-1
genome.1.haplotypes = c( "cluster 1", "cluster 2", "cluster 3", "cluster 4", "cluster 5", "cluster 7", "cluster 8", "genome 1", "genome 01")
genome.1.colors = brewer.pal(length(genome.1.haplotypes), "Blues")
names(genome.1.colors) = genome.1.haplotypes

## Genome-2
genome.2.haplotypes = c("cluster 9", "cluster 10", "cluster 11", "cluster 12", "cluster 13", "genome 2")
genome.2.colors = brewer.pal(length(genome.2.haplotypes), "YlOrRd")
names(genome.2.colors) = genome.2.haplotypes

## Cluster 6
cluster.6.haplotype = c("cluster 6")
cluster.6.color = brewer.pal(8, "Purples")[8]
names(cluster.6.color) = cluster.6.haplotype

## Combine
haplotype.colors = c(genome.1.colors, genome.2.colors, cluster.6.color)
# names(haplotype.colors) = gsub(" ", "_", names(haplotype.colors ))
haplotype.colors = c(haplotype.colors , "Anc" = "black")
# cluster 1 is too light
haplotype.colors['cluster 1'] = "#bdbebf"

## Set the Compartment Colors 
# Frontal Cortex
FCs = c("Frontal Cortex 1", "Frontal Cortex 2", "Frontal Cortex 3")
FCs.colors = brewer.pal(length(FCs), "Greens")
names(FCs.colors) = FCs
# Internal Compartments
Internal = c("Cerebellum Nucleus", "Cerebellum", "Midbrain", "Internal Capsule", "Brain Stem", "UBS", "Hippocampus")
Internal.colors = brewer.pal(length(Internal), "YlOrRd")
names(Internal.colors) = Internal
# Other Lobes
Other =  c("Occipital Lobe", "Temporal Lobe", "Parietal Lobe")
Other.colors = brewer.pal(length(Other), "Blues")
names(Other.colors) = Other
# Combine
Compartment.cols = c(FCs.colors, Internal.colors, Other.colors)
names(Compartment.cols) = gsub(" ", "_", names(Compartment.cols))

```

```{r Spruce Trees, message=F, warning=F, fig.align='center', fig.width=25, fig.height=25, echo=T}

names(haplotype.colors) = gsub(" ", "_", names(haplotype.colors ))
ggnetwork(network(tree.df, multiple = TRUE), by = "treenum")  %>%
  ggplot(aes(x = x, y = y, xend = xend, yend = yend)) +
    theme_blank() +
    geom_edges(curvature = 0.15) +
    geom_nodes(aes(fill = vertex.names), size = 10, pch = 21) +
    scale_fill_manual(values = haplotype.colors) +
    geom_nodetext(aes(y = y - 0.06, label =  vertex.names), color = "black", size = 2) +
    facet_wrap(~treenum) +
    theme(legend.position = "none")

```
The first tree fit's our current assumptions about the data. 

```{r Example Tree, message=F, warning=F, fig.align='center', fig.width=6, fig.height=6, echo=T}

tree.df = tree.df %>% 
  mutate(from = str_replace_all(from, "_", " "),
         to = str_replace_all(to, "_", " ")
         )

names(haplotype.colors) = gsub("_", " ", names(haplotype.colors))

ggnetwork(network(tree.df %>% filter(treenum == 1), multiple = TRUE), by = "treenum")  %>%
  ggplot(aes(x = x, y = y, xend = xend, yend = yend)) +
    theme_blank() +
    geom_edges(curvature = 0.15) +
    geom_nodes(aes(fill = vertex.names), size = 15, pch = 21) +
    scale_fill_manual(values = haplotype.colors) +
    geom_nodetext(aes(y = y - 0.06, label =  vertex.names), color = "black", size = 4) +
    theme(legend.position = "none")

ggsave("./example_tree.png", width = 8, height = 8)

```

## Visualize with IQ-TREE

But, what we're missing is branch length. I took the sequences for each of these haplotypes and generated a phylogenetic tree using `iqtree`. 

```{r Import Data, echo=T}

# Assigned SNPs with haplotype labels 
updated.df = read_csv(updated.data, show_col_types = FALSE)


# Get the background of the haplotypes
background.label = updated.df %>% 
  select(Haplotype_Name, Background) %>% 
  filter(!Haplotype_Name %in% c("subclonal", "genome 2", "genome 01", "genome 1", "both")) %>% 
  mutate(Haplotype_Name = str_replace(Haplotype_Name, " ", "_")) %>% 
  distinct()

# Phylogenetic tree 
tree = read.iqtree(phylo.data)

```

```{r Root the Tree, echo=T}

# Name of the out group for rooting the tree
root = "sspe_ancestor"

# Root the tree
tree@phylo = root(tree@phylo, outgroup = root)

```

```{r Add data to the tree, echo=T}

# Add the background labels
tree.df = left_join(as_tibble(tree), background.label, by=c("label" = "Haplotype_Name")) 

```

```{r Label Backgrounds, echo=T}

# Get the nodes in the tree
genome.1.node = MRCA(tree, filter(background.label, Background == "genome-1")$Haplotype_Name)
genome.2.node = MRCA(tree, filter(background.label, Background == "genome-2")$Haplotype_Name)
genome.1.offspring = offspring(tree, genome.1.node)
genome.2.offspring = offspring(tree, genome.2.node)

# Add the background labels
tree.df = as_tibble(tree) %>% 
  mutate(branch.length = if_else(UFboot < 70 & !is.na(UFboot), 0.0, branch.length)) %>% 
  mutate(Background = case_when(node %in% genome.1.offspring ~ "genome-1",
                                node %in% genome.2.offspring ~ "genome-2",
                                TRUE ~ "ancestral")) %>% 
  mutate(label = str_replace(label, "_", " ")) %>% 
  mutate(label = if_else(label == "sspe ancestor", "SSPE ancestor", label))

```

```{r Plot the Tree, fig.align="center", fig.width=32, fig.height=15}

tree.plt = tree.df %>% 
  as.treedata() %>% 
  ggtree(size = 3, aes(color = Background)) +
    geom_point2(aes(subset=(node==1)), size=12, color="#4d4d4d") + 
    geom_point2(aes(subset=(node==genome.1.node)), size=12, color="#307EC2") + 
    geom_point2(aes(subset=(node==genome.2.node)), size=12, color="#B63F3E") + 
    geom_point2(aes(subset=(node==23)), size=12, color="#199400") + 
    geom_tiplab(hjust = -.05, size = 8) + 
    scale_color_manual(values = c("#4d4d4d", "#307EC2", "#B63F3E")) + 
    theme(legend.position = "none") 

ggsave("../../results/phylogeny/sspe-phylogeny.png", width = 32, height = 15, dpi = 300)

```

```{r Make a Phylogeny for Each Tissue, echo=T}

mean.haplotype.freqs = updated.df %>% 
  filter(!Haplotype %in% c("subclonal", "both")) %>% 
  select(AF, Tissue, Background, Haplotype) %>% 
  group_by(Tissue, Background, Haplotype) %>% 
  mutate(AF = mean(AF)) %>% 
  distinct() %>% 
  mutate(Background = case_when(Haplotype == "cluster 13" ~ "cluster 9",
                                Haplotype %in% c("cluster 1",
                                                 "cluster 2",
                                                 "cluster 4",
                                                 "cluster 11",
                                                 "cluster 12",
                                                 "cluster 14") ~ "genome-1-1",
                                Haplotype == "cluster 3" ~ "cluster 2", 
                                TRUE ~ Background)) %>% 
  ungroup()

tissues = unique(mean.haplotype.freqs$Tissue)
tissue.trees = list()
for (i in 1:length(tissues)) {
  
  tissue = tissues[i]
  
  compartment.haplotypes = mean.haplotype.freqs %>% 
    filter(Tissue == tissue) %>% 
    select(AF, Haplotype, Tissue)
  
  compartment.tree = left_join(tree.df, compartment.haplotypes, by = c("label" = "Haplotype")) %>% 
    mutate(AF = if_else(is.na(AF), 0, AF), 
           Tissue = if_else(is.na(Tissue), tissue, Tissue),
           label = str_remove(label, "cluster "), 
           label = if_else(label == "SSPE ancestor", "", label))
  
  tissue.trees[[i]] = as.treedata(compartment.tree)
  
}

class(tissue.trees) = "multiPhylo"

```

```{r Standard, fig.align="center", fig.width=10, fig.height=10}

tissue.trees %>% 
  ggtree(size = 1, aes(color = Background)) +
    geom_point2(aes(subset=(node==1)), size=5, color="#4d4d4d") + 
    geom_tiplab(hjust = -.6, size = 5) + 
    geom_tippoint(aes(size = AF)) + 
    facet_wrap( ~Tissue,) + 
    scale_color_manual(values = c("#4d4d4d", "#307EC2", "#B63F3E")) +
    theme(legend.position = "none") 


ggsave("../../results/phylogeny/sspe-phylogeny-faceted.png", width = 14, height = 14, dpi = 300)

```

```{r With y-axis, fig.align="center", fig.width=7, fig.height=7}

haplotype.colors = c("1" = "#800000",
                     "2" = "#9A6324",
                     "3" = "#808000",
                     "4" = "#469990",
                     "5" = "#000075",
                     "6" = "#e6194B",
                     "7" = "#f58231",
                     "8" = "#ffe119",
                     "9" = "#bfef45",
                     "10" = "#911eb4",
                     "11" = "#dcbeff",
                     "12" = "#ffd8b1",
                     "13" = "#42d4f4",
                     "14" = "#aaffc3")

haplotypes = c("1",
               "2",
               "3",
               "4",
               "5",
               "6",
               "7",
               "8",
               "9",
               "10",
               "11",
               "12",
               "13",
               "14")


tissue.trees %>% 
  ggtree(size = 1, yscale = "AF", color="#4d4d4d") +
    geom_point2(aes(subset=(node==1)), size=5, color="#4d4d4d") + 
    geom_tippoint(aes(size = AF, color = label)) + 
    facet_wrap( ~Tissue,) + 
    scale_color_manual(breaks = haplotypes, values = haplotype.colors, name = "Haplotype") +
    theme_minimal() +
    theme(legend.position = "bottom") 


ggsave("../../results/phylogeny/sspe-phylogeny-faceted-yaxis.png", width = 14, height = 14, dpi = 300)

```

## "Muller" Plot

```{r Standardize to Backgrounds}

# Everything is on these backgrounds
backgrounds = unique(mean.haplotype.freqs$Background)

set.backgrounds.to.zero = mean.haplotype.freqs %>% 
  mutate(AF = case_when(
    Haplotype == "genome-1" ~ 0.0, 
    Haplotype == "genome-2" ~ 0.0, 
    Haplotype == "genome-1-1" ~ 0.0,
    TRUE ~ AF
  )) %>% 
  pivot_wider(names_from = "Tissue", values_from = "AF", values_fill = 0) %>% 
  pivot_longer(cols = !c("Haplotype", "Background"), names_to = "Tissue", values_to = "AF") %>% 
  arrange(Tissue)
  
final.df = data.frame()
tissues = unique(set.backgrounds.to.zero$Tissue)
for (i in 1:length(tissues)) {
  
  tissue = tissues[i]
  
  tissue.df = set.backgrounds.to.zero %>% 
    filter(Tissue == tissue)
  
  # Normalize to cluster parent
  cluster.13.af = tissue.df %>% filter(Haplotype == "cluster 13") %>% pull(AF)
  cluster.3.af = tissue.df %>% filter(Haplotype == "cluster 3") %>% pull(AF)
  
  normalize.df = tissue.df %>% 
    mutate(AF = case_when(
      Haplotype == "cluster 2" ~ AF - cluster.3.af,
      Haplotype == "cluster 9" ~ AF - cluster.13.af,
      TRUE ~ AF
    )) %>% 
    mutate(Background = case_when(
      Haplotype == "cluster 3" ~ "genome-1",
      Haplotype == "cluster 13" ~ "genome-1",
      TRUE ~ Background
    ))
  
  # Add it up to 1
  genome.1.total = mean.haplotype.freqs %>%
    filter(Tissue == tissue & Haplotype == "genome-1") %>%
    pull(AF)
  
  genome.2.total = mean.haplotype.freqs %>%
    filter(Tissue == tissue & Haplotype == "genome-2") %>%
    pull(AF)
  
  background.freqs = normalize.df %>%
    mutate(Background = if_else(Background == "genome-1-1", "genome-1", Background)) %>% 
    group_by(Background) %>% 
    summarise(AF = sum(AF))
  
  if (tissue != "Frontal Cortex 2") {
    
    normalize.df = normalize.df %>% 
      mutate(AF = case_when(
        Haplotype == "genome-1-1" ~ genome.1.total - filter(background.freqs, Background == "genome-1")$AF,
        Haplotype == "genome-2" ~ genome.2.total - filter(background.freqs, Background == "genome-2")$AF,
        TRUE ~ AF
      ))
    
  } else {
    
    normalize.df = normalize.df %>% 
      mutate(AF = case_when(
        Haplotype == "genome-1" ~ genome.1.total - filter(background.freqs, Background == "genome-1")$AF,
        Haplotype == "genome-2" ~ genome.2.total - filter(background.freqs, Background == "genome-2")$AF,
        TRUE ~ AF
      ))
    
  }
  
  final.df = rbind(final.df, normalize.df)
 
  
}


```

```{r Stacked Bar, fig.align="center", fig.width=10, fig.height=7}

haplotype.colors = c("cluster 1" = "#800000",
                     "cluster 2" = "#9A6324",
                     "cluster 3" = "#808000",
                     "cluster 4" = "#469990",
                     "cluster 5" = "#000075",
                     "cluster 6" = "#e6194B",
                     "cluster 7" = "#f58231",
                     "cluster 8" = "#ffe119",
                     "cluster 9" = "#bfef45",
                     "cluster 10" = "#911eb4",
                     "cluster 11" = "#dcbeff",
                     "cluster 12" = "#ffd8b1",
                     "cluster 13" = "#42d4f4",
                     "cluster 14" = "#aaffc3", 
                     "genome-1-1" = "#199400",
                     "genome-1" = "#307EC2",
                     "genome-2" = "#B63F3E")

require(RColorBrewer)

#G2 colors
g2s <- brewer.pal(6, "YlOrRd")
names(g2s) <- c("cluster 5", "cluster 6", "cluster 7", "cluster 8", "cluster 10", "genome-2")

#G1
g11cols <- brewer.pal(9, "Blues")
names(g11cols) <- c( "cluster 1", "cluster 2", "cluster 3", "cluster 4", "cluster 11", "cluster 12", "cluster 14", "genome-1-1", "genome-1")

#cluster 9 and 13
c913 <- brewer.pal(8, "Purples")[c(7,8)]
names(c913) <- c("cluster 9", "cluster 13")

cols <- c(g2s, g11cols, c913)


haplotype.levels = c("cluster 1",
                     "cluster 2",
                     "cluster 3",
                     "cluster 4",
                     "cluster 11",
                     "cluster 12",
                     "cluster 14", 
                     "cluster 9",
                     "cluster 13",
                     "genome-1-1",
                     "genome-1",
                     "genome-2",
                     "cluster 5",
                     "cluster 6",
                     "cluster 7",
                     "cluster 8",
                     "cluster 10")

final.df %>% 
  mutate(AF = if_else(AF < 0, 0.0, AF)) %>% 
  ggplot(aes(x = Tissue, y = AF, fill = factor(Haplotype, levels = haplotype.levels))) + 
    geom_bar(stat = "identity", position = position_stack()) + 
    scale_fill_manual(values = cols) + 
    scale_y_continuous(breaks = c(0, .5, 1.0)) +
    theme_classic() + 
    labs(fill="Haplotype") +
    xlab("Tissue") + 
    ylab("Allele Frequency") +
    theme_bw(22) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
    theme(strip.text.x = element_text(size = 12)) + 
    theme(panel.background = element_rect(fill = "grey"))



ggsave("../../results/figures/stacked.png", width = 10, height = 7, dpi = 300)

```



```{r}
write_csv(mean.haplotype.freqs, "mean-hapotypes-frequencies.csv")
```

```{r fig.width=15, fig.height=10}

tissue.frequencies = final.df %>% 
  mutate(Haplotype = case_when(
    Haplotype == "genome-1" ~ "genome 1",
    Haplotype == "genome-2" ~ "genome 2",
    Haplotype == "genome-1-1" ~ "genome 1_1",
    TRUE ~ Haplotype
  )) %>% 
  select(!Background) %>% 
  spread(Tissue, AF, fill = 0) %>% 
  column_to_rownames(var = "Haplotype") 


gheatmap(tree.plt, tissue.frequencies,
        colnames_angle=-30,
        colnames_position = "bottom",
        colnames_offset_y = 1.3,
        width=.6,
        offset =.0005,
        font.size = 8,
        low = "white",
        high = "red",
        hjust = 0,
        legend_title = "Allele Frequency") +
    theme(legend.position = "none")
  
ggsave("../../results/figures/sspe-phylogeny-with-heatmap.png", width = 23, height = 20, dpi = 300)

```

```{r fig.width=15, fig.height=10}

tree.df %>% 
  mutate(ADAR = case_when(
    node == genome.1.node ~ "X X X X X",
    node == genome.2.node ~ "",
    TRUE ~ ""
  )) %>% 
  as.treedata() %>% 
  ggtree(size = 3, aes(color = Background)) +
    geom_point2(aes(subset=(node==1)), size=12, color="#4d4d4d") + 
    geom_point2(aes(subset=(node==genome.1.node)), size=12, color="#307EC2") + 
    geom_point2(aes(subset=(node==genome.2.node)), size=12, color="#B63F3E") + 
    geom_point2(aes(subset=(node==23)), size=12, color="#199400") + 
    geom_tiplab(hjust = -.05, size = 8)  + 
    geom_text(aes(x=branch, label=ADAR), size=10, 
              vjust=-.3, color="firebrick") +
    scale_color_manual(values = c("#4d4d4d", "#307EC2", "#B63F3E")) + 
    theme(legend.position = "none") 


```

```{r Tree with ADAR, fig.width=15, fig.height=10}

ADAR.df = updated.df %>% 
  mutate(ADAR = case_when(
    REF == 'T' & ALT == 'C' ~ TRUE, 
    REF == 'A' & ALT == 'G' ~ TRUE,
    TRUE ~ FALSE
  )) %>% 
  select(Haplotype, SNP, ADAR) %>% 
  distinct() %>% 
  group_by(Haplotype, ADAR) %>% 
  count() %>% 
  filter(ADAR == TRUE) %>% 
  ungroup() %>% 
  select(!ADAR) %>% 
  filter(!Haplotype %in% c("both", "subclonal")) %>% 
  mutate(ADAR = strrep("X", n)) %>% 
  left_join(., select(tree.df, label, node), by = c("Haplotype" = "label")) %>% 
  mutate(node = case_when(
    Haplotype == "genome-2" ~ as.integer(genome.2.node),
    Haplotype == "genome-1" ~ as.integer(genome.1.node),
    Haplotype == "genome-1-1" ~ as.integer(23),
    Haplotype == "cluster 9" ~ as.integer(22),
    Haplotype == "cluster 2" ~ as.integer(29),
    TRUE ~ node
  )) %>% 
  select(ADAR, node)
  

tree.plt.adar = tree.df %>% 
  left_join(., ADAR.df, by = "node") %>%
  as.treedata() %>% 
  ggtree(size = 3, aes(color = Background)) +
    geom_point2(aes(subset=(node==1)), size=12, color="#4d4d4d") + 
    geom_point2(aes(subset=(node==genome.1.node)), size=12, color="#307EC2") + 
    geom_point2(aes(subset=(node==genome.2.node)), size=12, color="#B63F3E") + 
    geom_point2(aes(subset=(node==23)), size=12, color="#199400") + 
    geom_tiplab(hjust = -.05, size = 8)  + 
    geom_text(aes(x=branch, label=ADAR), size=6, 
              vjust=-.3, color="firebrick") +
    scale_color_manual(values = c("#4d4d4d", "#307EC2", "#B63F3E")) + 
    theme(legend.position = "none") 


gheatmap(tree.plt.adar, tissue.frequencies,
        colnames_angle=-30,
        colnames_position = "bottom",
        colnames_offset_y = 1.4,
        width=.6,
        offset =.0005,
        font.size = 8,
        low = "white",
        high = "red",
        hjust = 0,
        legend_title = "Allele Frequency") +
    theme(legend.position = "none")
  
ggsave("../../results/figures/sspe-phylogeny-with-heatmap-and-adar.png", width = 23, height = 20, dpi = 300)

```

