---
title: "Between Sample Bottleneck"
author: "Will Hannon"
date: "1/31/2022"
output: html_document
---

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```

```{r Required Packages, message=FALSE, warning=FALSE, echo=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed

packages = c("tidyverse", "UpSetR", "ComplexHeatmap", "cluster")
## Check that packages are installed, if not, install them
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}
## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```

```{r Inputs, include=FALSE}

## ==== File paths input data ==== ##

# Data from lofreq and varscan labeled with subclonal haplotypes
updated.data = "../../results/spatial/labeled_subhaplotypes.csv"

# Annotations 
annotations.filepath = "../../config/annotations.csv"

## ==== File paths output data ==== ##

# Path to save the figures for lab meeting 
output.path = "../../results/spatial"

```

```{r}
source("../../workflow/scripts/calculate_transmission_bottleneck.R")
```


```{r Process the variants, warning=F, message=F}

variant.df = read_csv(updated.data)

unique(variant.df$Tissue)[!unique(variant.df$Tissue) %in% "Temporal Lobe"]
```


```{r}

tissues = unique(variant.df$Tissue)

tv.df = data.frame()
results.df = data.frame()

for (tissue.one in tissues) {
  for (tissue.two in tissues) {
    
    if (tissue.one !=  tissue.two) {
      
      donor.df = variant.df %>% 
        filter(Tissue == tissue.one) %>% 
        select(SNP, AF, DP)
      
      contact.df = variant.df %>% 
        filter(Tissue == tissue.two) %>% 
        select(SNP, AF, DP)
      
      bottleneck.df = full_join(donor.df, contact.df,
                                by = 'SNP',
                                suffix = c(".donor", ".contact")) %>% 
        mutate_each(list( ~ replace(., which(is.na(.)), 0))) %>% 
        mutate(Donor = tissue.one,
               Contact = tissue.two)
      
      tv.df = rbind(tv.df, bottleneck.df)
      
      # -- Calculate the Bottleneck -- #  
      bottleneck.calc.df = bottleneck.df %>% 
        select(donor_freqs = AF.donor,
               recip_freqs = AF.contact,
               recip_total_reads = DP.contact) %>% 
        mutate(recip_var_reads = recip_total_reads*recip_freqs) %>% 
        mutate(recip_var_reads = round(recip_var_reads))
    
      
      print(paste("Calculating bottleneck for", unique(bottleneck.df$Donor), "vs.", unique(bottleneck.df$Contact)))

      
      results = estimate.transmission.bottleneck(bottleneck.calc.df,
                                                 var_calling_threshold = 0.02,
                                                 Nb_min = 1,
                                                 Nb_max = 2000,
                                                 confidence_interval = 0.95,
                                                 method = "exact")
    
      row = data.frame("donor" = tissue.one,
                       "recipient" = tissue.two,
                       "lower_ci" = results[[2]],
                       "upper_ci" = results[[3]],
                       "bottleneck" = results[[1]])
      
      results.df = rbind(results.df, row)
  
    }
  }
}

write_csv(results.df, "../../results/spatial/bottleneck.csv")

```

