---
title: "2. Determine Major Genotypes"
author:
    - "Will Hannon"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
  highlight: tango
  number_sections: no
  theme: default
  toc: yes
  toc_depth: 3
  toc_float:
    collapsed: no
    smooth_scroll: yes
---

As observed by the Cattaneo lab, there are clearly sets of mutations that move together in frequency across every spatial region of the brain. In particular, there are two sets of mutations who's average frequency appears to add up to 100% in every tissue. These two set of mutations appear to be ancestral to all virus variants in the brain.

The goal of this notebook is to formalize an approach to phase these sets of mutations by clustering variants that have strongly correlated frequencies across multiple tissue samples using k-medoids clustering on a correlation matrix of allele frequency 

In this notebook, I identify the 'genotypes' `Genome 1` and `Genome 2`. In addition to these two 'genotypes', I identify a third set of mutations that appears to arise on `Genome 1`, but is at very low frequency in the Frontal Cortex 2 sample **The output of this notebook is the list of variants annotated by whether they are `Genome 1`, `Genome 2`, `Genome-1-1`, or `subclonal.`**

```{r Setup, include = F}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```

```{r Required Packages, message = F, warning = F}

## ==== Install Required Packages ==== ##

## List of all packages needed
packages = c("tidyverse", "ComplexHeatmap", "cluster", "knitr", "ggrepel")

## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```

The inputs for this analysis are (1) the filtered and combined variants from `01-process-variant-calls.Rmd`, (2) annotations of the positions of the MeV open reading frames, and (3) the mutations in the SSPE consensus genome relative to the initial 'MeVChiTok' composite D-subtype reference. 

```{r Input, echo = T}

## ==== File paths input data ==== ##

if (exists("snakemake")) {
  
  # Filtered and unified variant calls
  variant.data = snakemake@params[['incsv']]
  
  # Pilot Sample
  pilot.data = snakemake@input[[1]]
  
  # Annotations for the standard Measles ORFs 
  annotations.filepath = snakemake@params[['annotations']]
  
} else {
  
  # Filtered and unified variant calls
  variant.data = "../../results/variants/filtered_variants.csv"
  
  # Pilot Sample
  pilot.data = "../../results/pilot/pilot_variants.csv"
  
  # Annotations for the standard Measles ORFs 
  annotations.filepath = "../../config/annotations.csv"
  
}

```

```{r Outputs, echo = T}

## ==== File paths output data ==== ##

if (exists("snakemake")) {
  
  # Variant calls annotated "genome-1", "genome-2", "genome-1-1", or "subclonal"
  output.path = snakemake@params[['outcsv']]
  
  # Path to save figures
  figure.dir = paste0(snakemake@params[['figures']], "/")

} else {
  
  # Variant calls annotated "genome-1", "genome-2", "genome-1-1", or "subclonal"
  output.path = "../../results/variants/genotyped_variants.csv"
  
  # Path to save figures
  figure.dir = "../../results/figures/"
  
}

# Make the figure directory if it doesn't exist
if (!file.exists(figure.dir)) {
  dir.create(figure.dir, recursive = TRUE)
  print("Directory created.")
} else {
  print("Directory already exists.")
}

```

## Clusters of Variants

This is what the underlying variant calling data we have looks like. 

```{r Read in Data, echo = F, warning = F, message = F}

selected.variants.df = read_csv(variant.data, show_col_types = FALSE)

head(selected.variants.df) %>% 
  kable(., caption = "Filtered Variants")

```
## Pilot Analysis 

For the pilot analysis we're just looking at the varscan variant calls from the single sample `SSPE 1` aligned to the `MeVChiTok` specimen, not the patient specific reference.

It's clear that there groups of mutations that are found at the same frequency across tissues. This was originally observed in this pilot tissue specimen that was chipped from the frozen brain. 

```{r Process Pilot Data, warning = F, message = F, echo = T}

# Minimum allele frequency to be included in analysis. 
min_AF = 0.02
# Minimum read depth to be included in analysis. 
min_DP = 200

sspe.pilot.df = read_csv(pilot.data, show_col_types = FALSE) %>% 
  filter(Caller == 'varscan') %>% 
  mutate(SNP = paste0(REF, POS, ALT)) %>% 
  mutate(Type = case_when(nchar(REF) > 1 ~ "Del",
                          nchar(ALT) > 1 ~ "Ins",
                          nchar(REF) == 1 ~ "SNP")) %>% 
  filter(Type == "SNP") %>% 
  filter(DP >= min_DP) %>% 
  filter(AF >= min_AF) %>% 
  select(!c("Caller", "Virus", "Host", "R1", "R2", "Type", "Sample")) 

```

It's clear that there groups of mutations that are found at the same frequency across tissues. This was originally observed in a pilot tissue specimen that was chipped from the frozen brain. This sample is called `SSPE 1`. 

```{r Annotations, message = F, warning = F, echo = F}

# Import genome annotations
annotations.df = read_csv(annotations.filepath, show_col_types = FALSE)

# Get the interval positions for these genes 
N = annotations.df %>% dplyr::filter(Locus == "N")
P.V.C = annotations.df %>% dplyr::filter(`Protein Name` == "phosphoprotein")
M = annotations.df %>% dplyr::filter(Locus == "M")
F. = annotations.df %>% dplyr::filter(Locus == "F")
H = annotations.df %>% dplyr::filter(Locus == "H")
L = annotations.df %>% dplyr::filter(Locus == "L")

# Add gene names to the variants
selected.variants.df = selected.variants.df %>% 
  mutate(Gene = case_when(POS < N$Start ~ "3'UTR",
                          POS > L$Stop ~ "5'UTR", 
                          TRUE ~ Gene_Name)) 

tissue_colors = c("#ef476f","#ffd166","#06d6a0","#118ab2","#073b4c","#DDA0DD")

# Tissue order - roughly by the location in the brain 
tissue_order = c(
  "SSPE 1", 
  "SSPE 2",
  "Frontal Cortex 2", 
  "Frontal Cortex 1", 
  "Frontal Cortex 3", 
  "Parietal Lobe",
  "Temporal Lobe", 
  "Occipital Lobe",
  "Hippocampus",
  "Internal Capsule", 
  "Midbrain", 
  "UBS",
  "Brain Stem",
  "Cerebellum",
  "Cerebellum Nucleus"
)

```

We can see these mutations clearly in the SSPE 1 pilot sample if we draw 'gates' of frequency around mutations on the plot. Below, I'm drawing 'gates' around 65-80% and 20-35% of the genome respectively.

```{r Format Pilot Data, echo = F, warning = F}

# Get the mutations by frequency gate
sspe.1.df = sspe.pilot.df %>% 
  # Select the pilot specimen
  filter(Tissue == "SSPE 1") %>% 
  # Bin variants by frequency
  mutate(Genotype = case_when(
    AF >= 0.9 ~ "CBA",
    AF >= 0.60 & AF <= 0.75 ~ "CG1", 
    AF >= 0.3 & AF <= 0.45 ~ "CG2",
    TRUE ~ "subclonal"
  ))

CBA.snps = sspe.1.df %>% filter(Genotype == "CBA") %>% pull(SNP) %>% unique()
CG1.snps = sspe.1.df %>% filter(Genotype == "CG1") %>% pull(SNP) %>% unique()
CG2.snps = sspe.1.df %>% filter(Genotype == "CG2") %>% pull(SNP) %>% unique()

# Format the data for plotting
sspe.df = sspe.pilot.df %>% 
  # Select the pilot specimen
  filter(Tissue %in% c("SSPE 1", "SSPE 2")) %>% 
  # Bin variants by frequency
  mutate(Genotype = case_when(
    SNP %in% CBA.snps ~ "CBA",
    SNP %in% CG1.snps ~ "CG1", 
    SNP %in% CG2.snps ~ "CG2",
    TRUE ~ "subclonal"
  )) %>% 
  # Point size should be small for mutation we don't know about
  mutate(Size = if_else(Genotype == "subclonal", "small", "normal")) %>% 
  # Point shape depends on whether the mutation is a transition or transversion
  mutate(Shape = case_when(
    REF == "A" & ALT == "G" ~ "A to G / U to C",
    REF == "T" & ALT == "C" ~ "A to G / U to C",
    REF == "G" & ALT == "A" ~ "Other Transitions",
    REF == "C" & ALT == "T" ~ "Other Transitions",
    TRUE ~ "Transversions"
  )) %>% 
  # We want to highlight mutations that are candidate 'Drivers' of spread
  mutate(Driver = case_when(
    SNP == "G3812A" ~ "M-W125*",
    SNP == "T6817A" ~ "F-L454M",
    SNP == "T3586C" ~ "M-F50S",
    # SNP == "T7293C" ~ "H-I8T",
    SNP == "C7036T" ~ "F-Q527*",
    TRUE ~ ""
  )) 
  
# Binned SNPS (i.e. CBA, CG1, and CG2)
binnned.snps = sspe.df %>% 
  filter(Genotype %in% c("CBA", "CG1", "CG2") & Driver == "")
# Un-binned SNPS (i.e. subclonal)
subclonal.snps = sspe.df %>% 
  filter(Genotype == 'subclonal' & Driver == "")
# Putative driver SNPs
driver.snps = sspe.df %>% 
  filter(Driver != "")

```

```{r Plot Pilot Data, warning=F, echo = F, fig.align = 'center', fig.width = 10, fig.height = 10}

# Plot the pilot data with sample SSPE 1 and SSPE 2
ggplot(data = binnned.snps, aes(x = POS, y = AF, color = Genotype, shape = Shape)) +
  geom_point(size = 2) + 
  geom_point(data = subclonal.snps, color = "#878484", size = 1) + 
  geom_point(data = driver.snps, color = "black", size = 3) + 
  geom_text_repel(data = driver.snps, aes(label = Driver), color = "black", point.padding = 1) + 
  facet_wrap(~Tissue, nrow = 2) +
  xlab("Position") + 
  ylab("Allele Frequency") +
  scale_color_manual(values = c("#d9cb34", "#307EC2", "#B63F3E")) +
  scale_shape_discrete(name="Mutation") +
  
  # == Annotations of genes == #
  
  # 3' Start
  annotate(geom = "text", x = (N$Start - 400), y =  -.05, label = "3'", size = 5) +
   
  # Nucleoprotein
  annotate("rect", xmin = N$Start-1, xmax = N$Stop, ymin = -.1, ymax = 0 ,
             alpha = .5, fill = "#d97634", col = "black", size = .1) +
  annotate(geom = "text", x = (N$Stop + N$Start)/2, y =  -.05, label = "N", size = 5) +
  annotate("rect", xmin = N$Start-1, xmax = N$Stop, ymin = 0, ymax = 1.1,
                 alpha = .05, fill = "#d97634", size = .1) +  
 
  # P/V/C
  annotate("rect", xmin = P.V.C$Start, xmax = P.V.C$Stop, ymin = -.1, ymax = 0 ,
             alpha = .5, fill = "#fcff3d", col = "black", size = .1) +
  annotate(geom = "text", x = (P.V.C$Stop + P.V.C$Start)/2, y =  -.05, label = "P/V/C", size = 5) +
  annotate("rect", xmin = P.V.C$Start, xmax = P.V.C$Stop, ymin = 0, ymax = 1.1,
             alpha = .05, fill = "#fcff3d", size = .1) +
 
  # Matrix protein
  annotate("rect", xmin = M$Start, xmax = M$Stop, ymin = -.1, ymax = 0 ,
             alpha = .5, fill = "#0db51b", col = "black", size = .1) +
  annotate(geom = "text", x = (M$Stop + M$Start)/2, y =  -.05, label = "M", size = 5)  +
  annotate("rect", xmin = M$Start, xmax = M$Stop, ymin = 0, ymax = 1.1,
             alpha = .05, fill = "#0db51b", size = .1) +
 
  # Fusion protein
  annotate("rect", xmin = F.$Start, xmax = F.$Stop, ymin = -.1, ymax = 0 ,
             alpha = .5, fill = "#6A3D9A", col = "black", size = .1) +
  annotate(geom = "text", x = (F.$Stop + F.$Start)/2, y =  -.05, label = "F", size = 5) +
  annotate("rect", xmin = F.$Start, xmax = F.$Stop, ymin = 0, ymax = 1.1,
             alpha = .05, fill = "#6A3D9A", size = .1) +

  # Hemagluttinin
  annotate("rect", xmin = H$Start, xmax = H$Stop, ymin = -.1, ymax = 0 ,
             alpha = .5, fill = "#0251d9", col = "black", size = .1) +
  annotate(geom = "text", x = (H$Stop + H$Start)/2, y =  -.05, label = "H", size = 5) +
  annotate("rect", xmin = H$Start, xmax = H$Stop, ymin = 0, ymax = 1.1,
             alpha = .05, fill = "#0251d9", size = .1) +

  # Large protein
  annotate("rect", xmin = L$Start, xmax = L$Stop, ymin = -.1, ymax = 0 ,
             alpha = .5, fill = "#bf1515", col = "black", size = .1) +
  annotate(geom = "text", x = (L$Stop + L$Start)/2, y = -.05, label = "L", size = 5) +
  annotate("rect", xmin = L$Start, xmax = L$Stop, ymin = 0, ymax = 1.1,
             alpha = .05, fill = "#bf1515", size = .1) +
  
  # 5' Start
  annotate(geom = "text", x = (L$Stop + 400), y =  -.05, label = "5'", size = 5) +
 
  
  # Themes
  theme_classic() +
  theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
  theme(legend.position="bottom", legend.box = "horizontal") +
  theme(legend.box.background =element_blank()) +
  theme(text=element_text(size=16)) + 
  theme(legend.text = element_text(size=12)) +
  theme(axis.text.x = element_text(size=16)) + 
  theme(axis.title.x = element_text(size=16)) + 
  scale_x_continuous(limits=c(-400, 16200), breaks=c(0, 5000, 10000, 15000)) +
  scale_y_continuous(breaks=c(0, 0.25, 0.5, 0.75, 1.0)) +
  theme(legend.position = "bottom") 

ggsave(paste(figure.dir, "figure-2.tiff", sep = ""), device = "tiff", width = 11, height = 10, dpi = 300)

```

What are the counts of mutations in each of these three bins (`CBA`: "Candidate Brain Ancestor", `CG1`: "Candidate Genome 1", and `CG2`: "Candidate Genome 2")?

```{r Mutations in Each Genotype, echo = F}

# What's the count of SNPs in each category?
sspe.df %>% 
  group_by(Genotype) %>% 
  count() %>%
  kable(., caption = "Candidate Genotype Variant Counts")

```

There are a total of {r nrow(sspe.1.df)} variable positions in the sample. 

These observations are what spurred the following approach to formalize a way to phase the mutations in these genotypes using all 15 tissues. 

## Phase Mutation Clusters

We know from experiments and preliminary observations (highlighted above) that there are **at least two sets of mutations present, but not fixed in every tissue sample**. These two sets of mutations are on the background of all viral sequences from this patient. Since these two sets of mutations are in every sample and their frequencies are anti-correlated, we should be able to identify these mutations by clustering them based on their co/anti-correlation. 

```{r Shared Mutations, warning = F, message = F, echo = F}

# All fifteen of the tissue samples 
tissues = selected.variants.df %>% 
  pull(Tissue) %>% 
  unique()

# Get the mutations in every single tissue
all.mutations = selected.variants.df %>% 
  count(SNP) %>%
  filter(n == length(tissues)) %>%
  pull(SNP)

print(paste("There are", length(all.mutations), "mutations found in every tissue sample"))

```

Although the reference seems to capture nearly every fixed mutation, there are four that aren't included in the reference. These are all in one continuous four nucleotide stretch in Nucleoprotein (N). I don't know why these don't get captured in the reference assembly, but it likely has to do with issues with the alignment of these reads. **I'll exclude these from the downstream analysis and add these back into the SSPE reference.** 

```{r Remaining Fixed Mutations, warning = F, message = F, echo = T}

# Remove the fixed mutations and get the long form of mutations in all samples. 
fixed.in.all = selected.variants.df %>% 
  filter(SNP %in% all.mutations) %>%
  filter(AF >= 0.95) %>% 
  group_by(SNP) %>% 
  count() %>% 
  filter(n == length(tissues)) %>%
  pull(SNP)

print(paste("There are", length(fixed.in.all), "fixed mutations in every tissue sample that aren't in the reference."))
print(paste("Fixed but not in reference:", fixed.in.all))
```

First, I'll make a correlation matrix between all of the SNP frequencies for SNPs that are found in every tissue and aren't fixed in every tissue (i.e. in the reference). 

```{r SNP Correlation, warning=F, message=F, echo=T, fig.align='center', fig.width=11, fig.height=9}

# Convert these mutations to long form where each column is a SNP and the row is it's frequency.
frequency.by.column = selected.variants.df %>% 
  filter(SNP %in% all.mutations) %>%
  filter(!SNP %in% fixed.in.all) %>% 
  select(AF, SNP, Tissue) %>% 
  pivot_wider(names_from = SNP, values_from = AF, values_fill = 0) %>% 
  select(!Tissue) 

# Calculate R between every pair of columns handling NAs as a correlation matrix
SNP.correlation = cor(frequency.by.column, use="pairwise.complete.obs")

Heatmap(SNP.correlation , 
        rect_gp = gpar(col = "white", lwd = 1),
        column_names_gp = grid::gpar(fontsize = 10),
        row_names_gp = grid::gpar(fontsize = 10),
        show_column_dend = FALSE, 
        show_row_dend = FALSE)
```

Most are strongly correlated (or anti-correlated), but some aren't strongly correlated with one another. Below is a histogram showing this more clearly. I'll filter out SNPs that aren't so well correlated with one another to make it easier to cluster the mutations on `Genome-1` and `Genome-2`. 

```{r Correlation Histogram, echo=F, fig.align='center', fig.width=6, fig.height=4}

# Get the R-squared from the correlation matrix
SNP.correlation.squared = SNP.correlation^2

# Plot the histogram
strong.corr.df = data.frame(SNP = names(rowMeans(SNP.correlation.squared)), Corr_Mean = rowMeans(SNP.correlation.squared))

strong.corr.df %>% 
  ggplot(aes(x = Corr_Mean)) +
    geom_histogram(bins = 40) +
    geom_vline(xintercept = 0.75, color = "red", linetype = 2) +
    xlab("R-squared") + 
    ylab("Count") +
    theme_bw(15)

ggsave(paste(figure.dir, "correlation.png", sep=""), width = 6, height = 4)

```

```{r Filter Correlation, echo=T, warning=F, message=F}

# Take only SNPs with greater than .75 based on the above histogram
filtered.SNPs = rowMeans(SNP.correlation.squared)[rowMeans(SNP.correlation.squared) >= 0.75]

# Remove the fixed mutations and get the long form
filtered.SNPs.df = selected.variants.df %>% 
  filter(!SNP %in% fixed.in.all) %>% 
  filter(SNP %in% names(filtered.SNPs)) %>% 
  select(AF, SNP, Tissue) %>% 
  pivot_wider(names_from = SNP, values_from = AF, values_fill = 0) %>% 
  select(!Tissue) 
  
# Calculate R between every pair of columns handling NAs 
filtered.SNP.correlation = cor(filtered.SNPs.df, use="pairwise.complete.obs")

# Convert to distrance (postive corr is close to 0)
filtered.SNP.dist = as.dist(1 - filtered.SNP.correlation)

# Create k-medoids clustering with 3 clusters
filtered.SNP.kmedoids = pam(filtered.SNP.dist, 3) 
kclusters = filtered.SNP.kmedoids$cluster

kclusters.df = data.frame(SNP = names(kclusters), cluster = kclusters)

# Assign the clusters to the original data frame
kmedoids.SNPs = selected.variants.df %>% 
  filter(!SNP %in% fixed.in.all) %>% 
  filter(SNP %in% all.mutations) %>% 
  left_join(., kclusters.df, by = "SNP") %>% 
  mutate(cluster = if_else(is.na(cluster), "no cluster", as.character(cluster)))

# Mean of each tissue
tissue.mean = kmedoids.SNPs %>% 
  group_by(Tissue, cluster) %>% 
  summarize(AF = mean(AF))

```

What's interesting, is that these mutations aren't best explained with 2 clusters (i.e. `Genome-1` and `Genome-2`) but rather with *3 clusters*! 

```{r Plot the clusters, echo=F, fig.align='center', fig.width=14, fig.height=10}

kmedoids.SNPs %>% 
  ggplot(aes(x = reorder(Tissue, -AF), y = AF)) +
    geom_line(aes(group = SNP)) +
    geom_line(data = tissue.mean, aes(x = Tissue, y = AF, group = 1, col = (cluster)), size = 1) +
    facet_wrap(~cluster) + 
    xlab("Tissue") + 
    ylab("Allele Frequency") +
    theme_bw(20) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    theme(legend.position = "none") 

ggsave(paste(figure.dir, "main-clusters.png", sep=""), width = 14, height = 10)
```

It's clear from the above plot that mutations grouped in clusters have very similar frequencies across tissues as expected. What's interesting is that clusters 2 and 3 are found at very similar frequencies in every tissue but the `Frontal Cortex 2` sample. It is likely that this represents an early divergence of one of the major genotypes.

We'll call cluster 3 `Genome-1`, cluster 2 `Genome-1-1`, and cluster 1 `Genome-2`. The 'no cluster' mutations are simply mutations found in every single tissue that don't neatly fit on `Genome-1`,`Genome-2`, or `Genome-1-1`. These will take some more work to resolve.

```{r Extract G1/G2, warning=F, message=F, echo=T}

g1.SNPs = kmedoids.SNPs %>% 
  filter(cluster == 3) %>% 
  pull(SNP)

sub.g1.SNPs = kmedoids.SNPs %>% 
  filter(cluster == 2) %>% 
  pull(SNP)

g2.SNPs = kmedoids.SNPs %>% 
  filter(cluster == 1) %>% 
  pull(SNP)

labeled.df = selected.variants.df %>% 
  filter(!SNP %in% fixed.in.all) %>% # excluding these four odd 'fixed mutations'. These are better explored in IGV
  mutate(Background = case_when(SNP %in% g1.SNPs ~ "genome-1", 
                                SNP %in% g2.SNPs ~ "genome-2",
                                SNP %in% sub.g1.SNPs ~ "genome-1",
                                TRUE ~ "subclonal")) %>% 
  mutate(Haplotype = case_when(SNP %in% g1.SNPs ~ "genome-1", 
                               SNP %in% g2.SNPs ~ "genome-2",
                               SNP %in% sub.g1.SNPs ~ "genome-1-1",
                               TRUE ~ "subclonal"))

```

I'll write out the newly annotated variant data frame for further analyses. 

```{r Write Out, warning=F, message=F, echo=F}

# Write these out to a file 
print(paste("Writing the results to", output.path))
write_csv(labeled.df, output.path)

```

## Exploring Genome-1 and Genome-2 

These two clusters of mutations (`Genome-1` and `Genome-2`) are ancestral to all other sequences in the brain. We can see that by summing their frequency in every compartment and seeing that they add up to 1. 

```{r G1 vs. G2 combined, fig.align='center', fig.width=10, fig.height=8, warning=F, message=F}

genotype_order = c("genome-2", "genome-1")
genome.colors = c("#307EC2", "#B63F3E")

genotype.plot.df = labeled.df %>% 
  filter(Haplotype %in% c("genome-1", "genome-2"))  %>% 
  select(Haplotype, Tissue, AF) %>% 
  group_by(Haplotype, Tissue) %>% 
  summarize(AF = mean(AF))

padding = genotype.plot.df %>% 
  group_by(Tissue) %>% 
  summarize(Padding = (1 - sum(AF))/2)

genotype.plot.df = genotype.plot.df %>% 
  left_join(., padding) %>% 
  mutate(AF = AF + Padding) %>% 
  select(!Padding)

genotype.plot.df$Haplotype = factor(genotype.plot.df$Haplotype, level = genotype_order)
  
genotype.plot.df %>% 
  ggplot(aes(x = factor(Tissue, levels = tissue_order), y = AF, fill = Haplotype, group = Haplotype)) + 
  geom_area() +
    xlab("Tissue") +
    ylab("Allele Frequency") +
    scale_fill_manual(values = genome.colors) +
    theme_bw(20) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
    theme(plot.title = element_text(hjust = 0.5))

```

Now, let's revisit the distribution of mutations in the genome colored by the mutations that have been identified as `Genome-1` or `Genome-2`. We can see how these mutations are distributed in the brain. 

```{r G1/G2 Distribution, warning=F, message=F, fig.align='center', fig.width=18, fig.height=12}

labeled.df %>% 
  filter(Background %in% c('genome-1', 'genome-2')) %>% 
  ggplot(aes(x = POS, y = factor(Tissue, levels = rev(tissue_order)))) +
    geom_point(aes(col = Background), shape = 108, size = 6) +
    xlab("Position") +
    # Annotations of genes #
    # Nucleoprotein
    annotate("rect", xmin = N$Start-1, xmax = N$Stop, ymin = 0, ymax = .5 ,
               alpha = .2, fill = "#d97634", col = "black", size = .1) +
    annotate(geom = "text", x = (N$Stop + N$Start)/2, y = .25, label = "N", family = "helvetica", size = 10) +
    annotate("rect", xmin = N$Start-1, xmax = N$Stop, ymin = .5, ymax = 16 ,
                   alpha = .1, fill = "#d97634", size = .1) +  
    # P/V/C
    annotate("rect", xmin = P.V.C$Start, xmax = P.V.C$Stop, ymin = 0, ymax = .5 ,
               alpha = .2, fill = "#fcff3d", col = "black", size = .1) +
    annotate(geom = "text", x = (P.V.C$Stop + P.V.C$Start)/2, y = .25, label = "P/V/C", family = "helvetica", size = 10) +
    annotate("rect", xmin = P.V.C$Start, xmax = P.V.C$Stop, ymin = 0, ymax = 16 ,
               alpha = .1, fill = "#fcff3d", size = .1) +
    # Matrix protein
    annotate("rect", xmin = M$Start, xmax = M$Stop, ymin = 0, ymax = .5 ,
               alpha = .2, fill = "#0db51b", col = "black", size = .1) +
    annotate(geom = "text", x = (M$Stop + M$Start)/2, y = .25, label = "M", family = "helvetica", size = 10)  +
    annotate("rect", xmin = M$Start, xmax = M$Stop, ymin = 0, ymax = 16 ,
               alpha = .1, fill = "#0db51b", size = .1) +
    # Fusion protein
    annotate("rect", xmin = F.$Start, xmax = F.$Stop, ymin = 0, ymax = .5 ,
               alpha = .2, fill = "#6A3D9A", col = "black", size = .1) +
    annotate(geom = "text", x = (F.$Stop + F.$Start)/2, y = .25, label = "F", family = "helvetica", size = 10) +
    annotate("rect", xmin = F.$Start, xmax = F.$Stop, ymin = 0, ymax = 16 ,
               alpha = .1, fill = "#6A3D9A", size = .1) +
    # Hemagluttinin
    annotate("rect", xmin = H$Start, xmax = H$Stop, ymin = 0, ymax = .5 ,
               alpha = .2, fill = "#0251d9", col = "black", size = .1) +
    annotate(geom = "text", x = (H$Stop + H$Start)/2, y = .25, label = "H", family = "helvetica", size = 10) +
    annotate("rect", xmin = H$Start, xmax = H$Stop, ymin = 0, ymax = 16,
               alpha = .1, fill = "#0251d9", size = .1) +
    # Large protein
    annotate("rect", xmin = L$Start, xmax = L$Stop, ymin = 0, ymax = .5 ,
               alpha = .2, fill = "#bf1515", col = "black", size = .1) +
    annotate(geom = "text", x = (L$Stop + L$Start)/2, y = .25, label = "L", family = "helvetica", size = 10) +
    annotate("rect", xmin = L$Start, xmax = L$Stop, ymin = 0, ymax = 16 ,
               alpha = .1, fill = "#bf1515", size = .1) +
    scale_color_manual(values = c("#307EC2", "#B63F3E", "grey")) +
    theme_classic() +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=24)) + 
    theme(axis.text.x = element_text(size=28)) + 
    theme(axis.title.x = element_text(size=30)) + 
    theme(plot.title = element_text(hjust = 0.5))  +
    theme(axis.title.y = element_blank()) +
    scale_x_continuous(limits=c(0,16000)) +
    theme(legend.position = "bottom") 

```

Interestingly, only `Genome-1` has mutations in the Matrix protein. This is kind of surprising given how hyper-mutated the Matrix protein is. 

Overall, there are more mutation in `Genome-1` than there are in `Genome-2`, but this is mostly driven by this large block of mutations in the Matrix. 

```{r Mutations per genotype, echo=F}

# All mutations 
all.mutations.included = labeled.df %>%
  filter(Background %in% c('genome-1', 'genome-2')) %>% 
  select(Background, SNP) %>% 
  distinct() %>% 
  group_by(Background) %>% 
  count()

# Excluding the mutations in Matrix
excluding.only.M = labeled.df %>%
  filter(Background %in% c('genome-1', 'genome-2')) %>% 
  filter(Gene_Name != "M") %>%
  select(Background, SNP) %>% 
  distinct() %>% 
  group_by(Background) %>% 
  count()

print(paste("There are",
            filter(all.mutations.included, Background == "genome-1")$n,
            "mutations in genome-1 and",
            filter(all.mutations.included, Background == "genome-2")$n,
            "mutations in genome-2"))

print(paste("There are",
            filter(excluding.only.M, Background == "genome-1")$n,
            "mutations in genome-1 and",
            filter(excluding.only.M, Background == "genome-2")$n,
            "mutations in genome-2 when you exclude Matrix"))

```

This is very important observation about the haplotypes since Matrix likely accumulates mutations at a different rate than the rest of the genome since there is no selective pressure to maintain the function of Matrix. 

## Which mutations don't cluster? 

Some mutations that are present in every tissue aren't correlating well with either the `Genome-1`, `Genome-2`, or `Genome-1-1` mutations. Let's take a look at these 'no cluster` mutations. 

```{r No Cluster Mutations, warning=F, message=F, fig.align='center', fig.width=12, fig.height=6}

tissue.mean = labeled.df %>% 
  filter(Haplotype %in% c("genome-1", "genome-2")) %>% 
  group_by(Tissue, Haplotype) %>% 
  summarize(AF.mean = mean(AF, na.rm = TRUE), 
            SD = sd(AF, na.rm = TRUE),
            N = n()) %>% 
  mutate(SE = SD / sqrt(N),
         Lower.CI = qt(1 - (0.05 / 2), N - 1) * SE,
         Upper.CI = qt(1 - (0.05 / 2), N - 1) * SE) %>% 
  rename("AF" = AF.mean)


kmedoids.SNPs %>% 
  filter(cluster == 'no cluster') %>% 
  mutate(title = paste(Gene_Name, AA_Change, sep = "-")) %>% 
  ggplot(aes(x = factor(Tissue, levels = tissue_order), y = AF, group = SNP )) + 
    geom_ribbon(data = tissue.mean, aes(x=factor(Tissue, levels = tissue_order),
                                        ymin=AF-SD,
                                        ymax=AF+SD,
                                        group = Haplotype,
                                        fill = Haplotype), alpha=0.2, colour = NA) +
    geom_line() + 
    scale_fill_manual(values=c("#424ef5", "#cf1919")) + 
    xlab("Tissue") + 
    ylab("Allele Frequency") +
    facet_wrap(~title) + 
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 

```

Three of these are particularly interesting, a stop codon towards the terminus of F, a mutation in the C-terminal of H, and a missense mutation in the N protein. 

```{r Interesting Mutations, warning=F, message=F, fig.align='center', fig.width=12, fig.height=4}

tissue.mean = labeled.df %>% 
  filter(Haplotype %in% c("genome-1", "genome-2")) %>% 
  group_by(Tissue, Haplotype) %>% 
  summarize(AF.mean = mean(AF, na.rm = TRUE), 
            SD = sd(AF, na.rm = TRUE),
            N = n()) %>% 
  mutate(SE = SD / sqrt(N),
         Lower.CI = qt(1 - (0.05 / 2), N - 1) * SE,
         Upper.CI = qt(1 - (0.05 / 2), N - 1) * SE) %>% 
  rename("AF" = AF.mean)

kmedoids.SNPs %>% 
  filter(AA_Change %in% c("Gln527*", "Ile8Thr", "Ala235Thr")) %>% 
  mutate(title = paste(Gene_Name, AA_Change, sep = "-")) %>% 
  ggplot(aes(x = factor(Tissue, levels = tissue_order), y = AF, group = SNP )) + 
    geom_ribbon(data = tissue.mean, aes(x=factor(Tissue, levels = tissue_order),
                                        ymin=AF-SD,
                                        ymax=AF+SD,
                                        group = Haplotype,
                                        fill = Haplotype), alpha=0.2, colour = NA) +
    geom_line() + 
    scale_fill_manual(values=c("#424ef5", "#cf1919")) + 
    xlab("Tissue") + 
    ylab("Allele Frequency") +
    facet_wrap(~title) + 
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) 

```
