---
title: "5. Connect Major Haplotypes"
author:
    - "Will Hannon"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
  highlight: tango
  number_sections: no
  theme: default
  toc: yes
  toc_depth: 3
  toc_float:
    collapsed: no
    smooth_scroll: yes
---

The goal of this notebook is to determine the relationship between the clusters of mutations identified in the previous notebooks (`cluster-subclonal-mutations.Rmd` and `genotype-subclonal-snps.Rmd`). With these relationships established, we can formulate a basic phylogeny of the viral haplotypes in the brain. 

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```

```{r Required Packages, message=FALSE, warning=FALSE, echo=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed
packages = c("tidyverse", "foreach", "emdbook", "RColorBrewer", "data.table", "ggnetwork", "network")

## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```

```{r Inputs, echo=T}

## ==== File paths input data ==== ##

# Data from lofreq and varscan labeled with subclonal haplotypes
updated.data = "../../results/variants/assigned_variants.csv"

# Data from the count of bridging reads
bridging.data = "../../results/bridging/bridging_reads.csv"

# Annotations 
annotations.filepath = "../../config/annotations.csv"

```

## Current Cluster Genotypes

First, using the previous notebook (`genotype-subclonal-snps.Rmd`), let's assign a putative background to each of the major clusters.

```{r Assign Genotypes, echo=T}

# Read in the variants with G1, G2, and G1-1 labeled
labeled.df = read_csv(updated.data, show_col_types = FALSE) %>% 
  mutate(Background = case_when(Haplotype == "cluster 1" ~ "genome-1",
                               Haplotype == "cluster 2" ~ "genome-1",
                               Haplotype == "cluster 3" ~ "genome-1",
                               Haplotype == "cluster 4" ~ "genome-1",
                               Haplotype == "cluster 5" ~ "genome-2",
                               Haplotype == "cluster 6" ~ "genome-2",
                               Haplotype == "cluster 7" ~ "genome-2",
                               Haplotype == "cluster 8" ~ "genome-1",
                               Haplotype == "cluster 9" ~ "genome-1",
                               Haplotype == "cluster 10" ~ "genome-1",
                               Haplotype == "cluster 11" ~ "genome-2",
                               Haplotype == "cluster 12" ~ "genome-1",
                               Haplotype == "cluster 13" ~ "genome-1",
                               Haplotype == "cluster 14" ~ "genome-1",
                               Haplotype == "cluster 15" ~ "genome-2",
                               Haplotype == "cluster 16" ~ "genome-2",
                              TRUE ~ Background)) 


```

Now, I'll plot all of the current haplotypes colored by whether they're likely `genome-1` or `genome-2`.
```{r All Polished Haplotypes, message=F, warning=F, fig.align='center', fig.width=19, fig.height=10, echo=T}

haplotypes.label = labeled.df %>% 
  select(SNP, Haplotype, Background) %>% 
  distinct()

# Expand the mutations to have a frequency for every tissue.
expanded.df = labeled.df %>% 
  select(SNP, Tissue, AF) %>% 
  pivot_wider(names_from = "Tissue", values_from = "AF", values_fill = 0) %>% 
  pivot_longer(cols = !SNP, names_to = "Tissue", values_to = "AF") %>% 
  left_join(., select(labeled.df, c("SNP", "Tissue", "DP")), by = c("SNP", "Tissue")) %>% 
  mutate(DP = if_else(is.na(DP), 0, DP)) %>% 
  left_join(., haplotypes.label, by = "SNP") 

# Get the mean frequency of the major genomes 
tissue.mean = labeled.df %>% 
  filter(Haplotype %in% c("genome-1", "genome-2")) %>% 
  group_by(Tissue, Haplotype) %>% 
  summarize(AF.mean = mean(AF, na.rm = TRUE), 
            SD = sd(AF, na.rm = TRUE),
            N = n()) %>% 
  mutate(SE = SD / sqrt(N),
         Lower.CI = qt(1 - (0.05 / 2), N - 1) * SE,
         Upper.CI = qt(1 - (0.05 / 2), N - 1) * SE) %>% 
  rename("AF" = AF.mean)

#  Mean frequency in each haplotype
haplotype.mean = expanded.df %>% 
  group_by(Tissue, Haplotype) %>% 
  summarize(AF = mean(AF))

# Tissue order - roughly by the location in the brain 
tissue_order = c(
  "SSPE 1", 
  "SSPE 2",
  "Frontal Cortex 2", 
  "Frontal Cortex 1", 
  "Frontal Cortex 3", 
  "Parietal Lobe",
  "Temporal Lobe", 
  "Occipital Lobe",
  "Hippocampus",
  "Internal Capsule", 
  "Midbrain", 
  "UBS",
  "Brain Stem",
  "Cerebellum",
  "Cerebellum Nucleus"
)

haplotype.order = c("cluster 1",
                    "cluster 2",
                    "cluster 3",
                    "cluster 4",
                    "cluster 5",
                    "cluster 6",
                    "cluster 7",
                    "cluster 8",
                    "cluster 9",
                    "cluster 10",
                    "cluster 11",
                    "cluster 12",
                    "cluster 13",
                    "cluster 14", 
                    "cluster 15",
                    "cluster 16",
                    # "cluster 17",
                    # "cluster 18", 
                    # "cluster 19", 
                    "genome-1-1")

expanded.df %>% 
    filter(!Haplotype %in% c("fixed", "both", "subclonal", "genome-1", "genome-2")) %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue_order), y = AF, col = (Background))) +
      geom_line(aes(group = SNP)) +
      facet_wrap(~factor(Haplotype, levels = haplotype.order)) + 
      scale_color_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))


```

## Bridging Reads

Let's see how much information we can glean about cluster relationships from the reads that overlap SNPs shared by two clusters.

```{r Upload Bridging Reads, echo=T}
# Bridging reads
bridging.df = read_csv(bridging.data, show_col_types = FALSE)
head(bridging.df)
```

The columns are as follows: `snp_1` is the position in the genome of one SNP, `snp_2` is the position of the second SNP, `00` are reads that overlap both positions but have neither SNP, `10` are reads that have only `snp_1`, `01` are reads that have only `snp_2`, and `11` are reads with both SNPs. 

Below, I'll calculate the probability that any two SNPs are either linked or forbidden over a log-spaced range of `taus`. 

```{r Calculate Linked and Forbidden SNPs, echo=T}

taus = lseq(0.005, .5, length.out = 50) # Log spaced values for tau
min_obsv = 100 # Minimum number of bridging reads (of any type)
min_snp_obsv = 25 # Minimum number of bridging reads with SNPs
min.sig = 0.05 # Minimum significance before multiple testing correction
  
# Assign to genome for each value of tau
SNPassignmentsByTau = foreach(tau = taus, .combine = "rbind")%do%{

  bridging.df %>% 
    mutate(total = `00` + `10` + `01` + `11`) %>% 
    filter(total >= min_obsv) %>% 
    filter(((`01` + `11`) >= min_snp_obsv) & ((`10` + `11`) >= min_snp_obsv)) %>% 
    # Multiple testing done by tissue, not polymorphic position
    group_by(snp_1, snp_2) %>% 
    # Calculate the probability of being linked
    mutate(linked = pbinom(`11`, total, tau, lower.tail = F), 
           linked_sig = linked < min.sig/n()) %>% 
      # Calculate the probability of being forbidden
    mutate(forbidden = pbinom(`11`, total, tau, lower.tail = T), 
           forbidden_sig = forbidden < min.sig/n()) %>% 
    mutate(linked_sig = ifelse(total == 0, NA, linked_sig),
           forbidden_sig = ifelse(total == 0, NA, forbidden_sig)) %>% 
    ungroup() %>% 
    mutate(tau = tau)

}

SNPassignmentsByTau %>% 
  select(!c("total", "linked", "forbidden")) %>% 
  head()

```

Then, I'll make a function that takes in the name of a cluster, the data frame of SNPs and their frequencies, the data frame of linked/forbidden SNPs as determined by bridging reads, and a minimum frequency at which both of the SNPs must be present in a given tissue's bridging reads. 

```{r Function to Summarise Bridging Reads, echo=T}

rank_bridging_reads = function(cluster.ID, snp.df, tau.df, min.freq = 0.1) {
    
  tau.df.by.tissue = tau.df %>% 
    mutate(snp_1_freq = (`10` + `11`) / total, 
           snp_2_freq = (`01` + `11`) / total) %>% 
    filter((snp_1_freq >= min.freq) & (snp_2_freq >= min.freq)) %>% 
    select(snp_1, snp_2, linked_sig, forbidden_sig, tau, Tissue) %>% 
    group_by(snp_1, snp_2, tau) %>% 
    mutate(linked = sum(linked_sig),
           forbidden = sum(forbidden_sig)) %>% 
    select(!c("Tissue", "linked_sig", "forbidden_sig")) %>% 
    distinct() %>% 
    ungroup()
  
  cluster.pos = snp.df %>% 
    filter(Haplotype == cluster.ID) %>% 
    pull(POS) %>% 
    unique()
  
  snp.1.df = tau.df.by.tissue %>% 
    filter(snp_1 %in% cluster.pos) %>% 
    dplyr::rename("cluster_snp" = snp_1, "query_snp" = snp_2)
  
  snp.2.df = tau.df.by.tissue %>% 
    filter(snp_2 %in% cluster.pos) %>% 
    dplyr::rename("cluster_snp" = snp_2, "query_snp" = snp_1)
  
  query.df = rbind(snp.1.df, snp.2.df) %>% 
    left_join(., distinct(select(snp.df, POS, Haplotype)), by = c("query_snp" = "POS")) %>% 
    filter(!Haplotype %in% c("subclonal", "both")) %>% 
    arrange(Haplotype, cluster_snp, tau) %>% 
    filter((linked > 0) | (forbidden > 0))
  
  return(query.df)
  
}

```

For all of the clusters plotted above, I'll print every known haplotype (including SNPs in G1 and G2) that is possibly linked to that cluster. As long as a SNP with a known haplotype is linked to the cluster's SNP in more than one tissue where both SNPs are present in greater than 10% of bridging reads, it will get printed below.

```{r Print Linked Matches, echo=T}

clusters = labeled.df %>% 
  filter(!Haplotype %in% c("genome-1", "genome-2", "both", "fixed", "subclonal")) %>% 
  pull(Haplotype) %>% 
  unique()

for (cluster.id in clusters) {
  
  linked.matches = rank_bridging_reads(cluster.id, labeled.df, SNPassignmentsByTau, min.freq = 0.1) %>% 
    filter(linked >= 1) %>% 
    pull(Haplotype) %>% 
    unique()
  
  print(paste(cluster.id, "is linked to:"))
  print(paste(linked.matches, collapse=", "))
  print("----")
  
}

```

Most of these conform with our expectations. The exceptions are cluster 10, 11, 13, 15, and 6. Let's investigate each of these independently. 

There are two reasons that SNPs might violate our assumptions about how clusters are linked together. First, they could be spuriously linked based on frequency. This could easily be the case for cluster 15, 10, and 13. Cluster 15 is a very low frequency cluster, there is good amount of internal variation, and there are only two SNPs. 

```{r Cluster 15, fig.align='center', fig.width=8, fig.height=5, echo=T}

expanded.df %>% 
    filter(Haplotype %in% c("cluster 15")) %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue_order), y = AF,)) +
      geom_line(aes(group = SNP)) +
      facet_wrap(~factor(Haplotype, levels = haplotype.order)) + 
      scale_color_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))

```

This probably shouldn't be a cluster, so let's remove it. 

```{r Remove cluster 15, echo=T}

labeled.df = labeled.df %>% 
  mutate(Background = case_when(Haplotype == "cluster 15" ~ "unknown",
                                TRUE ~ Background)) %>% 
  mutate(Haplotype = case_when(Haplotype == "cluster 15" ~ "subclonal",
                                TRUE ~ Haplotype))

```

Cluster 10 and 13 likely suffer from similar issues. In the case of cluster 10, we know that there are a set of variants on the background of genome-1 in the frontal cortex. However, since these SNPs are only present on the same haplotype in a single tissue, we can't as clearly link them by frequency.

```{r Cluster 10, fig.align='center', fig.width=8, fig.height=5, echo=T}

expanded.df %>% 
    filter(Haplotype %in% c("cluster 10")) %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue_order), y = AF,)) +
      geom_line(aes(group = SNP)) +
      facet_wrap(~factor(Haplotype, levels = haplotype.order)) + 
      scale_color_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))

```

This cluster of SNPs likely also suffers from the second possible issue that can lead to bridging reads disagreeing with our assumptions, and that's having lots of SNPs in Matrix. The mutations in Matrix are frequent and have the same spectrum due to ADAR. Therefore, it's conceivable that the same mutation could occur multiple times on different halpotypes. Deconvolving this is beyond the scope of our haplotyping approach. 

```{r Percentage of Matrix SNPs}

labeled.df %>% 
  filter (Haplotype == "cluster 10") %>% 
  select(SNP, Gene) %>% 
  distinct() %>% 
  group_by(Gene) %>% 
  count() %>% 
  arrange(-n)


```

This is easily visualized by plotting only the non-Matrix mutations. The correlation is a lot more plausible and the fact that this cluster is excluded from genome-1-1 is more clear. 

```{r Cluster 10 Non-matrix, fig.align='center', fig.width=8, fig.height=5, echo=T}

expanded.df %>% 
    filter(Haplotype %in% c("cluster 10")) %>%
    left_join(., distinct(select(labeled.df, SNP, Gene))) %>% 
    filter(Gene != "M") %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue_order), y = AF,)) +
      geom_line(aes(group = SNP)) +
      facet_wrap(~factor(Haplotype, levels = haplotype.order)) + 
      scale_color_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))


```

One solution to this issue, short of removing all Matrix mutations, is to remove the matrix mutations that are *also* linked to genome-1-1 or genome-2. This makes the correlation much better. 

```{r Exclude Reccurent SNPs from Cluster 10, fig.align='center', fig.width=8, fig.height=5, echo=T}

exclude.from.cluster.10 = rank_bridging_reads("cluster 10", labeled.df, SNPassignmentsByTau, min.freq = 0.1) %>% 
  filter(linked >= 1) %>% 
  filter(Haplotype %in% c("genome-1-1", "genome-2")) %>% 
  select(cluster_snp) %>% 
  pull() %>% 
  unique()

expanded.df %>% 
    filter(Haplotype %in% c("cluster 10")) %>%
    left_join(., distinct(select(labeled.df, SNP, Gene, POS))) %>% 
    filter(!POS %in% exclude.from.cluster.10) %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue_order), y = AF,)) +
      geom_line(aes(group = SNP)) +
      facet_wrap(~factor(Haplotype, levels = haplotype.order)) + 
      scale_color_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))

print(paste("Excluding", length(exclude.from.cluster.10), "SNPs from cluster 10."))

labeled.df = labeled.df %>% 
  mutate(Background = case_when(POS %in% exclude.from.cluster.10 ~ "unknown",
                                TRUE ~ Background)) %>% 
  mutate(Haplotype = case_when(POS %in% exclude.from.cluster.10 ~ "subclonal",
                                TRUE ~ Haplotype))
```

Cluster 13 is interesting, because from frequency alone, we know it's not possible that this exists as a single haplotype. This is because we believe that it's linked to genome-1 and therefore descended from cluster 10, yet, it's found at a higher frequency than cluster 10 in the Midbrain. 

```{r Cluster 13, fig.align='center', fig.width=8, fig.height=5, echo=T}

expanded.df %>% 
    filter(Haplotype %in% c("cluster 13")) %>%
    left_join(., distinct(select(labeled.df, SNP, Gene, POS))) %>% 
    filter(!POS %in% exclude.from.cluster.10) %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue_order), y = AF,)) +
      geom_line(aes(group = SNP)) +
      facet_wrap(~factor(Haplotype, levels = haplotype.order)) + 
      scale_color_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))

```

This cluster suffers from several probable issues. For one, it's pretty low frequency so error is proportionately large. Second, it's only high frequency in two tissues - the frontal cortex and mid brain - so the correlation could be spurious Finally, it's only Matrix mutations, so the likelihood that they're recurrent is high. 

```{r Percentage of Matrix SNPs C13, echo=F}

labeled.df %>% 
  filter (Haplotype == "cluster 13") %>% 
  select(SNP, Gene) %>% 
  distinct() %>% 
  group_by(Gene) %>% 
  count() %>% 
  arrange(-n)

```

This is likely two or more haplotypes, and unsurprisingly, all found are linked to both genome-1-1 and cluster 10. This suggests that these are recurrent SNPs and shouldn't be considered a haplotype. 

```{r Remove Cluster 13, echo=T}

rank_bridging_reads("cluster 13", labeled.df, SNPassignmentsByTau, min.freq = 0.1) %>% 
  filter(linked >= 1) %>% 
  filter(Haplotype %in% c("genome-1-1")) %>% 
  select(cluster_snp) %>% 
  pull() %>% 
  unique()

labeled.df = labeled.df %>% 
  mutate(Background = case_when(Haplotype == "cluster 13" ~ "unknown",
                                TRUE ~ Background)) %>% 
  mutate(Haplotype = case_when(Haplotype == "cluster 13" ~ "subclonal",
                                TRUE ~ Haplotype))

```

Now let's see about cluster 11. These are SNPs present in the lower brain where the read depth is pretty low, and the correlation is somewhat loose. 

```{r Cluster 11, fig.align='center', fig.width=8, fig.height=5, echo=T}

expanded.df %>% 
    filter(Haplotype %in% c("cluster 11")) %>%
    left_join(., distinct(select(labeled.df, SNP, Gene, POS))) %>% 
    filter(!POS %in% exclude.from.cluster.10) %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue_order), y = AF,)) +
      geom_line(aes(group = SNP)) +
      facet_wrap(~factor(Haplotype, levels = haplotype.order)) + 
      scale_color_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))

```

This cluster is half matrix mutations and half non-matrix mutations
```{r Percentage of Matrix SNPs C11, echo=F}

labeled.df %>% 
  filter (Haplotype == "cluster 11") %>% 
  select(SNP, Gene) %>%
  distinct() %>% 
  group_by(Gene) %>% 
  count() %>% 
  arrange(-n)

```
Below, I'll clean up cluster 11 by removing SNPs that are linked to genome-1 (since cluster 11 is a genome-2 cluster) and removing Matrix mutations unless they're linked to other cluster 11 mutations or genome-2. Finally, there is a very low depth F mutation that visibly looks like an outlier. I'll remove this as well. 
```{r Clean Up Cluster 11, fig.align='center', fig.width=8, fig.height=5, echo=T}

linked.to.cluster11 = rank_bridging_reads("cluster 11", labeled.df, SNPassignmentsByTau, min.freq = 0.1) %>% 
  filter(linked >= 1) %>% 
  filter(Haplotype %in% c("cluster 11", "genome-2")) %>% 
  select(cluster_snp) %>% 
  pull() %>% 
  unique()

unlinked.from.cluster.11 = rank_bridging_reads("cluster 11", labeled.df, SNPassignmentsByTau, min.freq = 0.1) %>% 
  filter(linked >= 1) %>% 
  filter(Haplotype %in% c("genome-1-1", "genome-1")) %>% 
  select(cluster_snp) %>% 
  pull() %>% 
  unique()

keep.in.cluster.11 = linked.to.cluster11[which(!linked.to.cluster11 %in% unlinked.from.cluster.11)]

non.matrix.cluster.11 = labeled.df %>% 
  filter(Haplotype == "cluster 11") %>% 
  filter(Gene != "M") %>% 
  pull(POS) %>% 
  unique()

keep.in.cluster.11 = c(keep.in.cluster.11, non.matrix.cluster.11)

low.depth.outlier = 4923

keep.in.cluster.11 = keep.in.cluster.11[which(!keep.in.cluster.11 %in% low.depth.outlier)]

cluster.11.snps = labeled.df %>% 
  filter(Haplotype %in% c("cluster 11")) %>%
  pull(POS) %>% 
  unique()  

exclude.from.cluster.11 = cluster.11.snps[which(!cluster.11.snps %in% keep.in.cluster.11)]


expanded.df %>% 
    filter(Haplotype %in% c("cluster 11")) %>%
    left_join(., distinct(select(labeled.df, SNP, Gene, POS))) %>% 
    filter((!POS %in% exclude.from.cluster.11)) %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue_order), y = AF,)) +
      geom_line(aes(group = SNP)) +
      facet_wrap(~factor(Haplotype, levels = haplotype.order)) + 
      scale_color_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))


labeled.df = labeled.df %>% 
  mutate(Background = case_when((POS %in% exclude.from.cluster.11) ~ "unknown",
                                TRUE ~ Background)) %>% 
  mutate(Haplotype = case_when((POS %in% exclude.from.cluster.11) ~ "subclonal",
                                TRUE ~ Haplotype))

```

Finally, unrelated to the bridging reads, Cluster 2 and Cluster 9 are the same cluster, but Cerebellum has very low depth. The SNPs in cluster 2 are present in Cerebellum at a similar frequency to cluster 9, but they fail the depth filters. Below, I'll amend that by setting the frequency to the mean frequency of cluster 9 and merging them.

```{r Merge Cluster 2 into 9, echo=T}

c9.mean.cerebellum.freq = labeled.df %>% 
  filter(Haplotype == "cluster 9" & Tissue == "Cerebellum") %>% 
  pull(AF) %>% 
  mean()

cluster.2.cerebellum.dummy.alleles = labeled.df %>% 
  filter(Haplotype == "cluster 2") %>% 
  filter(Tissue == "Cerebellum Nucleus") %>% 
  mutate(AF = c9.mean.cerebellum.freq, 
         Tissue = "Cerebellum", 
         Accession = "Cerebellum",
         DP = 100)


labeled.df = rbind(labeled.df, cluster.2.cerebellum.dummy.alleles) %>% 
  mutate(Haplotype = case_when(
    Haplotype == "cluster 2" ~ "cluster 9",
    TRUE ~ Haplotype)) 


```

Okay, so what do the clusters look like now after cleaning them up using the bridging reads? Before I proceed, I'll rename clusters according to the Cattaneo lab's name scheme.

```{r Cattaneo Nameing Scheme, echo=T}


labeled.df = labeled.df %>% 
  mutate(Haplotype_Name = case_when(
    Haplotype == "cluster 1" ~ "cluster 1",
    Haplotype == "cluster 3" ~ "cluster 2",
    Haplotype == "cluster 4" ~ "cluster 3",
    Haplotype == "cluster 8" ~ "cluster 4",
    Haplotype == "cluster 9" ~ "cluster 5",
    Haplotype == "cluster 10" ~ "cluster 6",
    Haplotype == "cluster 12" ~ "cluster 7",
    Haplotype == "cluster 14" ~ "cluster 8",
    Haplotype == "cluster 5" ~ "cluster 9",
    Haplotype == "cluster 6" ~ "cluster 10",
    Haplotype == "cluster 7" ~ "cluster 11",
    Haplotype == "cluster 11" ~ "cluster 12",
    Haplotype == "cluster 16" ~ "cluster 13",
    Haplotype == "cluster 14" ~ "cluster 14",
    Haplotype == "genome-1-1" ~ "genome 1",
    Haplotype == "genome-1" ~ "genome 01",
    Haplotype == "genome-2" ~ "genome 2",
    TRUE ~ Haplotype)) 


```


```{r Exapand the SNPs, message=F, warning=F, fig.align='center', fig.width=19, fig.height=10, echo=T}

haplotypes.label = labeled.df %>% 
  select(SNP, Haplotype, Haplotype_Name, Background) %>% 
  distinct()

# Expand the mutations to have a frequency for every tissue.
expanded.df = labeled.df %>% 
  select(SNP, Tissue, AF) %>% 
  pivot_wider(names_from = "Tissue", values_from = "AF", values_fill = 0) %>% 
  pivot_longer(cols = !SNP, names_to = "Tissue", values_to = "AF") %>% 
  left_join(., select(labeled.df, c("SNP", "Tissue", "DP")), by = c("SNP", "Tissue")) %>% 
  mutate(DP = if_else(is.na(DP), 0, DP)) %>% 
  left_join(., haplotypes.label, by = "SNP") 

# Tissue order - roughly by the location in the brain 
tissue_order = c(
  "SSPE 1", 
  "SSPE 2",
  "Frontal Cortex 2", 
  "Frontal Cortex 1", 
  "Frontal Cortex 3", 
  "Parietal Lobe",
  "Temporal Lobe", 
  "Occipital Lobe",
  "Hippocampus",
  "Internal Capsule", 
  "Midbrain", 
  "UBS",
  "Brain Stem",
  "Cerebellum",
  "Cerebellum Nucleus"
)

haplotype.order = c("cluster 1",
                    "cluster 2",
                    "cluster 3",
                    "cluster 4",
                    "cluster 5",
                    "cluster 6",
                    "cluster 7",
                    "cluster 8",
                    "cluster 9",
                    "cluster 10",
                    "cluster 11",
                    "cluster 12",
                    "cluster 13",
                    "cluster 14", 
                    "genome 1")

expanded.df %>% 
    filter(!Haplotype %in% c("fixed", "both", "subclonal", "genome-1", "genome-2")) %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue_order), y = AF, col = (Background))) +
      geom_line(aes(group = SNP)) +
      facet_wrap(Background~factor(Haplotype_Name, levels = haplotype.order)) + 
      scale_color_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))
```

## Frequency by compartment

These look a lot clearer than before thanks to the bridging reads. Now, I'm going to look at the frequency in each compartment and use the 'pigeon-hole' principle to see if there are violations in the sum of frequencies that point towards clusters being related to one another. 

```{r Color Palletes, echo=F}

# Make a color scheme for the haplotypes
## Genome-1
genome.1.haplotypes = c( "cluster 1", "cluster 2", "cluster 3", "cluster 4", "cluster 5", "cluster 7", "cluster 8", "genome 1", "genome 01")
genome.1.colors = brewer.pal(length(genome.1.haplotypes), "Blues")
names(genome.1.colors) = genome.1.haplotypes

## Genome-2
genome.2.haplotypes = c("cluster 9", "cluster 10", "cluster 11", "cluster 12", "cluster 13", "genome 2")
genome.2.colors = brewer.pal(length(genome.2.haplotypes), "YlOrRd")
names(genome.2.colors) = genome.2.haplotypes

## Cluster 6
cluster.6.haplotype = c("cluster 6")
cluster.6.color = brewer.pal(8, "Purples")[8]
names(cluster.6.color) = cluster.6.haplotype

## Combine
haplotype.colors = c(genome.1.colors, genome.2.colors, cluster.6.color)
# names(haplotype.colors) = gsub(" ", "_", names(haplotype.colors ))
haplotype.colors = c(haplotype.colors , "Anc" = "black")
# cluster 1 is too light
haplotype.colors['cluster 1'] = "#bdbebf"

## Set the Compartment Colors 
# Frontal Cortex
FCs = c("Frontal Cortex 1", "Frontal Cortex 2", "Frontal Cortex 3")
FCs.colors = brewer.pal(length(FCs), "Greens")
names(FCs.colors) = FCs
# Internal Compartments
Internal = c("Cerebellum Nucleus", "Cerebellum", "Midbrain", "Internal Capsule", "Brain Stem", "UBS", "Hippocampus")
Internal.colors = brewer.pal(length(Internal), "YlOrRd")
names(Internal.colors) = Internal
# Other Lobes
Other =  c("Occipital Lobe", "Temporal Lobe", "Parietal Lobe")
Other.colors = brewer.pal(length(Other), "Blues")
names(Other.colors) = Other
# Combine
Compartment.cols = c(FCs.colors, Internal.colors, Other.colors)
names(Compartment.cols) = gsub(" ", "_", names(Compartment.cols))

```

```{r Mean Haplotype Frequency}

haplotype.mean = expanded.df %>% 
  group_by(Haplotype_Name, Tissue, Background) %>% 
  summarize(AF = mean(AF))

```

```{r Genome 2 Clusters, warning=F, fig.align='center', fig.width=8, fig.height=5, echo=T}

genome.2.clusters = haplotype.mean %>% 
  filter(Background == "genome-2" & (!Haplotype_Name %in% c("genome 2"))) %>% 
  pull(Haplotype_Name) %>% 
  unique()

genome.2.mean = haplotype.mean %>% 
  filter(Haplotype_Name == "genome 2")

haplotype.mean %>% 
    filter(Haplotype_Name %in% genome.2.clusters) %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue_order), y = AF, fill = (factor(Haplotype_Name, levels = haplotype.order)), group = Haplotype_Name)) +
      geom_area() +
      geom_line(data = genome.2.mean, aes(x = factor(Tissue, levels = tissue_order), col = (Haplotype_Name), y = AF, group = Haplotype_Name), linewidth = 2) +
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Cluster") +
      scale_fill_manual(values = haplotype.colors) +
      scale_color_manual(values = haplotype.colors) +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))

```

Genome 2 doesn't have any obvious violations! What about Genome 1? 

```{r Genome 1 Clusters, warning=F, fig.align='center', fig.width=8, fig.height=5, echo=T}

genome.1.clusters = haplotype.mean %>% 
  filter(Background == "genome-1" & (!Haplotype_Name %in% c("genome 1", "genome 01"))) %>% 
  pull(Haplotype_Name) %>% 
  unique()

genome.1.mean = haplotype.mean %>% 
  filter(Haplotype_Name == "genome 01")

haplotype.mean %>% 
    filter(Haplotype_Name %in% genome.1.clusters) %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue_order), y = AF, fill = (factor(Haplotype_Name, levels = haplotype.order)), group = Haplotype_Name)) +
      geom_area() +
      geom_line(data = genome.1.mean, aes(x = factor(Tissue, levels = tissue_order), col = (Haplotype_Name), y = AF, group = Haplotype_Name), linewidth = 2) +
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Cluster") +
      scale_fill_manual(values = haplotype.colors) +
      scale_color_manual(values = haplotype.colors) +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))

```

The only obvious violation is caused by cluster 8 and cluster 4. These are likely related to one another in some way. Perhaps cluster 4 is on the background of cluster 8? 

## Prepare reads for Spruce

Finally, to build the phylogeny, I'll use the algorithm SPRUCE as implemented by MACHINA. First, I need to prepare the data for MACHINA by using the mean allele frequency +/- the binomial variance, times a down scaling factor.

```{r Format Data for MACHINA, echo=T}

# Add in the depth data
merged.depth = "../../results/coverage/merged.depth"

merged.depth.df = fread(merged.depth) %>% 
  select(!V1)

expanded.df = expanded.df %>% 
  select(!DP) %>% 
  left_join(., distinct(select(labeled.df, POS, SNP, Accession)), by = c("SNP")) %>% 
  left_join(., merged.depth.df, by = c("POS", "Accession"))

# Make tissue index
Tissue_index = expanded.df %>% 
  select(Tissue) %>% 
  distinct() %>% 
  mutate(tind = 1:n() - 1)

# Make cluster index 
Cluster_index = expanded.df %>% 
  select(Haplotype_Name) %>% 
  distinct() %>% 
  filter(!Haplotype_Name %in% c("both", "subclonal")) %>% 
  mutate(cind = 1:n() - 1) %>%
  add_row(Haplotype_Name = "Anc", cind = 16) 

# Calculate MAF, var, lb, and ub for MACHINA
var.margin = 0.18

snpinf = expanded.df %>%
  group_by(Haplotype_Name, Tissue) %>% 
  filter(!Haplotype_Name %in% c("both", "subclonal")) %>% 
  # use this formulation for cluster allele frequency, because it puts less weight on low read depth positions
  summarize(MAF = sum(AF * DP)/sum(DP),
            var = MAF * (1 - MAF)) %>% 
  # MACHINA wants an upper and lower bound on what the allele frequencies are
  mutate(lb = pmax(0, MAF - var.margin * var), ub = pmin(1, MAF + var.margin * var))
 
# Add the ancestor back in or it won't connect g1 and g2
withanc = bind_rows(snpinf, 
                    Tissue_index %>%
                      select(-tind) %>% 
                      mutate(Haplotype_Name = "Anc", MAF = NA, SD = NA, lb = 1, ub = 1)
                    )

# Merge the average cluster frequencies with the index information that spruce wants
toPrint = left_join(left_join(withanc, Tissue_index), Cluster_index) %>% 
    ungroup() %>%
    # need to substitute out spaces
    mutate(Tissue = gsub(" ", "_", Tissue)) %>%
    mutate(Haplotype = gsub(" ", "_", Haplotype_Name)) %>%
    mutate(sind = tind, sample = Tissue) %>%
    select(tind, Tissue, sind, sample, cind, Haplotype, lb, ub)  %>% 
    arrange(tind, cind) %>% 
    mutate(ub = pmax(lb, ub)) 

head(toPrint) 
```

Then, I'll write these results out to run with MACHINA 

```{r Write out Machina inputs}

## ==== File paths output data ==== ##

# Data from lofreq and varscan labeled with G1 and G2 mutations
machina.input.data = "./spruce_frequencies.tsv"
# Output data from MACHINA 
machina.output.data = "./spruce_out.tsv"


# Header needs to specify these variables
write(
  paste0(
    nrow(Tissue_index),
    "\n", nrow(Tissue_index),
    "\n", nrow(Cluster_index),
    "\n#", collapse = "" ),
  machina.input.data
)

# Then write the cluster frequencies 
write.table(toPrint,
            machina.input.data,
            sep = "\t", 
            quote = FALSE, 
            row.names = FALSE,
            col.names = FALSE,
            append = TRUE)

```

Now, run the following command in an environment with MACHINA installed. 

```{r Run Machina}

print(paste0("generatemutationtrees ", machina.input.data, "> ", machina.output.data))

```
What do the trees look like? 

```{r Process Spruce Output, echo=T}

# Read in the output of MACHINA
machina.output = readLines(machina.output.data)

# Process the output trees into edges
treenum = NA
tree.df = foreach(i = 1:length(machina.output), .combine = "rbind") %do% {
    toParse = machina.output[i]
    if(grepl("^#", toParse)){ return(NULL) }
    if(grepl("#trees", toParse)){ return(NULL) }
    if(grepl("#edges", toParse)){ 
        treenum = gsub("[0-9]+ \\#edges, ", "", toParse)
        return(NULL) 
    }
    tibble(toParse) %>%
      separate(toParse, into = c("from", "to"), sep = " ") %>% 
        mutate(tree = treenum)
}

tree.df = tree.df %>% 
  mutate(treenum = as.numeric(gsub("tree ", "", tree)))


```

```{r Spruce Trees, message=F, warning=F, fig.align='center', fig.width=10, fig.height=10, echo=T}

names(haplotype.colors) = gsub(" ", "_", names(haplotype.colors ))

ggnetwork(network(tree.df %>% filter(treenum <= 13), multiple = TRUE), by = "treenum")  %>%
  ggplot(aes(x = x, y = y, xend = xend, yend = yend)) +
    theme_blank() +
    geom_edges(curvature = 0.15) +
    geom_nodes(aes(fill = vertex.names), size = 10, pch = 21) +
    scale_fill_manual(values = haplotype.colors) +
    geom_nodetext(aes(y = y - 0.06, label =  vertex.names), color = "black", size = 2) +
    facet_wrap(~treenum) +
    theme(legend.position = "none") 

```

## Potential Driver Mutations

```{r Load Drivers}

driver.snps = read_csv("driver_snps.csv", show_col_types = FALSE) %>% 
  pull(SNP) %>% 
  unique()

expanded.df %>% 
  filter(SNP %in% driver.snps) %>% 
  select(SNP, Haplotype_Name) %>% 
  distinct() %>% 
  group_by(Haplotype_Name) %>% 
  count() %>% 
  arrange(-n)
  

```

