---
title: "4. Genotype Subclonal SNPs"
author:
    - "Will Hannon"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
  highlight: tango
  number_sections: no
  theme: default
  toc: yes
  toc_depth: 3
  toc_float:
    collapsed: no
    smooth_scroll: yes
---

The goal of this notebook is to establish a method to genotype SNPs as either belonging to `genome-1` or to `genome-2`. 

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```

```{r Required Packages, message=FALSE, warning=FALSE, echo=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed
packages = c("tidyverse", "foreach", "emdbook")

## Check that packages are installed, if not, install them
# installed_packages <- packages %in% rownames(installed.packages())
# if (any(installed_packages == FALSE)) {
#   install.packages(packages[!installed_packages])
# }

## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```

```{r Inputs, echo=T}

## ==== File paths input data ==== ##

# Data from lofreq and varscan labeled with subclonal haplotypes
updated.data = "../../results/variants/clustered_variants.csv"
# updated.data = snakemake@params[["incsv"]]

# Counts of SNPs on reads belonging to genome-1 or genome-2
reads.data = "../../results/bridging/genotyped.csv"
# reads.data = snakemake@input[[1]]
  
# Annotations for MeV genes
annotations.filepath = "../../config/annotations.csv"
# annotations.filepath = snakemake@params[["annotations"]]

```

```{r Outputs, echo=T}

## ==== File paths output data ==== ##

# Path to save the results
output.path = "../../results/variants/assigned_variants.csv"

# output.path = snakemake@params[["outcsv"]]

```

```{r Import and Format Data, echo = T}

# Download the variants
haplotypes = read_csv(updated.data, show_col_types = FALSE) %>% 
  select(Haplotype, SNP) %>% 
  distinct()

# Download the counts of SNPs on genome-1 and genome-2 
genotyping.df = read_csv(reads.data, show_col_types = FALSE) %>% 
  select(!AF) %>% 
  rename(total = DP, 
         var = OBSV) %>% 
  mutate(genotype = if_else(genotype == "genome-1", "g1", "g2")) %>% 
  pivot_wider(names_from = genotype,
              values_from = c(total, var), 
              names_sep = "_", 
              values_fill = 0) %>% 
  left_join(., haplotypes, by = "SNP") 

# Haplotyped SNPs for linking subclonal SNPs
updated.df = read_csv(updated.data, show_col_types = FALSE)

```

```{r Assign Genotypes for each Tau, echo=T}

# Log-spaced sequence of taus
taus = lseq(0.00001, 1, length.out = 200)

# Minimum significance
min.sig = 0.05

# Minimum number of variant observations? 

# Assign to genome for each value of tau
assignmentsByTau = foreach(tau = taus, .combine = "rbind")%do%{

    allSNPs = genotyping.df %>% 
        gather(cat, count, -POS, -REF, -ALT, -SNP, -Tissue, -Haplotype) %>% 
        separate(cat, into = c("cat", "G"), sep = "_") %>% 
        spread(cat, count) %>% 
        group_by(SNP, G) %>% 
        mutate(linked= pbinom( var, total, tau, lower.tail = FALSE), 
               sig = linked < min.sig/n()) %>% 
        mutate(sig = ifelse(total == 0, NA, sig)) %>% 
        summarize(TissueCount = sum(sig, na.rm = TRUE), .groups = "drop") %>% 
        spread(G, TissueCount) %>% 
        mutate(G1orG2 = ifelse(g1 > 0 & g2 == 0, "G1",
                        ifelse(g2 > 0 & g1 == 0, "G2", 
                        ifelse(g1 > 0 & g2 > 0, "G1/G2", 
                        ifelse(g1 == 0 & g2 == 0, "-", "*"))))) %>% 
        ungroup()


    left_join(genotyping.df, allSNPs, by = "SNP") %>% select( -total_g1, -total_g2, -var_g1, -var_g2) %>% mutate(tau = tau)

}


```

## G1/G2 assignment for Haplotypes

What are the `genome1/2` assignments for each identified haplotype for every value of `tau` between `0.00001` and `1`? 
```{r Assignments for all Haplotypes, warning=F, message=F, fig.align='center', fig.width=18, fig.height=14}

assignment.order = c('G1', 'G2', 'G1/G2', '-')
assignment.colors = c("#307EC2", "#B63F3E", "purple", "darkgrey")
  
assignmentsByTau %>%
  select(-Tissue) %>% 
  unique() %>%
  group_by(G1orG2, Haplotype, tau) %>%
  summarize(n = n()) %>% 
  spread(G1orG2, n) %>%
  gather(type, count, -Haplotype, -tau) %>% 
  mutate(count = ifelse(is.na(count), 0, count)) %>%
  spread(type, count) %>% 
  gather(type, count, -Haplotype, -tau) %>%
  group_by(Haplotype, tau) %>%
  mutate(total = sum(count)) %>% 
  mutate(prop = count/total) %>% 
    ggplot(aes(x = tau, y = prop, col = factor(type, level = assignment.order), group = type)) +
      geom_path(size = 1.5) + 
      scale_x_log10() +
      facet_wrap(~Haplotype) +
      labs(y = "Proportion of SNPs assigned") +  
      scale_color_manual(values = assignment.colors, name = "Genome Assignment") +
      theme_classic() +
      theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
      theme(legend.position="bottom", legend.box = "horizontal") +
      theme(legend.box.background = element_rect(colour = "black")) +
      theme(text=element_text(size=18)) + 
      theme(plot.title = element_text(hjust = 0.5)) 


```

## Genotype Subclonal SNPs

How does this look for subclonal SNPs? Can we assign most to either `genome-1` or `genome-2`? 

```{r Annotations, message=FALSE, warning=FALSE}

# Import genome annotations
annotations.df = read_csv(annotations.filepath, show_col_types = FALSE)

# Get the interval positions for these genes 
N = annotations.df %>% dplyr::filter(Locus == "N")
P.V.C = annotations.df %>% dplyr::filter(`Protein Name` == "phosphoprotein")
M = annotations.df %>% dplyr::filter(Locus == "M")
F. = annotations.df %>% dplyr::filter(Locus == "F")
H = annotations.df %>% dplyr::filter(Locus == "H")
L = annotations.df %>% dplyr::filter(Locus == "L")

```


```{r Subclonal Assignments, warning=F, message=F, echo = T}

subclonal.assignments = assignmentsByTau %>%
  filter(Haplotype == "subclonal") %>%
  select(-Tissue) %>% 
  unique() %>%
  group_by(G1orG2, SNP, tau) %>%
  summarize(n = n()) %>% 
  spread(G1orG2, n) %>% 
  gather(type, count, -SNP, -tau) %>% 
  mutate(count = ifelse(is.na(count), 0, count)) %>%
  spread(type, count) %>% 
  gather(type, count, -SNP, -tau) %>% group_by(SNP, tau) %>%
  mutate(total = sum(count)) %>% 
  mutate(prop = count/total) %>%
  group_by(SNP) %>%
  filter(type == "G1" | type == "G2") %>% 
  filter(prop > 0) %>%
  select(SNP, type) %>% 
  unique()

```

There isn't much of a difference between the assignment of SNPs to `genome-1` or `genome-2` across the MeV genome. 

```{r Subclonal Assignments Positions, warning=F, message=F, fig.align='center', fig.width=10, fig.height=8}

# Import genome annotations
annotations.df = read_csv(annotations.filepath, show_col_types = FALSE)

# Get the interval positions for these genes 
N = annotations.df %>% dplyr::filter(Locus == "N")
P.V.C = annotations.df %>% dplyr::filter(`Protein Name` == "phosphoprotein")
M = annotations.df %>% dplyr::filter(Locus == "M")
F. = annotations.df %>% dplyr::filter(Locus == "F")
H = annotations.df %>% dplyr::filter(Locus == "H")
L = annotations.df %>% dplyr::filter(Locus == "L")

left_join(subclonal.assignments, genotyping.df) %>%
  select(SNP, type, POS) %>%
  unique() %>% 
  ggplot(aes(x = POS, y = type, col = type)) +
    geom_point(pch = "|", size = 3) +
    labs(x = "Position", y = "Genome")+
    scale_color_manual(values = assignment.colors, name = "Genome Assignment") +
  
    # == Annotations of genes == #
  
    # Nucleoprotein
    annotate("rect", xmin = N$Start-1, xmax = N$Stop, ymin = 0, ymax = .1 ,
               alpha = .2, fill = "dodgerblue2", col = "black", size = .1) +
    annotate(geom = "text", x = (N$Stop + N$Start)/2, y = .05, label = "N", size = 8) +
    annotate("rect", xmin = N$Start-1, xmax = N$Stop, ymin = 0, ymax = 2.5 ,
                   alpha = .1, fill = "dodgerblue2", size = .1) +  
    # P/V/C
    annotate("rect", xmin = P.V.C$Start, xmax = P.V.C$Stop, ymin = 0, ymax = .1 ,
               alpha = .2, fill = "yellow3", col = "black", size = .1) +
    annotate(geom = "text", x = (P.V.C$Stop + P.V.C$Start)/2, y = .05, label = "P/V/C", size = 8) +
    annotate("rect", xmin = P.V.C$Start, xmax = P.V.C$Stop, ymin = 0, ymax = 2.5 ,
               alpha = .1, fill = "yellow3", size = .1) +
    # Matrix protein
    annotate("rect", xmin = M$Start, xmax = M$Stop, ymin = 0, ymax = .1 ,
               alpha = .2, fill = "#FF7F00", col = "black", size = .1) +
    annotate(geom = "text", x = (M$Stop + M$Start)/2, y = .05, label = "M", size = 8)  +
    annotate("rect", xmin = M$Start, xmax = M$Stop, ymin = 0, ymax = 2.5 ,
               alpha = .1, fill = "#FF7F00", size = .1) +
    # Fusion protein
    annotate("rect", xmin = F.$Start, xmax = F.$Stop, ymin = 0, ymax = .1 ,
               alpha = .2, fill = "#6A3D9A", col = "black", size = .1) +
    annotate(geom = "text", x = (F.$Stop + F.$Start)/2, y = .05, label = "F", size = 8) +
    annotate("rect", xmin = F.$Start, xmax = F.$Stop, ymin = 0, ymax = 2.5 ,
               alpha = .1, fill = "#6A3D9A", size = .1) +
    # Hemagluttinin
    annotate("rect", xmin = H$Start, xmax = H$Stop, ymin = 0, ymax = .1 ,
               alpha = .2, fill = "green4", col = "black", size = .1) +
    annotate(geom = "text", x = (H$Stop + H$Start)/2, y = .05, label = "H", size = 8) +
    annotate("rect", xmin = H$Start, xmax = H$Stop, ymin = 0, ymax = 2.5,
               alpha = .1, fill = "green4", size = .1) +
    # Large protein
    annotate("rect", xmin = L$Start, xmax = L$Stop, ymin = 0, ymax = .1 ,
               alpha = .2, fill = "#E31A1C", col = "black", size = .1) +
    annotate(geom = "text", x = (L$Stop + L$Start)/2, y = .05, label = "L", size = 8) +
    annotate("rect", xmin = L$Start, xmax = L$Stop, ymin = 0, ymax = 2.5 ,
               alpha = .1, fill = "#E31A1C", size = .1) +
    theme_classic() +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=24)) + 
    theme(axis.text.x = element_text(size=28)) + 
    theme(axis.title.x = element_text(size=30)) + 
    theme(plot.title = element_text(hjust = 0.5))  +
    theme(axis.title.y = element_blank()) +
    scale_x_continuous(limits=c(0,16000)) +
    theme(legend.position = "none") 

ggsave("../../results/figures/subclonal-snps-in-g1-and-g2.png", width = 10, height = 8, dpi = 300)

```


```{r Write out the assignments}

fully.assigned.df = left_join(updated.df, subclonal.assignments, by = "SNP") %>% 
  rename(Proposed_Background = type) %>% 
  mutate(Proposed_Background = case_when(Proposed_Background == "G1" ~ "genome-1",
                                         Proposed_Background == "G2" ~ "genome-2",
                                         is.na(Proposed_Background) ~ "unknown"))

write_csv(fully.assigned.df, output.path)

```



