{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2f602716",
   "metadata": {},
   "source": [
    "# Genotype Mutations\n",
    "\n",
    "The goal of this notebook is to genotype mutations as either `genome-1` or `genome-2` by splitting reads onto either of those backgrounds. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "59e4961c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from collections import Counter, defaultdict\n",
    "import pysam \n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "from Bio import SeqIO "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cf20579",
   "metadata": {},
   "source": [
    "## Inputs\n",
    "\n",
    "These will ultimatley be wrapped into a `Snakemake` pipeline. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c08a71af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Bam files - will be supplied by snakemake \n",
    "bams = list()\n",
    "for root, dirs, files in os.walk(\"../../results/realigned/\"):\n",
    "    for file in files:\n",
    "        if file.endswith(\".bam\"):\n",
    "             bams.append(os.path.join(root, file))\n",
    "                \n",
    "# Base inputs for analysis\n",
    "ref_path = \"../../config/ref/MeVChiTok-SSPE.fa\" \n",
    "contig = \"MeVChiTok\"\n",
    "minimum_qual = 25\n",
    "\n",
    "# Import the dataframe with mutations labled by identity\n",
    "genomes_path = \"../../results/spatial/labeled_variants.csv\" # Path to the major haplotypes\n",
    "genomes_df = pd.read_csv(genomes_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64f75929",
   "metadata": {},
   "source": [
    "## Functions \n",
    "\n",
    "Modularized code for running the analysis. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "0edaeee0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_read(read):\n",
    "    \"\"\"\n",
    "    Helper function to decide what reads should\n",
    "    be keep when parsing alignment file with `pysam`. \n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    read : AlignedSegment\n",
    "        read from alignment file parsed with `pysam`.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    bool\n",
    "        True/False if read should be included\n",
    "        \n",
    "    \"\"\"\n",
    "    # Exclude Quality Failures\n",
    "    if read.is_qcfail:\n",
    "        return False\n",
    "    # Exclude Secondary Mappings\n",
    "    if read.is_secondary:\n",
    "        return False\n",
    "    # Exclude Unmapped Reads\n",
    "    if read.is_unmapped:\n",
    "        return False\n",
    "    else:\n",
    "        return True\n",
    "    \n",
    "\n",
    "def assign_reads(genome_dict, bampath):\n",
    "    \"\"\"\n",
    "    Save a list of read names associated with either genome-1 or genome-2 SNPs\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    genome_dict : dict\n",
    "        Dictionary of either genome-1 or genome-2 only SNPs.\n",
    "        \n",
    "    bampath: str\n",
    "        Path to bamfile to assign reads for. \n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    dict\n",
    "        Dictionary of read name sets assigned to either genome-1 or genome-2 key. \n",
    "    \"\"\"\n",
    "    \n",
    "    # Assigning reads to either genome 1 or genome 2\n",
    "    assert len(genome_dict.keys()) == 2, \"There must be two genotypes.\"\n",
    "    assert \"genome-1\" in genome_dict.keys() and \"genome-2\" in genome_dict.keys(), \"Must be genome-1 and genome-2.\"\n",
    "            \n",
    "    # Empty list to store qnames\n",
    "    qnames = defaultdict(list)\n",
    "    \n",
    "    # Open the alignment file\n",
    "    with pysam.AlignmentFile(bampath, \"rb\") as bamfile:\n",
    "        # Iterate through the bams and get reads with genome 1 or genome 2 SNPs \n",
    "        for genotype, snps in genome_dict.items():\n",
    "            # Iterate over the pileup column at each position\n",
    "            for pileupcolumn in bamfile.pileup(stepper = 'nofilter', flag_filter = 0, min_base_quality = 25):\n",
    "                # Check if the position contains a genotypeable position\n",
    "                if pileupcolumn.reference_pos + 1 in set(pos for pos, snp in snps):\n",
    "                    for pileupread in pileupcolumn.pileups:\n",
    "                        # Check that it's a good read\n",
    "                        if check_read(pileupread.alignment) and not pileupread.is_del and not pileupread.is_refskip:\n",
    "                            # Save the qname\n",
    "                            if (pileupcolumn.reference_pos + 1, pileupread.alignment.query_sequence[pileupread.query_position]) in snps:\n",
    "                                qnames[genotype].append(pileupread.alignment.query_name)\n",
    "                            else:\n",
    "                                if genotype == \"genome-1\":\n",
    "                                    qnames[\"genome-2\"].append(pileupread.alignment.query_name)\n",
    "                                else:\n",
    "                                    qnames[\"genome-1\"].append(pileupread.alignment.query_name) \n",
    "                                \n",
    "    return {k:set(v) for k,v in qnames.items()}\n",
    "\n",
    "\n",
    "def subset_bam(bampath, outpath, qnames):\n",
    "    \"\"\"\n",
    "    Take a bamfile and subset it into a new bam file \n",
    "    based on a list of read names.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    bampath : str\n",
    "        Path pointing to an indexed Bam format file.\n",
    "        \n",
    "    read_names: list\n",
    "        A list of readnames to extract the Bam file. \n",
    "        \n",
    "    prefix: str\n",
    "        The prefix of the file to be written out during the analysis.\n",
    "        \n",
    "    \"\"\"\n",
    "    with pysam.AlignmentFile(bampath) as bam:\n",
    "        with pysam.AlignmentFile(outpath, \"w\", template = bam) as outbam:\n",
    "            for read in bam.fetch(until_eof=True):\n",
    "                if read.query_name in qnames:\n",
    "                    outbam.write(read)\n",
    "     # Message about status\n",
    "    print(f\"Wrote bam subset to {outpath}\\n\\nDone!\\n\")\n",
    "    \n",
    "\n",
    "def sort_bam(bampath):\n",
    "    \"\"\"\n",
    "    Take a bamfile and sort and index it.\n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    bampath : str\n",
    "        Path pointing to an indexed Bam format file.\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    # Sort the bam \n",
    "    print(\"Sorting Bam\\n\")\n",
    "    pysam.sort(\"-o\", f\"{os.path.splitext(bampath)[0]}.sorted.bam\", bampath)\n",
    "    \n",
    "    # Index the sorted bam\n",
    "    print(\"Indexing Bam\\n\")\n",
    "    pysam.index(f\"{os.path.splitext(bampath)[0]}.sorted.bam\")   \n",
    "    \n",
    "    print(\"Done!\\n\")\n",
    "        \n",
    "\n",
    "def count_coverage(bampath,\n",
    "                   snps,\n",
    "                   callback_function = check_read, \n",
    "                   contig = \"MeVChiTok\",\n",
    "                   ref_path = \"../../config/ref/MeVChiTok-SSPE.fa\", \n",
    "                   minimum_qual = 25):\n",
    "    \"\"\"\n",
    "    Count the coverage of all SNPs of interest in a filtered BAM file. \n",
    "    \n",
    "    Parameters\n",
    "    ----------\n",
    "    bampath : str\n",
    "        Path pointing to an indexed Bam format file.\n",
    "\n",
    "    snps : list\n",
    "        List of all SNPs to filter resulting dataframe by\n",
    "        \n",
    "    Returns\n",
    "    -------\n",
    "    pd.DataFrame\n",
    "        Dataframe of SNP counts and total coverage in an alignment file. \n",
    "        \n",
    "    \"\"\"\n",
    "    with pysam.AlignmentFile(bampath, \"rb\") as bamfile:\n",
    "        \n",
    "        # The count_coverage method counts the occurances of each base at each position. \n",
    "        # It excludes reads based on the callback function.\n",
    "        count_df = pd.DataFrame.from_dict({base:counts for base, counts in zip(\"ACGT\",\n",
    "                                                                               bamfile.count_coverage(contig=contig,\n",
    "                                                                               read_callback=callback_function,\n",
    "                                                                               quality_threshold=minimum_qual))})\n",
    "        # Add the depth at each position\n",
    "        count_df['DP'] = count_df.sum(axis = 1)\n",
    "        # Add the position (1-indexed)\n",
    "        count_df['POS'] = count_df.index + 1\n",
    "        # Add the reference allele\n",
    "        count_df['REF'] = [base.upper() for base in list(SeqIO.parse(ref_path, \"fasta\"))[0].seq]\n",
    "        # convert counts to frequency \n",
    "        count_df.iloc[:,0:4] = count_df.iloc[:,0:4].div(count_df.DP, axis = 0)\n",
    "        # handle any NaNs created by dividing by 0 coverage\n",
    "        count_df = count_df.fillna(0)\n",
    "        # Melt the data frame to a longer ('tidy') form\n",
    "        count_df = pd.melt(count_df, \n",
    "                           id_vars=['POS', 'DP', 'REF'],\n",
    "                           value_vars=[base for base in 'ATGC'],\n",
    "                           value_name='AF',\n",
    "                           var_name='ALT')\n",
    "        # Remove anything with 0 coverage.\n",
    "        count_df = count_df[count_df['AF'] > 0]\n",
    "        # TRUE/FALSE if it's a SNP\n",
    "        count_df['IS_SNP'] = np.where(count_df['ALT'] != count_df['REF'], True, False)\n",
    "        # Add the number of times a given allele is observed.\n",
    "        count_df['OBSV'] = count_df.DP * count_df.AF\n",
    "        \n",
    "        # Filter to only get SNPs\n",
    "        count_df.loc[count_df['IS_SNP']]\n",
    "        count_df = count_df.drop(columns=['IS_SNP'])\n",
    "        \n",
    "        # New colunm of the SNP to filter by\n",
    "        count_df['SNP'] = count_df[\"REF\"] + count_df[\"POS\"].astype(str) + count_df[\"ALT\"]\n",
    "        \n",
    "        return count_df[count_df['SNP'].isin(snps)]\n",
    "        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "504cda01",
   "metadata": {},
   "source": [
    "## Analysis \n",
    "\n",
    "**TEST**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "c2d52173",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Assigning reads for Frontal Cortex 3\n",
      "0.9032430727469103 percent of reads are in both sets of genomes.\n",
      "\n",
      "Wrote bam subset to genome-1.bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Sorting Bam\n",
      "\n",
      "Indexing Bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Wrote bam subset to genome-2.bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Sorting Bam\n",
      "\n",
      "Indexing Bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Assigning reads for Internal Capsule\n",
      "0.4547952580163973 percent of reads are in both sets of genomes.\n",
      "\n",
      "Wrote bam subset to genome-1.bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Sorting Bam\n",
      "\n",
      "Indexing Bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Wrote bam subset to genome-2.bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Sorting Bam\n",
      "\n",
      "Indexing Bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Assigning reads for Brain Stem\n",
      "0.2639651851323393 percent of reads are in both sets of genomes.\n",
      "\n",
      "Wrote bam subset to genome-1.bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Sorting Bam\n",
      "\n",
      "Indexing Bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Wrote bam subset to genome-2.bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Sorting Bam\n",
      "\n",
      "Indexing Bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Assigning reads for Cerebellum Nucleus\n",
      "0.35071668191521804 percent of reads are in both sets of genomes.\n",
      "\n",
      "Wrote bam subset to genome-1.bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Sorting Bam\n",
      "\n",
      "Indexing Bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Wrote bam subset to genome-2.bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Sorting Bam\n",
      "\n",
      "Indexing Bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Assigning reads for Frontal Cortex 1\n",
      "0.9410133697964144 percent of reads are in both sets of genomes.\n",
      "\n",
      "Wrote bam subset to genome-2.bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Sorting Bam\n",
      "\n",
      "Indexing Bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Wrote bam subset to genome-1.bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Sorting Bam\n",
      "\n",
      "Indexing Bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Assigning reads for UBS\n",
      "0.33547397680437074 percent of reads are in both sets of genomes.\n",
      "\n",
      "Wrote bam subset to genome-1.bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Sorting Bam\n",
      "\n",
      "Indexing Bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Wrote bam subset to genome-2.bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Sorting Bam\n",
      "\n",
      "Indexing Bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Assigning reads for Cerebellum\n",
      "0.24767801857585142 percent of reads are in both sets of genomes.\n",
      "\n",
      "Wrote bam subset to genome-1.bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Sorting Bam\n",
      "\n",
      "Indexing Bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Wrote bam subset to genome-2.bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Sorting Bam\n",
      "\n",
      "Indexing Bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Assigning reads for Hippocampus\n",
      "0.4996852376455776 percent of reads are in both sets of genomes.\n",
      "\n",
      "Wrote bam subset to genome-2.bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Sorting Bam\n",
      "\n",
      "Indexing Bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Wrote bam subset to genome-1.bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Sorting Bam\n",
      "\n",
      "Indexing Bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Assigning reads for Parietal Lobe\n",
      "0.328241025398477 percent of reads are in both sets of genomes.\n",
      "\n",
      "Wrote bam subset to genome-1.bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Sorting Bam\n",
      "\n",
      "Indexing Bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Wrote bam subset to genome-2.bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Sorting Bam\n",
      "\n",
      "Indexing Bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Assigning reads for Midbrain\n",
      "0.2326485142102127 percent of reads are in both sets of genomes.\n",
      "\n",
      "Wrote bam subset to genome-2.bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Sorting Bam\n",
      "\n",
      "Indexing Bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Wrote bam subset to genome-1.bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Sorting Bam\n",
      "\n",
      "Indexing Bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Assigning reads for Occipital Lobe\n",
      "0.5448431035734159 percent of reads are in both sets of genomes.\n",
      "\n",
      "Wrote bam subset to genome-2.bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Sorting Bam\n",
      "\n",
      "Indexing Bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Wrote bam subset to genome-1.bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Sorting Bam\n",
      "\n",
      "Indexing Bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Assigning reads for Temporal Lobe\n",
      "0.6327333412802479 percent of reads are in both sets of genomes.\n",
      "\n",
      "Wrote bam subset to genome-2.bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Sorting Bam\n",
      "\n",
      "Indexing Bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Wrote bam subset to genome-1.bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Sorting Bam\n",
      "\n",
      "Indexing Bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Assigning reads for Frontal Cortex 2\n",
      "0.6093571842464345 percent of reads are in both sets of genomes.\n",
      "\n",
      "Wrote bam subset to genome-2.bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Sorting Bam\n",
      "\n",
      "Indexing Bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Wrote bam subset to genome-1.bam\n",
      "\n",
      "Done!\n",
      "\n",
      "Sorting Bam\n",
      "\n",
      "Indexing Bam\n",
      "\n",
      "Done!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Isolate the genome-1 and genome-2 only mutations. \n",
    "genome_dict = {\n",
    "'genome-1': set(zip(genomes_df.loc[genomes_df.Haplotype == \"genome-1\"].POS, genomes_df.loc[genomes_df.Haplotype == \"genome-1\"].ALT)),\n",
    "'genome-2': set(zip(genomes_df.loc[genomes_df.Haplotype == \"genome-2\"].POS, genomes_df.loc[genomes_df.Haplotype == \"genome-2\"].ALT))\n",
    "}\n",
    "\n",
    "tissue_df_list = []\n",
    "for bam in bams: \n",
    "    \n",
    "    tissue = \" \".join(os.path.basename(bam).split(\".\")[0].split(\"_\"))\n",
    "    print(f'Assigning reads for {tissue}')\n",
    "    \n",
    "    # Get the reads from a bam belonging to each tissue. \n",
    "    qnames_dict = assign_reads(genome_dict, bam)\n",
    "\n",
    "\n",
    "    # How many reads get assigned to both genomes (could be recombinats, reversions, or template switching)\n",
    "    reads_in_both = qnames_dict['genome-1'] & qnames_dict['genome-2']\n",
    "    reads_in_one_only = (qnames_dict['genome-1'] | qnames_dict['genome-2']) - (qnames_dict['genome-1'] & qnames_dict['genome-2'])\n",
    "\n",
    "    print(f\"{len(reads_in_both)/len(qnames_dict['genome-1'] | qnames_dict['genome-2']) * 100} percent of reads are in both sets of genomes.\\n\")\n",
    "\n",
    "    genotype_df_list = []\n",
    "    # For each genotype, get the count of SNPs\n",
    "    for genotype, qnames in qnames_dict.items():\n",
    "        qnames = qnames - reads_in_both\n",
    "        # Subset the bam for only these reads\n",
    "        subset_bam(bam, f\"{genotype}.bam\", qnames)\n",
    "        # Sort and index for variant id\n",
    "        sort_bam(f\"{genotype}.bam\")\n",
    "        # Count the coverage for SNPs of interest \n",
    "        coverage_df = count_coverage(f\"{genotype}.sorted.bam\",\n",
    "                                     snps=genomes_df.SNP.to_list(),\n",
    "                                     callback_function = check_read, \n",
    "                                     contig = \"MeVChiTok\",\n",
    "                                     ref_path = \"../../config/ref/MeVChiTok-SSPE.fa\", \n",
    "                                     minimum_qual = 25)\n",
    "\n",
    "        coverage_df['genotype'] = f\"{genotype}\"\n",
    "        genotype_df_list.append(coverage_df)\n",
    "\n",
    "    genotype_df = pd.concat(genotype_df_list)\n",
    "    genotype_df['Tissue'] = f\"{tissue}\"\n",
    "    \n",
    "    tissue_df_list.append(genotype_df)\n",
    "    \n",
    "final_df = pd.concat(tissue_df_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "2420da0a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>DP</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AF</th>\n",
       "      <th>OBSV</th>\n",
       "      <th>SNP</th>\n",
       "      <th>genotype</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>152</td>\n",
       "      <td>44</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.022727</td>\n",
       "      <td>1.0</td>\n",
       "      <td>G152A</td>\n",
       "      <td>genome-1</td>\n",
       "      <td>Frontal Cortex 3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809</th>\n",
       "      <td>810</td>\n",
       "      <td>7284</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.007688</td>\n",
       "      <td>56.0</td>\n",
       "      <td>G810A</td>\n",
       "      <td>genome-1</td>\n",
       "      <td>Frontal Cortex 3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1631</th>\n",
       "      <td>1632</td>\n",
       "      <td>6</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.166667</td>\n",
       "      <td>1.0</td>\n",
       "      <td>G1632A</td>\n",
       "      <td>genome-1</td>\n",
       "      <td>Frontal Cortex 3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2369</th>\n",
       "      <td>2370</td>\n",
       "      <td>6302</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.000635</td>\n",
       "      <td>4.0</td>\n",
       "      <td>G2370A</td>\n",
       "      <td>genome-1</td>\n",
       "      <td>Frontal Cortex 3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2798</th>\n",
       "      <td>2799</td>\n",
       "      <td>3793</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.006591</td>\n",
       "      <td>25.0</td>\n",
       "      <td>G2799A</td>\n",
       "      <td>genome-1</td>\n",
       "      <td>Frontal Cortex 3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59818</th>\n",
       "      <td>12137</td>\n",
       "      <td>179</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.050279</td>\n",
       "      <td>9.0</td>\n",
       "      <td>T12137C</td>\n",
       "      <td>genome-1</td>\n",
       "      <td>Frontal Cortex 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62290</th>\n",
       "      <td>14609</td>\n",
       "      <td>3092</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.002587</td>\n",
       "      <td>8.0</td>\n",
       "      <td>T14609C</td>\n",
       "      <td>genome-1</td>\n",
       "      <td>Frontal Cortex 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62347</th>\n",
       "      <td>14666</td>\n",
       "      <td>4826</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000829</td>\n",
       "      <td>4.0</td>\n",
       "      <td>T14666C</td>\n",
       "      <td>genome-1</td>\n",
       "      <td>Frontal Cortex 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62766</th>\n",
       "      <td>15085</td>\n",
       "      <td>4804</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.037052</td>\n",
       "      <td>178.0</td>\n",
       "      <td>T15085C</td>\n",
       "      <td>genome-1</td>\n",
       "      <td>Frontal Cortex 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63058</th>\n",
       "      <td>15377</td>\n",
       "      <td>4749</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.000842</td>\n",
       "      <td>4.0</td>\n",
       "      <td>T15377C</td>\n",
       "      <td>genome-1</td>\n",
       "      <td>Frontal Cortex 2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6823 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         POS    DP REF ALT        AF   OBSV      SNP  genotype  \\\n",
       "151      152    44   G   A  0.022727    1.0    G152A  genome-1   \n",
       "809      810  7284   G   A  0.007688   56.0    G810A  genome-1   \n",
       "1631    1632     6   G   A  0.166667    1.0   G1632A  genome-1   \n",
       "2369    2370  6302   G   A  0.000635    4.0   G2370A  genome-1   \n",
       "2798    2799  3793   G   A  0.006591   25.0   G2799A  genome-1   \n",
       "...      ...   ...  ..  ..       ...    ...      ...       ...   \n",
       "59818  12137   179   T   C  0.050279    9.0  T12137C  genome-1   \n",
       "62290  14609  3092   T   C  0.002587    8.0  T14609C  genome-1   \n",
       "62347  14666  4826   T   C  0.000829    4.0  T14666C  genome-1   \n",
       "62766  15085  4804   T   C  0.037052  178.0  T15085C  genome-1   \n",
       "63058  15377  4749   T   C  0.000842    4.0  T15377C  genome-1   \n",
       "\n",
       "                 Tissue  \n",
       "151    Frontal Cortex 3  \n",
       "809    Frontal Cortex 3  \n",
       "1631   Frontal Cortex 3  \n",
       "2369   Frontal Cortex 3  \n",
       "2798   Frontal Cortex 3  \n",
       "...                 ...  \n",
       "59818  Frontal Cortex 2  \n",
       "62290  Frontal Cortex 2  \n",
       "62347  Frontal Cortex 2  \n",
       "62766  Frontal Cortex 2  \n",
       "63058  Frontal Cortex 2  \n",
       "\n",
       "[6823 rows x 9 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "8ea40479",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df.to_csv(\"../../config/genotyped.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1633e529",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
