---
title: "Haplotype Clustering"
author: "Will Hannon"
date: "2/6/2022"
output: html_document
---

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```

```{r Required Packages, message=FALSE, warning=FALSE, echo=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed

packages = c("tidyverse", "cluster", "ggrepel", "gridExtra", 'grid', 'uwot')
## Check that packages are installed, if not, install them
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}
## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```

```{r Inputs, include=FALSE}

## ==== File paths input data ==== ##

# Data from lofreq and varscan labeled with G1 and G2 mutations
labeled.data = "../../results/variants/final_variants.csv"

# Data from lofreq and varscan labeled with subclonal haplotypes
updated.data = "../../results/variants/clustered_variants.csv"

# Haplotypes parsed from CliqueSNV
haplotype.data = "../../results/haplotypes/haplotypes.csv"

# Assigned genotype from reads 
assigned.data = "../../results/split/genotyped.csv"

# Import the sample data
sample.data = "../../config/samples.csv"

# Annotations 
annotations.filepath = "../../config/annotations.csv"

```

```{r Tissue Ordering, include=FALSE}

# Tissue order
tissue_order = c(
  "Frontal Cortex 2", 
  "Frontal Cortex 1", 
  "Frontal Cortex 3", 
  "Temporal Lobe", 
  "Parietal Lobe",
  "Occipital Lobe",
  "Hippocampus",
  "Internal Capsule", 
  "Cerebellum",
  "Cerebellum Nucleus",
  "Midbrain", 
  "UBS",
  "Brain Stem")

```

## Clustering SNPs to infer Haplotypes

Here, I'm attempting to do a little more unbiased way to infer haplotypes by clustering with PCA + Kmeans. 

```{r Prep for Clustering}

# Using the maximally lable data
updated.haplotype.df = read_csv(updated.data)

haplotypes.label = updated.haplotype.df %>% 
  select(SNP, Haplotype, Background) %>% 
  distinct()

# Expand the mutations to have a frequency for every tissue.
expanded.df = updated.haplotype.df %>% 
  select(SNP, Tissue, AF) %>% 
  pivot_wider(names_from = "Tissue", values_from = "AF", values_fill = 0) %>% 
  pivot_longer(cols = !SNP, names_to = "Tissue", values_to = "AF") %>% 
  left_join(., select(updated.haplotype.df, c("SNP", "Tissue", "DP")), by = c("SNP", "Tissue")) %>% 
  mutate(DP = if_else(is.na(DP), 0, DP)) %>% 
  left_join(., haplotypes.label, by = "SNP") 

# Get the mean frequency of the major genomes 
tissue.mean = updated.haplotype.df %>% 
  filter(Haplotype %in% c("genome-1", "genome-2")) %>% 
  group_by(Tissue, Haplotype) %>% 
  summarize(AF.mean = mean(AF, na.rm = TRUE), 
            SD = sd(AF, na.rm = TRUE),
            N = n()) %>% 
  mutate(SE = SD / sqrt(N),
         Lower.CI = qt(1 - (0.05 / 2), N - 1) * SE,
         Upper.CI = qt(1 - (0.05 / 2), N - 1) * SE) %>% 
  rename("AF" = AF.mean)

```

```{r SNPs to Haplotype}

# SNPs greater than 5% in X tissues
n.tissues = 2

snps.to.haplotype = updated.haplotype.df %>% 
  filter(!Haplotype %in% c("genome-1", "genome-2", "genome-1-1", "fixed", "both")) %>% 
  filter(AF >= 0.05) %>%
  group_by(SNP) %>% 
  count() %>% 
  filter(n >= n.tissues) %>%
  pull(SNP) 

print(paste("There are about", length(snps.to.haplotype), "SNPs that I can haplotype with clustering."))

```


```{r Cluster the SNPs, fig.width=25, fig.height=15}

set.seed(seed = 10)

# Number of inferred clusters
n.clusters = 20


# Prepare the SNPs as column and Allele frequency as row
snp.by.freq = updated.haplotype.df %>% 
  filter(SNP %in% snps.to.haplotype) %>% 
  select(AF, SNP, Tissue) %>% 
  pivot_wider(names_from = SNP, values_from = AF, values_fill = 0)  %>% 
  remove_rownames %>% 
  column_to_rownames(var = "Tissue") %>% 
  t()

# Scaled and centered PCA 
snp.pca = prcomp(snp.by.freq, center = TRUE, scale. = TRUE)

# Cluster with Kmeans
clusters = kmeans(snp.pca$x, centers = n.clusters)
cluster.df = clusters$cluster %>% 
  as.list() %>% 
  as_tibble() %>% 
  t() %>% 
  as.data.frame() %>% 
  rownames_to_column(var = "SNP") %>% 
  rename(cluster = "V1")

# Join the clusters to the dataframe for plotting
kmeans.cluster.haplotypes = expanded.df %>% 
  left_join(., cluster.df, by = "SNP") %>% 
  filter(!is.na(cluster)) %>% 
  mutate(cluster = as.character(cluster))
n.clusters.per.snp = kmeans.cluster.haplotypes %>% 
  select(SNP, cluster) %>%
  distinct() %>% 
  group_by(cluster) %>% 
  count() %>% 
  mutate(cluster_size = paste0(cluster, " (", n, " snps in cluster)"))
kmeans.cluster.haplotypes = left_join(kmeans.cluster.haplotypes, n.clusters.per.snp, by = "cluster")

# Mean of each tissue
kmeans.cluster.haplotypes.mean = kmeans.cluster.haplotypes %>% 
  group_by(Tissue, cluster, cluster_size) %>% 
  summarize(AF = mean(AF))

# Plot the clusters
kmeans.cluster.haplotypes %>% 
    ggplot(aes(x = Tissue, y = AF)) +
      geom_line(aes(group = SNP)) +
      geom_line(data = kmeans.cluster.haplotypes.mean, aes(x = Tissue, y = AF, group = 1, col = (cluster)), size = 1) +
      geom_ribbon(data = tissue.mean, aes(x=Tissue, ymin=AF-SD, ymax=AF+SD, group = Haplotype, fill = Haplotype), alpha=0.2, colour = NA) +
      facet_wrap(~cluster_size) + 
      xlim((tissue_order)) +
      scale_fill_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Haplotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))

ggsave(filename = "../../clusters.png", dpi = 300, width = 25, height = 15)

```

```{r}

updated.haplotype.df %>% 
  filter(!Haplotype %in% c("genome-1", "genome-2", "genome-1-1", "fixed", "both")) %>% 
  ggplot() +
    geom_histogram(aes(x = AF)) + 
    facet_wrap(~Tissue) +
    theme_bw()

```


```{r}
assigned.df = read_csv(assigned.data) %>% 
  left_join(., select(updated.haplotype.df, POS, REF), by = "POS") %>% 
  mutate(SNP = paste0(REF, POS, ALT)) %>% 
  distinct()
```

## Cluster 16 and 12

```{r Cluster X}

genome.2.af = tissue.mean %>% 
  filter(Haplotype == "genome-2") %>% 
  select(Tissue, "g2_AF" = AF)

kmeans.cluster.haplotypes %>% 
  filter(cluster %in% c(16, 12, 18)) %>% 
  left_join(., genome.2.af, by = "Tissue") %>% 
  mutate(AF = AF/g2_AF) %>% 
  group_by(Tissue, cluster) %>% 
  summarize(AF = mean(AF)) %>% 
    ggplot(aes(x = Tissue, y = AF, group = cluster, fill = cluster)) +
      geom_area() +
      xlim((tissue_order)) +
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Haplotype", col = "Cluster") +
      theme_bw(18) +
      theme(axis.text.x = element_text(angle = 45, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))
  
ggsave(filename = "../../genome-2.png", dpi = 300, width = 8, height = 6)

  
```


```{r}

genome.1.af = tissue.mean %>% 
  filter(Haplotype == "genome-1") %>% 
  select(Tissue, "g1_AF" = AF)

kmeans.cluster.haplotypes %>% 
  filter(cluster %in% c(10, 11, 17, 19, 3)) %>% 
  left_join(., genome.1.af, by = "Tissue") %>% 
  mutate(AF = AF/g1_AF) %>% 
  group_by(Tissue, cluster) %>% 
  summarize(AF = mean(AF)) %>% 
    ggplot(aes(x = Tissue, y = AF, group = cluster, fill= cluster)) +
      geom_area() +
      xlim((tissue_order)) +
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Haplotype", col = "Cluster") +
      theme_bw(18) +
      theme(axis.text.x = element_text(angle = 45, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))
  
ggsave(filename = "../../genome-1.png", dpi = 300, width = 8, height = 6)


```

```{r}

cluster.snps = kmeans.cluster.haplotypes %>% 
  filter(cluster == 5) %>% 
  pull(SNP)
  
assigned.df %>% 
  filter(SNP %in% cluster.snps) %>% view()

kmeans.cluster.haplotypes %>% 
  filter(SNP %in% cluster.snps) %>% 
    ggplot(aes(x = Tissue, y = AF)) +
      geom_line(aes(group = SNP)) +
      geom_ribbon(data = tissue.mean, aes(x=Tissue, ymin=AF-SD, ymax=AF+SD, group = Haplotype, fill = Haplotype), alpha=0.2, colour = NA) +
      facet_wrap(~Haplotype) + 
      xlim((tissue_order)) +
      scale_fill_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Haplotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))

```

```{r}

kmeans.cluster.haplotypes %>% 
  filter(cluster %in% c(10, 16, 19, 12))
    ggplot(aes(x = Tissue, y = AF)) +
      # geom_line(aes(group = SNP)) +
      geom_line(data = filter(kmeans.cluster.haplotypes.mean, cluster %in% c(10, 16, 19, 12)), aes(x = Tissue, y = AF, group = 1, col = (cluster)), size = 1) +
      geom_ribbon(data = tissue.mean, aes(x=Tissue, ymin=AF-SD, ymax=AF+SD, group = Haplotype, fill = Haplotype), alpha=0.2, colour = NA) +
      facet_wrap(~cluster_size, ncol = 4) + 
      xlim((tissue_order)) +
      scale_fill_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Haplotype", col = "Cluster") +
      theme_bw(16) +
      theme(axis.text.x = element_text(angle = 45, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))

ggsave(filename = "../../clusters-fixed.png", dpi = 300, width = 14, height = 5)

```

```{r}
# Prepare the SNPs as column and Allele frequency as row
frequency.by.snp = updated.haplotype.df %>% 
  filter(!Haplotype %in% c("fixed", "both")) %>%
  select(AF, SNP, Tissue) %>% 
  pivot_wider(names_from = SNP, values_from = AF, values_fill = 0)  %>% 
  remove_rownames %>% 
  column_to_rownames(var = "Tissue")

# Do the PCA 
pcobj = prcomp(frequency.by.snp, center = TRUE, scale. = TRUE)

# PCs to plot
choices = 1:2

# Scaling factors 
scale = 1
var.scale = scale 
obs.scale = 1 - scale

# Variables for plotting
circle.prob = 0.69
varname.adjust = 1.5
varname.size = 3

# ----------

# Pull the data from the PCA object
nobs.factor = sqrt(nrow(pcobj$x) - 1)
d = pcobj$sdev
u = sweep(pcobj$x, 2, 1 / (d * nobs.factor), FUN = '*')
v = pcobj$rotation
 
# Scores
choices = pmin(choices, ncol(u))
df.u = as.data.frame(sweep(u[,choices], 2, d[choices]^obs.scale, FUN='*'))

# Directions
v = sweep(v, 2, d^var.scale, FUN='*')
df.v = as.data.frame(v[, choices])

names(df.u)  = c('xvar', 'yvar')
names(df.v) = names(df.u)

df.u = df.u * nobs.factor

# Scale directions
r = sqrt(qchisq(circle.prob, df = 2)) * prod(colMeans(df.u^2))^(1/4)
v.scale = rowSums(v^2)
df.v = r * df.v / sqrt(max(v.scale))

# Append the proportion of explained variance to the axis labels
u.axis.labs = paste('PC', choices, sep='')
u.axis.labs = paste(u.axis.labs, 
                    sprintf('(%0.1f%% explained var.)', 
                            100 * pcobj$sdev[choices]^2/sum(pcobj$sdev^2)))

# Variables for text label placement
df.v$varname = rownames(v)
df.v$angle = with(df.v, (180/pi) * atan(yvar / xvar))
df.v$hjust = with(df.v, (1 - varname.adjust * sign(xvar)) / 2)


# Add the data from the haplotypes
df.u = df.u %>% 
  rownames_to_column(var = "Tissue") 

df.v = df.v %>% 
  rownames_to_column(var = "SNP") %>% 
  left_join(., select(updated.haplotype.df, SNP, Background, Haplotype, Gene)) %>% 
  distinct()
  

with.loadings.and.color = ggplot(data = df.u, aes(x = xvar, y = yvar)) + 
  xlab(u.axis.labs[1]) + 
  ylab(u.axis.labs[2]) +
  coord_equal() +
  geom_segment(data = df.v,
               aes(x = 0, y = 0, xend = xvar, yend = yvar, col = Haplotype),
               arrow = arrow(length = unit(1/2, 'picas'))) + 
  geom_point(size = 1.5) + 
  geom_text_repel(aes(label = Tissue)) +
  # geom_text(data = df.v, 
  #           aes(label = varname, x = xvar, y = yvar, 
  #               angle = angle, hjust = hjust), 
  #           color = 'darkred', size = varname.size) + 
  # scale_color_manual(values = c("#307EC2", "#B63F3E", "#bdbdbd")) + 
  theme_bw(18) + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + 
  theme(legend.position = "none")


```

```{r}
high.freq.snps = kmeans.cluster.haplotypes %>% 
  filter(cluster %in% c(10, 16, 19, 12, 17)) %>% 
  pull(SNP)

updated.haplotype.df %>% 
  filter(SNP %in% high.freq.snps) %>% view()
```

