{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1a759fd5",
   "metadata": {},
   "source": [
    "# Phylogenetic Analysis\n",
    "\n",
    "The goal of this notebook is to perform phylogenetic inference on the haplotypes identified in the brain. \n",
    "\n",
    "**Requirements:**\n",
    "Make sure you have `Biopython`, `MAFFT`, and `ETE3` installed in your environment. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b20b6636",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "from subprocess import call \n",
    "from Bio import SeqIO, AlignIO, SeqRecord, Seq\n",
    "from Bio.Align.Applications import MafftCommandline \n",
    "\n",
    "# from ete3 import Tree, TreeStyle, NodeStyle, TextFace, SequenceFace, CircleFace, faces, AttrFace "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1541c3cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path to output directory\n",
    "outpath = \"../../results/phylogeny\"\n",
    "\n",
    "if not os.path.exists(outpath):\n",
    "    os.mkdir(outpath)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d1f1ac8",
   "metadata": {},
   "source": [
    "## Make haplotypes from mutations\n",
    "\n",
    "I'll take the mutations from the variant calling analysis and make the haplotypes using the reference genome. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b8fa5014",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>AF</th>\n",
       "      <th>DP</th>\n",
       "      <th>Effect</th>\n",
       "      <th>Gene_Name</th>\n",
       "      <th>AA_Change</th>\n",
       "      <th>Accession</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>SNP</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Background</th>\n",
       "      <th>Haplotype</th>\n",
       "      <th>Proposed_Background</th>\n",
       "      <th>Haplotype_Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>537</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.3265</td>\n",
       "      <td>132284</td>\n",
       "      <td>Missense</td>\n",
       "      <td>N</td>\n",
       "      <td>Ser144Pro</td>\n",
       "      <td>SSPE_1</td>\n",
       "      <td>SSPE 1</td>\n",
       "      <td>T537C</td>\n",
       "      <td>N</td>\n",
       "      <td>genome-2</td>\n",
       "      <td>genome-2</td>\n",
       "      <td>unknown</td>\n",
       "      <td>genome 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>684</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.1140</td>\n",
       "      <td>127945</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>N</td>\n",
       "      <td>Leu193Leu</td>\n",
       "      <td>SSPE_1</td>\n",
       "      <td>SSPE 1</td>\n",
       "      <td>C684T</td>\n",
       "      <td>N</td>\n",
       "      <td>genome-1</td>\n",
       "      <td>cluster 1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cluster 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1328</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.6809</td>\n",
       "      <td>131283</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>N</td>\n",
       "      <td>Ser407Ser</td>\n",
       "      <td>SSPE_1</td>\n",
       "      <td>SSPE 1</td>\n",
       "      <td>T1328C</td>\n",
       "      <td>N</td>\n",
       "      <td>genome-1</td>\n",
       "      <td>genome-1-1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>genome 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1632</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.3126</td>\n",
       "      <td>121322</td>\n",
       "      <td>Missense</td>\n",
       "      <td>N</td>\n",
       "      <td>Gly509Ser</td>\n",
       "      <td>SSPE_1</td>\n",
       "      <td>SSPE 1</td>\n",
       "      <td>G1632A</td>\n",
       "      <td>N</td>\n",
       "      <td>genome-2</td>\n",
       "      <td>cluster 7</td>\n",
       "      <td>unknown</td>\n",
       "      <td>cluster 11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2139</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.6473</td>\n",
       "      <td>62808</td>\n",
       "      <td>Synonymous</td>\n",
       "      <td>P/V/C</td>\n",
       "      <td>His111His</td>\n",
       "      <td>SSPE_1</td>\n",
       "      <td>SSPE 1</td>\n",
       "      <td>T2139C</td>\n",
       "      <td>P/V/C</td>\n",
       "      <td>genome-1</td>\n",
       "      <td>genome-1</td>\n",
       "      <td>unknown</td>\n",
       "      <td>genome 01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    POS REF ALT      AF      DP      Effect Gene_Name  AA_Change Accession  \\\n",
       "0   537   T   C  0.3265  132284    Missense         N  Ser144Pro    SSPE_1   \n",
       "1   684   C   T  0.1140  127945  Synonymous         N  Leu193Leu    SSPE_1   \n",
       "5  1328   T   C  0.6809  131283  Synonymous         N  Ser407Ser    SSPE_1   \n",
       "7  1632   G   A  0.3126  121322    Missense         N  Gly509Ser    SSPE_1   \n",
       "9  2139   T   C  0.6473   62808  Synonymous     P/V/C  His111His    SSPE_1   \n",
       "\n",
       "   Tissue     SNP   Gene Background   Haplotype Proposed_Background  \\\n",
       "0  SSPE 1   T537C      N   genome-2    genome-2             unknown   \n",
       "1  SSPE 1   C684T      N   genome-1   cluster 1             unknown   \n",
       "5  SSPE 1  T1328C      N   genome-1  genome-1-1             unknown   \n",
       "7  SSPE 1  G1632A      N   genome-2   cluster 7             unknown   \n",
       "9  SSPE 1  T2139C  P/V/C   genome-1    genome-1             unknown   \n",
       "\n",
       "  Haplotype_Name  \n",
       "0       genome 2  \n",
       "1      cluster 1  \n",
       "5       genome 1  \n",
       "7     cluster 11  \n",
       "9      genome 01  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# SSPE reference\n",
    "reference = [base for record in SeqIO.parse(\"../../config/ref/MeVChiTok-SSPE.fa\", \"fasta\") for base in record.seq]\n",
    "\n",
    "# Import the mutations\n",
    "haplotypes_df = pd.read_csv(\"../../results/variants/validated_variants.csv\")\n",
    "\n",
    "# We only need the mutations that have been assigned to a haplotype\n",
    "haplotypes_df = haplotypes_df[~haplotypes_df.Haplotype.isin(['subclonal', 'both'])]\n",
    "\n",
    "haplotypes_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "457d67a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_haplotype(haplotype_name, reference):\n",
    "    \"\"\"\n",
    "    Add the SNPs the reference to make a haplotype.\n",
    "    \"\"\"\n",
    "    \n",
    "    assert haplotype_name in set(haplotypes_df.Haplotype_Name.to_list())\n",
    "    \n",
    "    # SNPs for a given haplotype\n",
    "    snps_df = haplotypes_df[haplotypes_df.Haplotype_Name == haplotype_name]\n",
    "    \n",
    "    # get a list of tuples for the SNPs\n",
    "    snps = [(pos, alt) for pos, alt in zip(snps_df.POS, snps_df.ALT)]\n",
    "    \n",
    "    # Copy the reference \n",
    "    haplotype = deepcopy(reference)\n",
    "    \n",
    "    # Change the mutation in the list.\n",
    "    for pos, alt in snps:\n",
    "        haplotype[pos + 1] = alt\n",
    "        \n",
    "    return haplotype\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8f95e06",
   "metadata": {},
   "source": [
    "### Haplotype Relationship\n",
    "\n",
    "![haplotypes](example_tree.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b0286d5",
   "metadata": {},
   "source": [
    "I'll make the fasta files for each haplotype by cumulativley building them alone the above tree. Starting with the main backgrounds (Genome-1 and Genome-2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a8e7b257",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Make genome-1 and genome-2\n",
    "genome_1 = make_haplotype(\"genome 01\", reference)\n",
    "genome_2 = make_haplotype(\"genome 2\", reference)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "274d0805",
   "metadata": {},
   "source": [
    "Then, I'll add the main branch-point from Genome-1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "74a4b4f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2. Make genome-1-1 and cluster 6\n",
    "genome_1_1 = make_haplotype(\"genome 1\", genome_1)\n",
    "cluster_6 = make_haplotype(\"cluster 6\", genome_1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf3d18ad",
   "metadata": {},
   "source": [
    "Then I'll add the Genome-1-1 children nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3071d739",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4. Make genome-1-1 children \n",
    "cluster_1 = make_haplotype(\"cluster 1\", genome_1_1)\n",
    "cluster_2 = make_haplotype(\"cluster 2\", genome_1_1)\n",
    "cluster_4 = make_haplotype(\"cluster 4\", genome_1_1)\n",
    "cluster_5 = make_haplotype(\"cluster 5\", genome_1_1)\n",
    "cluster_7 = make_haplotype(\"cluster 7\", genome_1_1)\n",
    "cluster_8 = make_haplotype(\"cluster 8\", genome_1_1)\n",
    "\n",
    "# 5. Cluster 3 is descended from cluster 4\n",
    "cluster_3 = make_haplotype(\"cluster 3\", cluster_4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47196e27",
   "metadata": {},
   "source": [
    "Then, I'll add the Genome-2 children."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b5fda0d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6. Make genome-2 clusters\n",
    "cluster_9 = make_haplotype(\"cluster 9\", genome_2)\n",
    "cluster_10 = make_haplotype(\"cluster 10\", genome_2)\n",
    "cluster_11 = make_haplotype(\"cluster 11\", genome_2)\n",
    "cluster_12 = make_haplotype(\"cluster 12\", genome_2)\n",
    "cluster_13 = make_haplotype(\"cluster 13\", genome_2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be2bdf3a",
   "metadata": {},
   "source": [
    "Then, I'll write all of these out to a fasta file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "546c4094",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 7. Save all the termimal nodes in a dictionary\n",
    "haplotype_seqs = {\n",
    "    \"sspe_ancestor\":\"\".join(nt for nt in reference),\n",
    "    \"genome_1\":\"\".join(nt for nt in genome_1),\n",
    "    \"genome_2\":\"\".join(nt for nt in genome_2),\n",
    "    \"genome_1_1\":\"\".join(nt for nt in genome_1_1),\n",
    "    \"cluster_1\":\"\".join(nt for nt in cluster_1),\n",
    "    \"cluster_2\":\"\".join(nt for nt in cluster_2),\n",
    "    \"cluster_3\":\"\".join(nt for nt in cluster_3),\n",
    "    \"cluster_4\":\"\".join(nt for nt in cluster_4),\n",
    "    \"cluster_5\":\"\".join(nt for nt in cluster_5),\n",
    "    \"cluster_6\":\"\".join(nt for nt in cluster_6),\n",
    "    \"cluster_7\":\"\".join(nt for nt in cluster_7),\n",
    "    \"cluster_8\":\"\".join(nt for nt in cluster_8),\n",
    "    \"cluster_9\":\"\".join(nt for nt in cluster_9),\n",
    "    \"cluster_10\":\"\".join(nt for nt in cluster_10),\n",
    "    \"cluster_11\":\"\".join(nt for nt in cluster_11),\n",
    "    \"cluster_12\":\"\".join(nt for nt in cluster_12),\n",
    "    \"cluster_13\":\"\".join(nt for nt in cluster_13)\n",
    "}\n",
    "\n",
    "# 8. Write out to a fasta file\n",
    "haplotype_records = [SeqRecord.SeqRecord(Seq.Seq(seq), id = hap) for hap, seq in haplotype_seqs.items()]\n",
    "SeqIO.write(haplotype_records, os.path.join(outpath, \"haplotype-sequences.fa\"), \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5767103c",
   "metadata": {},
   "source": [
    "## Visualize the Tree"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50ad79b0",
   "metadata": {},
   "source": [
    "Make a tree with the Fasta file of each terminal haplotype. These are alredady aligned by the nature of how I made them. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3a8cb936",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IQ-TREE multicore version 2.1.4-beta COVID-edition for Linux 64-bit built Jun 24 2021\n",
      "Developed by Bui Quang Minh, James Barbetti, Nguyen Lam Tung,\n",
      "Olga Chernomor, Heiko Schmidt, Dominik Schrempf, Michael Woodhams.\n",
      "\n",
      "Host:    rhino02 (AVX512, FMA3, 754 GB RAM)\n",
      "Command: iqtree -s ../../results/phylogeny/haplotype-sequences.fa -m GTR+I+G -bb 1000 -redo\n",
      "Seed:    789293 (Using SPRNG - Scalable Parallel Random Number Generator)\n",
      "Time:    Thu Mar  2 20:40:40 2023\n",
      "Kernel:  AVX+FMA - 1 threads (72 CPU cores detected)\n",
      "\n",
      "HINT: Use -nt option to specify number of threads because your CPU has 72 cores!\n",
      "HINT: -nt AUTO will automatically determine the best number of threads to use.\n",
      "\n",
      "Reading alignment file ../../results/phylogeny/haplotype-sequences.fa ... Fasta format detected\n",
      "Alignment most likely contains DNA/RNA sequences\n",
      "Alignment has 17 sequences with 15894 columns, 69 distinct patterns\n",
      "47 parsimony-informative, 50 singleton sites, 15797 constant sites\n",
      "               Gap/Ambiguity  Composition  p-value\n",
      "   1  sspe_ancestor    0.00%    passed     99.27%\n",
      "   2  genome_1         0.00%    passed    100.00%\n",
      "   3  genome_2         0.00%    passed     99.78%\n",
      "   4  genome_1_1       0.00%    passed     99.84%\n",
      "   5  cluster_1        0.00%    passed     99.73%\n",
      "   6  cluster_2        0.00%    passed     99.80%\n",
      "   7  cluster_3        0.00%    passed     99.76%\n",
      "   8  cluster_4        0.00%    passed     99.89%\n",
      "   9  cluster_5        0.00%    passed     99.89%\n",
      "  10  cluster_6        0.00%    passed     99.99%\n",
      "  11  cluster_7        0.00%    passed     99.74%\n",
      "  12  cluster_8        0.00%    passed     99.80%\n",
      "  13  cluster_9        0.00%    passed     99.78%\n",
      "  14  cluster_10       0.00%    passed     99.62%\n",
      "  15  cluster_11       0.00%    passed     99.71%\n",
      "  16  cluster_12       0.00%    passed     99.84%\n",
      "  17  cluster_13       0.00%    passed     99.76%\n",
      "****  TOTAL            0.00%  0 sequences failed composition chi2 test (p-value<5%; df=3)\n",
      "NOTE: cluster_9 is identical to genome_2 but kept for subsequent analysis\n",
      "\n",
      "Create initial parsimony tree by phylogenetic likelihood library (PLL)... 0.000 seconds\n",
      "Generating 1000 samples for ultrafast bootstrap (seed: 789293)...\n",
      "\n",
      "NOTE: 0 MB RAM (0 GB) is required!\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "Thoroughly optimizing +I+G parameters from 10 start values...\n",
      "Init pinv, alpha: 0.000, 1.000 / Estimate: 0.000, 0.133 / LogL: -22895.760\n",
      "Init pinv, alpha: 0.110, 1.000 / Estimate: 0.110, 0.204 / LogL: -22895.731\n",
      "Init pinv, alpha: 0.221, 1.000 / Estimate: 0.221, 0.295 / LogL: -22895.713\n",
      "Init pinv, alpha: 0.331, 1.000 / Estimate: 0.331, 0.431 / LogL: -22895.707\n",
      "Init pinv, alpha: 0.442, 1.000 / Estimate: 0.442, 0.668 / LogL: -22895.702\n",
      "Init pinv, alpha: 0.552, 1.000 / Estimate: 0.552, 1.210 / LogL: -22895.699\n",
      "Init pinv, alpha: 0.663, 1.000 / Estimate: 0.663, 3.722 / LogL: -22895.695\n",
      "Init pinv, alpha: 0.773, 1.000 / Estimate: 0.773, 998.431 / LogL: -22895.712\n",
      "Init pinv, alpha: 0.883, 1.000 / Estimate: 0.883, 998.444 / LogL: -22896.215\n",
      "Init pinv, alpha: 0.994, 1.000 / Estimate: 0.967, 998.447 / LogL: -22902.602\n",
      "Optimal pinv,alpha: 0.663, 3.722 / LogL: -22895.695\n",
      "\n",
      "Parameters optimization took 0.086 sec\n",
      "Computing ML distances based on estimated model parameters...\n",
      "Computing ML distances took 0.006255 sec (of wall-clock time) 0.006256 sec(of CPU time)\n",
      "WARNING: Some pairwise ML distances are too long (saturated)\n",
      "Computing RapidNJ tree took 0.000216 sec (of wall-clock time) 0.000232 sec (of CPU time)\n",
      "Log-likelihood of RapidNJ tree: -22970.107\n",
      "--------------------------------------------------------------------\n",
      "|             INITIALIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Generating 98 parsimony trees... 0.034 second\n",
      "Computing log-likelihood of 98 initial trees ... 0.036 seconds\n",
      "Current best score: -22895.695\n",
      "\n",
      "Do NNI search on 20 best initial trees\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 1: -22895.693\n",
      "Iteration 10 / LogL: -22895.693 / Time: 0h:0m:0s\n",
      "Iteration 20 / LogL: -22895.693 / Time: 0h:0m:0s\n",
      "Finish initializing candidate tree set (20)\n",
      "Current best tree score: -22895.693 / CPU time: 0.135\n",
      "Number of iterations: 20\n",
      "--------------------------------------------------------------------\n",
      "|               OPTIMIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Iteration 30 / LogL: -22895.770 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 40 / LogL: -22895.796 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 50 / LogL: -22895.730 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -22906.401\n",
      "Iteration 60 / LogL: -22895.693 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 70 / LogL: -22895.787 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 80 / LogL: -22978.930 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 90 / LogL: -22895.693 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Iteration 100 / LogL: -22895.783 / Time: 0h:0m:0s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -22906.401\n",
      "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 0.898\n",
      "NOTE: UFBoot does not converge, continue at least 100 more iterations\n",
      "Iteration 110 / LogL: -22895.732 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "Iteration 120 / LogL: -22895.693 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "Iteration 130 / LogL: -22895.728 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "Iteration 140 / LogL: -22895.740 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "Iteration 150 / LogL: -22895.729 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -22906.633\n",
      "Iteration 160 / LogL: -22895.693 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "Iteration 170 / LogL: -22895.760 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "Iteration 180 / LogL: -22895.693 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "Iteration 190 / LogL: -22895.766 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "Iteration 200 / LogL: -22895.720 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -22906.634\n",
      "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 0.934\n",
      "NOTE: UFBoot does not converge, continue at least 100 more iterations\n",
      "Iteration 210 / LogL: -22895.789 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "Iteration 220 / LogL: -22895.756 / Time: 0h:0m:1s (0h:0m:0s left)\n",
      "Iteration 230 / LogL: -22895.693 / Time: 0h:0m:2s (0h:0m:0s left)\n",
      "Iteration 240 / LogL: -22895.768 / Time: 0h:0m:2s (0h:0m:0s left)\n",
      "Iteration 250 / LogL: -22895.762 / Time: 0h:0m:2s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -22906.517\n",
      "Iteration 260 / LogL: -22895.693 / Time: 0h:0m:2s (0h:0m:0s left)\n",
      "Iteration 270 / LogL: -22895.740 / Time: 0h:0m:2s (0h:0m:0s left)\n",
      "Iteration 280 / LogL: -22895.693 / Time: 0h:0m:2s (0h:0m:0s left)\n",
      "Iteration 290 / LogL: -22895.723 / Time: 0h:0m:2s (0h:0m:0s left)\n",
      "Iteration 300 / LogL: -22895.723 / Time: 0h:0m:2s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -22906.517\n",
      "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 0.979\n",
      "NOTE: UFBoot does not converge, continue at least 100 more iterations\n",
      "Iteration 310 / LogL: -22895.693 / Time: 0h:0m:2s (0h:0m:0s left)\n",
      "Iteration 320 / LogL: -22895.693 / Time: 0h:0m:2s (0h:0m:0s left)\n",
      "Iteration 330 / LogL: -22895.814 / Time: 0h:0m:2s (0h:0m:0s left)\n",
      "Iteration 340 / LogL: -22895.756 / Time: 0h:0m:3s (0h:0m:0s left)\n",
      "Iteration 350 / LogL: -22895.742 / Time: 0h:0m:3s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -22906.517\n",
      "Iteration 360 / LogL: -22978.861 / Time: 0h:0m:3s (0h:0m:0s left)\n",
      "Iteration 370 / LogL: -22895.758 / Time: 0h:0m:3s (0h:0m:0s left)\n",
      "Iteration 380 / LogL: -22895.763 / Time: 0h:0m:3s (0h:0m:0s left)\n",
      "Iteration 390 / LogL: -22895.786 / Time: 0h:0m:3s (0h:0m:0s left)\n",
      "Iteration 400 / LogL: -22895.760 / Time: 0h:0m:3s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -22906.517\n",
      "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 0.964\n",
      "NOTE: UFBoot does not converge, continue at least 100 more iterations\n",
      "Iteration 410 / LogL: -22895.693 / Time: 0h:0m:3s (0h:0m:0s left)\n",
      "Iteration 420 / LogL: -22895.693 / Time: 0h:0m:3s (0h:0m:0s left)\n",
      "Iteration 430 / LogL: -22895.749 / Time: 0h:0m:3s (0h:0m:0s left)\n",
      "Iteration 440 / LogL: -22895.693 / Time: 0h:0m:4s (0h:0m:0s left)\n",
      "Iteration 450 / LogL: -22895.736 / Time: 0h:0m:4s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -22906.696\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Iteration 460 / LogL: -22895.752 / Time: 0h:0m:4s (0h:0m:0s left)\n",
      "Iteration 470 / LogL: -22895.757 / Time: 0h:0m:4s (0h:0m:0s left)\n",
      "Iteration 480 / LogL: -22895.693 / Time: 0h:0m:4s (0h:0m:0s left)\n",
      "Iteration 490 / LogL: -22895.759 / Time: 0h:0m:4s (0h:0m:0s left)\n",
      "Iteration 500 / LogL: -22895.760 / Time: 0h:0m:4s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -22906.696\n",
      "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 0.957\n",
      "NOTE: UFBoot does not converge, continue at least 100 more iterations\n",
      "Iteration 510 / LogL: -22895.693 / Time: 0h:0m:4s (0h:0m:0s left)\n",
      "Iteration 520 / LogL: -22895.750 / Time: 0h:0m:4s (0h:0m:0s left)\n",
      "Iteration 530 / LogL: -22895.789 / Time: 0h:0m:4s (0h:0m:0s left)\n",
      "Iteration 540 / LogL: -22895.794 / Time: 0h:0m:5s (0h:0m:0s left)\n",
      "Iteration 550 / LogL: -22895.693 / Time: 0h:0m:5s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -22906.696\n",
      "Iteration 560 / LogL: -22895.794 / Time: 0h:0m:5s (0h:0m:0s left)\n",
      "Iteration 570 / LogL: -22895.693 / Time: 0h:0m:5s (0h:0m:0s left)\n",
      "Iteration 580 / LogL: -22895.736 / Time: 0h:0m:5s (0h:0m:0s left)\n",
      "Iteration 590 / LogL: -22895.776 / Time: 0h:0m:5s (0h:0m:0s left)\n",
      "Iteration 600 / LogL: -22895.723 / Time: 0h:0m:5s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -22906.696\n",
      "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 0.968\n",
      "NOTE: UFBoot does not converge, continue at least 100 more iterations\n",
      "Iteration 610 / LogL: -22895.723 / Time: 0h:0m:5s (0h:0m:0s left)\n",
      "Iteration 620 / LogL: -22895.776 / Time: 0h:0m:5s (0h:0m:0s left)\n",
      "Iteration 630 / LogL: -22895.749 / Time: 0h:0m:6s (0h:0m:0s left)\n",
      "Iteration 640 / LogL: -22895.751 / Time: 0h:0m:6s (0h:0m:0s left)\n",
      "Iteration 650 / LogL: -22895.800 / Time: 0h:0m:6s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -22906.719\n",
      "Iteration 660 / LogL: -22895.747 / Time: 0h:0m:6s (0h:0m:0s left)\n",
      "Iteration 670 / LogL: -22895.693 / Time: 0h:0m:6s (0h:0m:0s left)\n",
      "Iteration 680 / LogL: -22895.693 / Time: 0h:0m:6s (0h:0m:0s left)\n",
      "Iteration 690 / LogL: -22895.756 / Time: 0h:0m:6s (0h:0m:0s left)\n",
      "Iteration 700 / LogL: -22895.756 / Time: 0h:0m:6s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -22906.719\n",
      "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 0.981\n",
      "NOTE: UFBoot does not converge, continue at least 100 more iterations\n",
      "Iteration 710 / LogL: -22895.756 / Time: 0h:0m:7s (0h:0m:0s left)\n",
      "Iteration 720 / LogL: -22895.768 / Time: 0h:0m:7s (0h:0m:0s left)\n",
      "Iteration 730 / LogL: -22895.693 / Time: 0h:0m:7s (0h:0m:0s left)\n",
      "Iteration 740 / LogL: -22895.730 / Time: 0h:0m:7s (0h:0m:0s left)\n",
      "Iteration 750 / LogL: -22895.749 / Time: 0h:0m:7s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -22906.719\n",
      "Iteration 760 / LogL: -22895.693 / Time: 0h:0m:7s (0h:0m:0s left)\n",
      "Iteration 770 / LogL: -22895.693 / Time: 0h:0m:7s (0h:0m:0s left)\n",
      "Iteration 780 / LogL: -22895.693 / Time: 0h:0m:7s (0h:0m:0s left)\n",
      "Iteration 790 / LogL: -22895.749 / Time: 0h:0m:7s (0h:0m:0s left)\n",
      "Iteration 800 / LogL: -22895.781 / Time: 0h:0m:8s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -22906.719\n",
      "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 0.978\n",
      "NOTE: UFBoot does not converge, continue at least 100 more iterations\n",
      "Iteration 810 / LogL: -22895.823 / Time: 0h:0m:8s (0h:0m:0s left)\n",
      "Iteration 820 / LogL: -22895.732 / Time: 0h:0m:8s (0h:0m:0s left)\n",
      "Iteration 830 / LogL: -22895.751 / Time: 0h:0m:8s (0h:0m:0s left)\n",
      "Iteration 840 / LogL: -22895.745 / Time: 0h:0m:8s (0h:0m:0s left)\n",
      "Iteration 850 / LogL: -22895.693 / Time: 0h:0m:8s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -22906.875\n",
      "Iteration 860 / LogL: -22895.784 / Time: 0h:0m:8s (0h:0m:0s left)\n",
      "Iteration 870 / LogL: -22895.784 / Time: 0h:0m:9s (0h:0m:0s left)\n",
      "Iteration 880 / LogL: -22895.732 / Time: 0h:0m:9s (0h:0m:0s left)\n",
      "Iteration 890 / LogL: -22895.723 / Time: 0h:0m:9s (0h:0m:0s left)\n",
      "Iteration 900 / LogL: -22895.729 / Time: 0h:0m:9s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -22906.875\n",
      "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 0.973\n",
      "NOTE: UFBoot does not converge, continue at least 100 more iterations\n",
      "Iteration 910 / LogL: -22954.813 / Time: 0h:0m:9s (0h:0m:0s left)\n",
      "Iteration 920 / LogL: -22895.777 / Time: 0h:0m:9s (0h:0m:0s left)\n",
      "Iteration 930 / LogL: -22895.736 / Time: 0h:0m:9s (0h:0m:0s left)\n",
      "Iteration 940 / LogL: -22895.745 / Time: 0h:0m:9s (0h:0m:0s left)\n",
      "Iteration 950 / LogL: -22895.787 / Time: 0h:0m:10s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -22906.875\n",
      "Iteration 960 / LogL: -22895.748 / Time: 0h:0m:10s (0h:0m:0s left)\n",
      "Iteration 970 / LogL: -22895.723 / Time: 0h:0m:10s (0h:0m:0s left)\n",
      "Iteration 980 / LogL: -22895.693 / Time: 0h:0m:10s (0h:0m:0s left)\n",
      "Iteration 990 / LogL: -22895.693 / Time: 0h:0m:10s (0h:0m:0s left)\n",
      "Iteration 1000 / LogL: -22895.797 / Time: 0h:0m:10s (0h:0m:0s left)\n",
      "Log-likelihood cutoff on original alignment: -22906.875\n",
      "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 0.981\n",
      "NOTE: UFBoot does not converge, continue at least 100 more iterations\n",
      "WARNING: bootstrap analysis did not converge. You should rerun with higher number of iterations (-nm option)\n",
      "TREE SEARCH COMPLETED AFTER 1001 ITERATIONS / Time: 0h:0m:10s\n",
      "\n",
      "--------------------------------------------------------------------\n",
      "|                    FINALIZING TREE SEARCH                        |\n",
      "--------------------------------------------------------------------\n",
      "Performs final model parameters optimization\n",
      "Estimate model parameters (epsilon = 0.010)\n",
      "1. Initial log-likelihood: -22895.693\n",
      "Optimal log-likelihood: -22895.690\n",
      "Rate parameters:  A-C: 2.57807  A-G: 1.12159  A-T: 0.95501  C-G: 1.47915  C-T: 3.00643  G-T: 1.00000\n",
      "Base frequencies:  A: 0.291  C: 0.241  G: 0.235  T: 0.233\n",
      "Proportion of invariable sites: 0.663\n",
      "Gamma shape alpha: 3.734\n",
      "Parameters optimization took 1 rounds (0.002 sec)\n",
      "BEST SCORE FOUND : -22895.690\n",
      "Creating bootstrap support values...\n",
      "Split supports printed to NEXUS file ../../results/phylogeny/haplotype-sequences.fa.splits.nex\n",
      "Total tree length: 0.006\n",
      "\n",
      "Total number of iterations: 1001\n",
      "CPU time used for tree search: 10.380 sec (0h:0m:10s)\n",
      "Wall-clock time used for tree search: 10.643 sec (0h:0m:10s)\n",
      "Total CPU time used: 10.520 sec (0h:0m:10s)\n",
      "Total wall-clock time used: 10.804 sec (0h:0m:10s)\n",
      "\n",
      "Computing bootstrap consensus tree...\n",
      "Reading input file ../../results/phylogeny/haplotype-sequences.fa.splits.nex...\n",
      "17 taxa and 205 splits.\n",
      "Consensus tree written to ../../results/phylogeny/haplotype-sequences.fa.contree\n",
      "Reading input trees file ../../results/phylogeny/haplotype-sequences.fa.contree\n",
      "Log-likelihood of consensus tree: -22896.182\n",
      "\n",
      "Analysis results written to: \n",
      "  IQ-TREE report:                ../../results/phylogeny/haplotype-sequences.fa.iqtree\n",
      "  Maximum-likelihood tree:       ../../results/phylogeny/haplotype-sequences.fa.treefile\n",
      "  Likelihood distances:          ../../results/phylogeny/haplotype-sequences.fa.mldist\n",
      "\n",
      "Ultrafast bootstrap approximation results written to:\n",
      "  Split support values:          ../../results/phylogeny/haplotype-sequences.fa.splits.nex\n",
      "  Consensus tree:                ../../results/phylogeny/haplotype-sequences.fa.contree\n",
      "  Screen log file:               ../../results/phylogeny/haplotype-sequences.fa.log\n",
      "\n",
      "Date and Time: Thu Mar  2 20:40:51 2023\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Building the phylogeny with IQtree with 1000 bootstrap iterations with GTR+I+G (Invariable site plus discrete Gamma model) with asr\n",
    "alignfasta = os.path.join(outpath, \"haplotype-sequences.fa\")\n",
    "call(f\"iqtree -s {alignfasta} -m GTR+I+G -bb 1000 -redo\", shell=True)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
