---
title: "10. Expand Existing Haplotypes"
author:
    - "Will Hannon"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
  highlight: tango
  number_sections: no
  theme: default
  toc: yes
  toc_depth: 3
  toc_float:
    collapsed: no
    smooth_scroll: yes
---

The goal of this notebook is to go through all of the mutations above 5% frequency in any tissue to figure out how these are related by combining the previous correlation analyses with bridging reads. 

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```

```{r Required Packages, message=FALSE, warning=FALSE, echo=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed
packages = c("tidyverse", "data.table")

## Check that packages are installed, if not, install them
# installed_packages <- packages %in% rownames(installed.packages())
# if (any(installed_packages == FALSE)) {
#   install.packages(packages[!installed_packages])
# }

## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```

```{r Inputs, echo=T}

## ==== File paths input data ==== ##

# Data from lofreq and varscan labeled with subclonal haplotypes
updated.data = "../../results/variants/assigned_variants.csv"

# Data from the count of bridging reads
bridging.data = "../../results/bridging/bridging_reads.csv"

# Depth Data 
depth.data = "../../results/coverage/merged.depth"

# Annotations 
annotations.filepath = "../../config/annotations.csv"

```

```{r Import Data, echo=T, warning=F, message=F}

# Assigned SNPs with haplotype labels 
updated.df = read_csv(updated.data, show_col_types = FALSE)

haplotypes.label = updated.df %>% 
  select(SNP, Haplotype, Background, Gene_Name, AA_Change, Effect, POS, ALT, REF) %>% 
  distinct()

# Depth at each position
depth.df = fread(depth.data) %>% 
  select(!V1) %>% 
  left_join(., select(updated.df, Tissue, Accession)) %>% 
  select(!Accession)
  
# Expand the mutations to have a frequency for every tissue.
expanded.updated.df = updated.df %>% 
  select(SNP, Tissue, AF) %>% 
  pivot_wider(names_from = "Tissue", values_from = "AF", values_fill = 0) %>% 
  pivot_longer(cols = !SNP, names_to = "Tissue", values_to = "AF") %>% 
  left_join(., haplotypes.label, by = "SNP") %>% 
  left_join(., depth.df, by = c("POS", "Tissue")) %>% 
  distinct()


```

```{r Count and Frequency}

# Cut the haplotypes into different categories based on their maximum frequency in the brain
count.df = data.frame()
for (min_freq in seq(.05, .9, .05)) {
  
  count = expanded.updated.df %>% 
    filter(AF >= min_freq) %>% 
    group_by(SNP) %>% 
    count() %>% 
    mutate(min_freq = min_freq)
    
  count.df = rbind(count.df, count)
  
}

annotated_snp.df = expanded.updated.df %>% 
  mutate(Effect = if_else(is.na(Effect), "Synonymous", Effect)) %>% 
  mutate(ADAR = case_when(
    REF == 'T' & ALT == 'C' ~ TRUE, 
    REF == 'A' & ALT == 'G' ~ TRUE,
    TRUE ~ FALSE
  )) %>% 
  left_join(., count.df, by = "SNP") 

```

```{r fig.align='center', fig.width=15, fig.height=8}

final.df %>% 
  filter(n >= 1,
         min_freq >= 0.1,
         Haplotype == "subclonal"
         ) %>% 
    ggplot(aes(x = Tissue, y = AF)) +
      geom_line(aes(group = SNP)) +
      facet_wrap(~SNP) +
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))
```

```{r}

final.df %>% 
  filter(SNP %in% c("G3579A","G5961A","G6123A","T3761C") | Haplotype == "cluster 8")%>% 
    ggplot(aes(x = Tissue, y = AF, col=SNP)) +
      geom_line(aes(group = SNP)) +
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12))

cluster.15 = c("G3579A","G5961A","G6123A","T3761C")

```

```{r}

final.df %>% 
  filter(Haplotype != "subclonal") %>% 
  filter(Gene_Name %in% c("H", "F")) %>% 
  select(SNP, POS, Haplotype, Gene_Name, Tissue) %>% 
  distinct() %>% 
  view()

```

```{r}

final.df %>% 
  filter(Gene_Name == "M") %>% 
  filter(Haplotype != "subclonal") %>% 
  filter(Tissue == "Hippocampus") %>% 
  select(Haplotype, SNP, AF) %>% 
  distinct() %>% 
  group_by(Haplotype) %>% 
  summarize(mean_af = mean(AF)) %>% 
  arrange(-mean_af)
  
  
  
```

Every mutation above 10% frequency that's subclonal. 

```{r}

unclustered.snps = expanded.updated.df %>% 
  filter(Haplotype == "subclonal") %>% 
  filter(AF >= .1) %>% 
  pull(SNP) %>% 
  unique()


expanded.updated.df %>% 
  filter(SNP %in% unclustered.snps) %>% 
  ggplot(aes(x = Tissue, y = AF, group=SNP)) + 
    geom_line()


```


