---
title: "5. Process Bridging Reads"
author: "Alison Feder"
output: 
  html_document:
  highlight: tango
  number_sections: no
  theme: default
  toc: yes
  toc_depth: 3
  toc_float:
    collapsed: no
    smooth_scroll: yes
---

The goal of this notebook is to establish a method to genotype SNPs as either belonging to `genome-1` or to `genome-2`. Additionaly, using the same method, we will try to link known haplotypes to one another if there are reads that bridge sites between the haplotypes.

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```

```{r Required Packages, message=FALSE, warning=FALSE, echo=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed

packages = c("tidyverse", "foreach", "emdbook")
## Check that packages are installed, if not, install them
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}
## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```

```{r Inputs, echo=T}

## ==== File paths input data ==== ##

# Data from lofreq and varscan labeled with subclonal haplotypes
updated.data = "../../results/variants/clustered_variants.csv"

# Counts of SNPs on reads belonging to genome-1 or genome-2
reads.data = "../../results/split/genotyped.csv"

# Annotations for MeV genes
annotations.filepath = "../../config/annotations.csv"
  
```

```{r Import and Format Data, echo = T}

# Download the variants
haplotypes = read_csv(updated.data, show_col_types = FALSE) %>% 
  select(Haplotype, SNP) %>% 
  distinct()

# Download the counts of SNPs on genome-1 and genome-2 
genotyping.df = read_csv(reads.data, show_col_types = FALSE) %>% 
  select(!AF) %>% 
  rename(total = DP, 
         var = OBSV) %>% 
  mutate(genotype = if_else(genotype == "genome-1", "g1", "g2")) %>% 
  pivot_wider(names_from = genotype,
              values_from = c(total, var), 
              names_sep = "_", 
              values_fill = 0) %>% 
  left_join(., haplotypes, by = "SNP") 

# Haplotyped SNPs for linking subclonal SNPs
updated.df = read_csv(updated.data, show_col_types = FALSE)

```

```{r Assign Genotypes for each Tau, echo=T}

# Log-spaced sequence of taus
taus = lseq(0.00001, 1, length.out = 200)

# Minimum significance
min.sig = 0.05

# Minimum number of variant observations? 

# Assign to genome for each value of tau
assignmentsByTau = foreach(tau = taus, .combine = "rbind")%do%{

    allSNPs = genotyping.df %>% 
        gather(cat, count, -POS, -REF, -ALT, -SNP, -Tissue, -Haplotype) %>% 
        separate(cat, into = c("cat", "G"), sep = "_") %>% 
        spread(cat, count) %>% 
        group_by(SNP, G) %>% 
        mutate(linked= pbinom( var, total, tau, lower.tail = FALSE), 
               sig = linked < min.sig/n()) %>% 
        mutate(sig = ifelse(total == 0, NA, sig)) %>% 
        summarize(TissueCount = sum(sig, na.rm = TRUE), .groups = "drop") %>% 
        spread(G, TissueCount) %>% 
        mutate(G1orG2 = ifelse(g1 > 0 & g2 == 0, "G1",
                        ifelse(g2 > 0 & g1 == 0, "G2", 
                        ifelse(g1 > 0 & g2 > 0, "G1/G2", 
                        ifelse(g1 == 0 & g2 == 0, "-", "*"))))) %>% 
        ungroup()


    left_join(genotyping.df, allSNPs, by = "SNP") %>% select( -total_g1, -total_g2, -var_g1, -var_g2) %>% mutate(tau = tau)

}


```

### G1/G2 assignment for Haplotypes

What are the `genome1/2` assignments for each identified haplotype for every value of `tau` between `0.00001` and `1`? 
```{r Assignments for all Haplotypes, warning=F, message=F, fig.align='center', fig.width=18, fig.height=14}

assignment.order = c('G1', 'G2', 'G1/G2', '-')
assignment.colors = c("#307EC2", "#B63F3E", "purple", "darkgrey")
  
assignmentsByTau %>%
  select(-Tissue) %>% 
  unique() %>%
  group_by(G1orG2, Haplotype, tau) %>%
  summarize(n = n()) %>% 
  spread(G1orG2, n) %>%
  gather(type, count, -Haplotype, -tau) %>% 
  mutate(count = ifelse(is.na(count), 0, count)) %>%
  spread(type, count) %>% 
  gather(type, count, -Haplotype, -tau) %>%
  group_by(Haplotype, tau) %>%
  mutate(total = sum(count)) %>% 
  mutate(prop = count/total) %>% 
    ggplot(aes(x = tau, y = prop, col = factor(type, level = assignment.order), group = type)) +
      geom_path(size = 1.5) + 
      scale_x_log10() +
      facet_wrap(~Haplotype) +
      labs(y = "Proportion of SNPs assigned") +  
      scale_color_manual(values = assignment.colors, name = "Genome Assignment") +
      theme_classic() +
      theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
      theme(legend.position="bottom", legend.box = "horizontal") +
      theme(legend.box.background = element_rect(colour = "black")) +
      theme(text=element_text(size=18)) + 
      theme(plot.title = element_text(hjust = 0.5)) 


```

### Subclonal SNPs? 

How does this look for subclonal SNPs? Can we assign most to either `genome-1` or `genome-2`? 

```{r Annotations, message=FALSE, warning=FALSE}

# Import genome annotations
annotations.df = read_csv(annotations.filepath, show_col_types = FALSE)

# Get the interval positions for these genes 
N = annotations.df %>% dplyr::filter(Locus == "N")
P.V.C = annotations.df %>% dplyr::filter(`Protein Name` == "phosphoprotein")
M = annotations.df %>% dplyr::filter(Locus == "M")
F. = annotations.df %>% dplyr::filter(Locus == "F")
H = annotations.df %>% dplyr::filter(Locus == "H")
L = annotations.df %>% dplyr::filter(Locus == "L")

```


```{r Subclonal Assignments, warning=F, message=F, echo = T}

subclonal.assignments = assignmentsByTau %>%
  filter(Haplotype == "subclonal") %>%
  select(-Tissue) %>% 
  unique() %>%
  group_by(G1orG2, SNP, tau) %>%
  summarize(n = n()) %>% 
  spread(G1orG2, n) %>% 
  gather(type, count, -SNP, -tau) %>% 
  mutate(count = ifelse(is.na(count), 0, count)) %>%
  spread(type, count) %>% 
  gather(type, count, -SNP, -tau) %>% group_by(SNP, tau) %>%
  mutate(total = sum(count)) %>% 
  mutate(prop = count/total) %>%
  group_by(SNP) %>%
  filter(type == "G1" | type == "G2") %>% 
  filter(prop > 0) %>%
  select(SNP, type) %>% 
  unique()

```

There isn't much of a difference between the assignment of SNPs to `genome-1` or `genome-2` across the MeV genome. 

```{r Subclonal Assignments Positions, warning=F, message=F, fig.align='center', fig.width=18, fig.height=8}

# Import genome annotations
annotations.df = read_csv(annotations.filepath, show_col_types = FALSE)

# Get the interval positions for these genes 
N = annotations.df %>% dplyr::filter(Locus == "N")
P.V.C = annotations.df %>% dplyr::filter(`Protein Name` == "phosphoprotein")
M = annotations.df %>% dplyr::filter(Locus == "M")
F. = annotations.df %>% dplyr::filter(Locus == "F")
H = annotations.df %>% dplyr::filter(Locus == "H")
L = annotations.df %>% dplyr::filter(Locus == "L")

left_join(subclonal.assignments, genotyping.df) %>%
  select(SNP, type, POS) %>%
  unique() %>% 
  ggplot(aes(x = POS, y = type, col = type)) +
    geom_point(pch = "|", size = 3) +
    labs(x = "Position", y = "Genome")+
    scale_color_manual(values = assignment.colors, name = "Genome Assignment") +
  
    # == Annotations of genes == #
  
    # Nucleoprotein
    annotate("rect", xmin = N$Start-1, xmax = N$Stop, ymin = 0, ymax = .1 ,
               alpha = .2, fill = "dodgerblue2", col = "black", size = .1) +
    annotate(geom = "text", x = (N$Stop + N$Start)/2, y = .05, label = "N", size = 8) +
    annotate("rect", xmin = N$Start-1, xmax = N$Stop, ymin = 0, ymax = 2.5 ,
                   alpha = .1, fill = "dodgerblue2", size = .1) +  
    # P/V/C
    annotate("rect", xmin = P.V.C$Start, xmax = P.V.C$Stop, ymin = 0, ymax = .1 ,
               alpha = .2, fill = "yellow3", col = "black", size = .1) +
    annotate(geom = "text", x = (P.V.C$Stop + P.V.C$Start)/2, y = .05, label = "P/V/C", size = 8) +
    annotate("rect", xmin = P.V.C$Start, xmax = P.V.C$Stop, ymin = 0, ymax = 2.5 ,
               alpha = .1, fill = "yellow3", size = .1) +
    # Matrix protein
    annotate("rect", xmin = M$Start, xmax = M$Stop, ymin = 0, ymax = .1 ,
               alpha = .2, fill = "#FF7F00", col = "black", size = .1) +
    annotate(geom = "text", x = (M$Stop + M$Start)/2, y = .05, label = "M", size = 8)  +
    annotate("rect", xmin = M$Start, xmax = M$Stop, ymin = 0, ymax = 2.5 ,
               alpha = .1, fill = "#FF7F00", size = .1) +
    # Fusion protein
    annotate("rect", xmin = F.$Start, xmax = F.$Stop, ymin = 0, ymax = .1 ,
               alpha = .2, fill = "#6A3D9A", col = "black", size = .1) +
    annotate(geom = "text", x = (F.$Stop + F.$Start)/2, y = .05, label = "F", size = 8) +
    annotate("rect", xmin = F.$Start, xmax = F.$Stop, ymin = 0, ymax = 2.5 ,
               alpha = .1, fill = "#6A3D9A", size = .1) +
    # Hemagluttinin
    annotate("rect", xmin = H$Start, xmax = H$Stop, ymin = 0, ymax = .1 ,
               alpha = .2, fill = "green4", col = "black", size = .1) +
    annotate(geom = "text", x = (H$Stop + H$Start)/2, y = .05, label = "H", size = 8) +
    annotate("rect", xmin = H$Start, xmax = H$Stop, ymin = 0, ymax = 2.5,
               alpha = .1, fill = "green4", size = .1) +
    # Large protein
    annotate("rect", xmin = L$Start, xmax = L$Stop, ymin = 0, ymax = .1 ,
               alpha = .2, fill = "#E31A1C", col = "black", size = .1) +
    annotate(geom = "text", x = (L$Stop + L$Start)/2, y = .05, label = "L", size = 8) +
    annotate("rect", xmin = L$Start, xmax = L$Stop, ymin = 0, ymax = 2.5 ,
               alpha = .1, fill = "#E31A1C", size = .1) +
    theme_classic() +
    theme(panel.grid.major.y = element_line(colour="grey", linetype="dashed")) +
    theme(legend.position="bottom", legend.box = "horizontal") +
    theme(legend.box.background = element_rect(colour = "black")) +
    theme(text=element_text(size=24)) + 
    theme(axis.text.x = element_text(size=28)) + 
    theme(axis.title.x = element_text(size=30)) + 
    theme(plot.title = element_text(hjust = 0.5))  +
    theme(axis.title.y = element_blank()) +
    scale_x_continuous(limits=c(0,16000)) +
    theme(legend.position = "none") 


```


## Processing Pairs of SNPs

We also have counts of reads for all observed pairs of haplotypes for each tissue. The idea is to use these bridging reads to figure out how certain haplotypes are connected. 

```{r}

snp.pairs.df = read_csv("../../config/snp_pairs.csv", show_col_types = FALSE)

head(snp.pairs.df)

```

The columns are as follows: `snp_1` is the position of one SNP, `snp_2` is the position of the second SNP, `00` are reads that overlap both positions but have neither SNP, `10` are reads that have only `snp_1`, `01` are reads that have only `snp_2`, and `11` are reads with both SNPs. 


```{r Calculate Linked and Forbidden SNPs, echo = T}

tau = 0.02 # Tau, minimum haplotype frequency
min_obsv = 200 # Minimum number of bridging reads (of any type)
min.sig = 0.05 # Minimum significance before multiple testing correction
  
linked.and.forbidden.df = snp.pairs.df %>% 
  mutate(total = `00` + `10` + `01` + `11`) %>% 
  # Multiple testing done by tissue, not polymorphic position
  group_by(snp_1, snp_2) %>% 
  # Calculate the probability of being linked
  mutate(linked = pbinom(`11`, total, tau, lower.tail = F), 
         linked_sig = linked < min.sig/n()) %>% 
    # Calculate the probability of being forbidden
  mutate(forbidden = pbinom(`11`, total, tau, lower.tail = T), 
         forbidden_sig = forbidden < min.sig/n()) %>% 
  mutate(linked_sig = ifelse(total == 0, NA, linked_sig),
         forbidden_sig = ifelse(total == 0, NA, forbidden_sig)) %>% 
  filter(total > min_obsv) %>% 
  ungroup()

```

To see if it's possible to link clusters with read support, I'll get all of the bridging reads that overlap polymorphic positions from two distinct clusters or haplotypes. Then I can use the resulting data to uncover any evidence that clusters are joined. 

```{r Get Linked SNPs for all Clusters}

# All combinations of cluster excluding: g1/g2/both/fixed/subclonal (not clustered)
clusters = updated.df %>% 
  filter(!Haplotype %in% c("genome-1", "genome-2", "both", "fixed", "subclonal")) %>% 
  pull(Haplotype) %>% 
  unique() %>% 
  combn(., 2)

# For each combination of clusters, get the SNPs covered by bridging reads
comparison.df = data.frame()
for (i in 1:ncol(clusters)) {
  
  # Get the clusters 
  cluster.a = clusters[1, i] 
  cluster.b = clusters[2, i]
  
  cluster.a.snps = updated.df %>% 
    filter(Haplotype == cluster.a) %>% 
    pull(POS) %>% 
    unique()
  
  cluster.b.snps = updated.df %>% 
    filter(Haplotype == cluster.b) %>% 
    pull(POS) %>% 
    unique()
  
  cluster.a.df = linked.and.forbidden.df %>% 
    filter(snp_1 %in% cluster.a.snps | snp_2 %in% cluster.a.snps) 
  
  if (dim(cluster.a.df)[1] == 0) next
    
  cluster.b.df = cluster.a.df %>% 
    filter(snp_1 %in% cluster.b.snps | snp_2 %in% cluster.b.snps) 
  
  if (dim(cluster.b.df)[1] == 0) next
  
  overlap.df = cluster.b.df %>% 
    mutate(haplotype_1 = cluster.a,
           haplotype_2 = cluster.b)
    
  comparison.df = rbind(comparison.df, overlap.df)
  
}


```


```{r Percent linked and forbidden, fig.align="center", fig.width=20, fig.height=20}

comparison.df %>% 
  group_by(haplotype_1, haplotype_2, Tissue) %>% 
  mutate(linked_perc = (sum(linked_sig) / n()) * 100,
         forbidden_perc = (sum(forbidden_sig) / n()) * 100) %>% 
  select(Tissue, haplotype_1, haplotype_2, linked_perc, forbidden_perc) %>% 
  pivot_longer(cols = c(linked_perc, forbidden_perc), names_to = "Relation", values_to = "Percent") %>% 
  distinct() %>% 
  ggplot(aes(x = Tissue, y = Percent, fill = Relation)) + 
    geom_bar(stat = "identity", position = position_dodge(), size = 1.5) + 
    facet_grid(cols = vars(haplotype_1), rows = vars(haplotype_2)) +
    theme_bw()

```


```{r}
comparison.df %>% 
  filter(haplotype_1 == "cluster 11" | haplotype_2 == "cluster 11") %>% 
  filter(haplotype_1 == "cluster 1" | haplotype_2 == "cluster 1")
```
```{r}
comparison.df %>% 
  filter(haplotype_1 == "cluster 12" | haplotype_2 == "cluster 12") %>% 
  filter(haplotype_1 == "cluster 1" | haplotype_2 == "cluster 1")
```
```{r}

cluster.1 = "genome-1-1"
cluster.2 = "cluster 9"

comparison.df %>% 
  filter(haplotype_1 == cluster.1 | haplotype_2 == cluster.1) %>% 
  filter(haplotype_1 == cluster.2 | haplotype_2 == cluster.2) %>% 
  arrange(-linked_sig)


updated.df
  
  
```

It looks like clusters 11 and 12 are not on the background of cluster 1. It seems pretty clear that these are typically forbidden. 

There is another odd feature of these results. It looks like some of the SNPs that are part of cluster 9 are acrtually linked to genome-1-1. This shouldn't be possible - I need to explore this further. 


```{r}

# Tissue order ~ relative to position in the brain. 
tissue_order = c( "Frontal Cortex 1", 
                  "Frontal Cortex 3", 
                  "Frontal Cortex 2", 
                  "Temporal Lobe", 
                  "Parietal Lobe",
                  "Occipital Lobe",
                  "Hippocampus",
                  "Internal Capsule", 
                  "Cerebellum",
                  "Cerebellum Nucleus",
                  "Midbrain", 
                  "UBS",
                  "Brain Stem")

haplotypes.label = updated.df %>% 
  select(SNP, Haplotype, Background) %>% 
  distinct()

# Expand the mutations to have a frequency for every tissue.
expanded.df = updated.df %>% 
  select(SNP, Tissue, AF) %>% 
  pivot_wider(names_from = "Tissue", values_from = "AF", values_fill = 0) %>% 
  pivot_longer(cols = !SNP, names_to = "Tissue", values_to = "AF") %>% 
  left_join(., select(updated.df, c("SNP", "Tissue", "DP")), by = c("SNP", "Tissue")) %>% 
  mutate(DP = if_else(is.na(DP), 0, DP)) %>% 
  left_join(., haplotypes.label, by = "SNP") 

# Get the mean frequency of the major genomes 
tissue.mean = updated.df %>%  
  filter(Haplotype %in% c("genome-1", "genome-2")) %>% 
  group_by(Tissue, Haplotype) %>% 
  summarize(AF.mean = mean(AF, na.rm = TRUE), 
            SD = sd(AF, na.rm = TRUE),
            N = n()) %>% 
  mutate(SE = SD / sqrt(N),
         Lower.CI = qt(1 - (0.05 / 2), N - 1) * SE,
         Upper.CI = qt(1 - (0.05 / 2), N - 1) * SE) %>% 
  rename("AF" = AF.mean, "Genotype" = Haplotype)

```


```{r fig.align="center", fig.width=14, fig.height=8}

updated.df %>% 
  filter(POS %in% c(3550)) %>%
  filter(Haplotype %in% ('cluster 9')) %>% 
    ggplot(aes(x = factor(Tissue, levels = tissue_order), y = AF)) +
      geom_line(aes(group = SNP, col = Haplotype, shape = Background), size = .5) + 
      geom_ribbon(data = tissue.mean, aes(x=factor(Tissue, levels = tissue_order), ymin=AF-SD, ymax=AF+SD, group = Genotype, fill = Genotype), alpha=0.2, colour = NA) +
      scale_fill_manual(values=c("#424ef5", "#cf1919")) + 
      xlab("Tissue") + 
      ylab("Allele Frequency") +
      labs(fill="Genotype", col = "Cluster") +
      theme_bw(20) +
      theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
      theme(strip.text.x = element_text(size = 12)) 


```



