---
title: "8. Tissue Level Diversity"
author: "Will Hannon"
output: 
  html_document:
  highlight: tango
  number_sections: no
  theme: default
  toc: yes
  toc_depth: 3
  toc_float:
    collapsed: no
    smooth_scroll: yes
---


The goal of this notebook is to see how the within-host nucleotide diversity compares between tissue samples. Perhaps this could tell us how long Measles has been diversifying in a given tissue compartment. 

```{r Setup, include=FALSE}
require("knitr")
knitr::opts_chunk$set(echo = FALSE)
```

```{r Required Packages, message=FALSE, warning=FALSE, echo=FALSE}

## ==== Install Required Packages ==== ##

## List of all packages needed
packages = c("tidyverse", "ggrepel")
## Check that packages are installed, if not, install them
installed_packages <- packages %in% rownames(installed.packages())
if (any(installed_packages == FALSE)) {
  install.packages(packages[!installed_packages])
}
## Packages loading
invisible(lapply(c(packages), library, character.only = TRUE))

```

```{r Inputs, echo=T}

## ==== File paths input data ==== ##

# Data from lofreq and varscan labeled with G1 and G2 mutations and subclonal haplotypes
updated.data = "../../results/variants/clustered_variants.csv"

# Read in the variants with G1, G2, G1-1, and subclonal haplotypes labeled
updated.df = read_csv(updated.data, show_col_types = FALSE) 

# Tissue order ~ relative to position in the brain. 
tissue.order = c("Frontal Cortex 1", 
  "Frontal Cortex 3", 
  "Frontal Cortex 2", 
  "Temporal Lobe", 
  "Parietal Lobe",
  "Occipital Lobe",
  "Hippocampus",
  "Internal Capsule", 
  "Cerebellum",
  "Cerebellum Nucleus",
  "Midbrain", 
  "UBS",
  "Brain Stem")

```

To calculate nucleotide diversity, I'm referencing the formula from [this paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4316684/). 

Assuming that all sites are biallelic, we can calculate the diversity at a position $i$ as 

$$D_i = \frac{(REF * ALT)}{(DP^2 - DP) / 2}$$

Then, we can sum this for each polymorphic site $s$ to get the nucleotide diversity for a single tissue $\pi$. 

$$\sum_{i=1}^s D_i / L$$
```{r Calculate Nucleotide Diversity Everything, echo=T}

# Length of the measles genome. 
L = 15894

# Calculate the nucleotide diversity with every SNP included. 
pi.with.everything = updated.df %>% 
  select(POS, REF, ALT, AF, DP, Tissue) %>% 
  mutate(alt_count = round(AF * DP),
         var_count = round(DP - alt_count),
         total = alt_count + var_count,
         diversity = (alt_count * var_count) / ( (total^2 - total) / 2) 
         ) %>% 
  group_by(Tissue) %>% 
  summarize(pi = sum(diversity) / L)

# Calculate the nucleotide diversity without genome-1, genome-1-1, or genome-2
pi.without.g1.g2 = updated.df %>% 
  filter(!Haplotype %in% c("genome-1", "genome-1-1", "genome-2")) %>% 
  select(POS, REF, ALT, AF, DP, Tissue) %>% 
  mutate(alt_count = round(AF * DP),
         var_count = round(DP - alt_count),
         total = alt_count + var_count,
         diversity = (alt_count * var_count) / ( (total^2 - total) / 2) 
         ) %>% 
  group_by(Tissue) %>% 
  summarize(pi = sum(diversity) / L)

# Calculate the nucleotide diversity with only the subclonal SNPs included
pi.with.subclonal = updated.df %>% 
  filter(Haplotype == "subclonal") %>% 
  select(POS, REF, ALT, AF, DP, Tissue) %>% 
  mutate(alt_count = round(AF * DP),
         var_count = round(DP - alt_count),
         total = alt_count + var_count,
         diversity = (alt_count * var_count) / ( (total^2 - total) / 2) 
         ) %>% 
  group_by(Tissue) %>% 
  summarize(pi = sum(diversity) / L)

# Join the dataframes into one 
pi.df = rbind(
  mutate(pi.with.everything, Type = "everything"),
  mutate(pi.without.g1.g2, Type = "no g1 or g2"),
  mutate(pi.with.subclonal, Type = "only subclonal")
)

```


How does the nucleotide diversity look for each tissue for each condition? Nothing in particular stands out from this bar chart. There really aren't tremendous differences in the nucleotide diversity ($\pi$) between the different tissue compartments. 

```{r fig.align='center', fig.width=15, fig.height=8}

pi.df %>% 
  ggplot(aes(x = factor(Tissue, level = tissue.order), y = pi, fill = Type)) + 
    geom_bar(position = position_dodge(), stat = "identity") + 
    xlab("Tissue") + 
    ylab(expression(paste("Nucleotide Diversity (", pi, ")", sep = ""))) +
    labs(fill="SNPs included:") +
    theme_bw(20) +
    theme(axis.text.x = element_text(angle = 45,  hjust=1)) +
    theme(legend.position = "bottom")

```

How does this compare to the level of Measles RNA in each compartment? It's interesting that the diversity is so high in compartments like the Cerebellum that have such low Measles RNA. It's possible that this has to do with a higher proportion of false-positive SNPs owing to the low read depth. 

```{r fig.align='center', fig.width=8, fig.height=6}

# Back-of-the-hand measles viral reads % for now: 
percent.mev.df = data.frame(Tissue = tissue.order,
                            Percent_MeV = c(19.18, 12.48, 19.29, 11.64, 20.12, 12.01, 9.32, 3.18, 0.11, 0.53, 5.08, 4.82, 2.29))


# With everything included
pi.df %>% 
  filter(Type == "everything") %>% 
  left_join(., percent.mev.df, by = "Tissue") %>% 
  ggplot(aes(x = Percent_MeV, y = pi)) + 
    geom_point(size = 1.5) + 
    geom_text_repel(aes(label = Tissue)) + 
    ggtitle("With all SNPs included") + 
    ylab(expression(paste("Nucleotide Diversity (", pi, ")", sep = ""))) +
    xlab("Measles RNA (%)") + 
    theme_bw(16) + 
    theme(plot.title = element_text(hjust = 0.5))  

# Without genome-1 and genome-2 
pi.df %>% 
  filter(Type == "no g1 or g2") %>% 
  left_join(., percent.mev.df, by = "Tissue") %>% 
  ggplot(aes(x = Percent_MeV, y = pi)) + 
    geom_point(size = 1.5) + 
    geom_text_repel(aes(label = Tissue)) + 
    ggtitle("Without G1/G2/G1-1 SNPs") + 
    ylab(expression(paste("Nucleotide Diversity (", pi, ")", sep = ""))) +
    xlab("Measles RNA (%)") + 
    theme_bw(16) + 
    theme(plot.title = element_text(hjust = 0.5))  

# With only subclonal SNPs
pi.df %>% 
  filter(Type == "only subclonal") %>% 
  left_join(., percent.mev.df, by = "Tissue") %>% 
  ggplot(aes(x = Percent_MeV, y = pi)) + 
    geom_point(size = 1.5) + 
    geom_text_repel(aes(label = Tissue)) + 
    ggtitle("With only subclonal SNPs") + 
    ylab(expression(paste("Nucleotide Diversity (", pi, ")", sep = ""))) +
    xlab("Measles RNA (%)") + 
    theme_bw(16) + 
    theme(plot.title = element_text(hjust = 0.5))  


```








